// @arcgis/charts-components/dist/esm/arcgis-charts-config-bar-chart-series-popover-f767197a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { r as registerInstance, c as createEvent, g as getAssetPath, h, H as Host, a as getElement } from './index-accb48c0.js';
import { M as Mn, A as eu, L as Hn } from './rest-js-object-literals-a5a8eefc.js';
import { U as UIDefaults } from './chart-ui-utils-0ccca40a.js';

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    panelChildrenContainer: "panel-children-container",
};

const arcgisChartsConfigBarChartSeriesPopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.host{overflow-y:auto;width:-moz-min-content;width:min-content}.panel-children-container{display:flex;padding:0 1rem;margin-top:1rem}.panel-children-container calcite-label{flex:1 1 0px;display:flex;width:0px}";
const ArcgisChartsConfigBarChartSeriesPopoverStyle0 = arcgisChartsConfigBarChartSeriesPopoverCss;

const ArcgisChartsConfigBarChartSeriesPopover = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.arcgisChartsConfigSeriesPopoverNameChange = createEvent(this, "arcgisChartsConfigSeriesPopoverNameChange", 7);
        this.arcgisChartsConfigSeriesPopoverColorChange = createEvent(this, "arcgisChartsConfigSeriesPopoverColorChange", 7);
        this.arcgisChartsConfigPopoverClose = createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        this.arcgisChartsConfigPopoverOpenChange = createEvent(this, "arcgisChartsConfigPopoverOpenChange", 7);
        /**
         * @category Event handler
         */
        this.seriesNameChange = (event) => {
            const name = event.target.value;
            if (this.selectedName !== name) {
                this.selectedName = name;
                this.arcgisChartsConfigSeriesPopoverNameChange.emit({ name });
            }
        };
        /**
         * @category Event handler
         */
        this.closePopover = () => {
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit();
        };
        /**
         * @category Event handler
         */
        this.colorChange = (e) => {
            const color = e instanceof CustomEvent ? e.detail : e;
            const [r, g, b] = color;
            const a = Hn * 255;
            this.selectedColor = [r, g, b, a];
            this.arcgisChartsConfigSeriesPopoverColorChange.emit({ color: [r, g, b, a] });
        };
        this.referenceElement = undefined;
        this.selectedName = undefined;
        this.selectedColor = undefined;
        this.placement = undefined;
        this.open = true;
    }
    /**
     * Watch on open popover.
     * @category Private
     */
    openChange() {
        this.arcgisChartsConfigPopoverOpenChange.emit({ open: this.open });
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await Mn(this.hostElement, getAssetPath(`.`)));
    }
    /**
     * @category Lifecycle
     */
    componentDidLoad() {
        this.openChange();
    }
    /**
     * @category Lifecycle
     */
    componentDidUpdate() {
        setTimeout(() => {
            this.popoverElement?.reposition();
        }, UIDefaults.PopoverTimer);
    }
    componentDidRender() {
        eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    render() {
        const rgba = this.selectedColor;
        return (h(Host, { key: '37911103ca43fcb8b76e173e91e869cdb41f8861', class: HTMLClasses.jsAppFlyout }, h("calcite-popover", { key: '9c947bea170f6ae44ff250fd1196d098a3052b2a', referenceElement: this.referenceElement, placement: this.placement, open: this.open, onCalcitePopoverClose: this.closePopover, ref: (e) => {
                this.popoverElement = e;
            }, label: "" }, h("calcite-panel", { key: 'b76203fbb93c091ebe1cd506cdf935c1e60799f4', heading: this.strings.seriesElement, closable: true, closed: !this.open, onCalcitePanelClose: this.closePopover }, h("div", { key: '88c83c4667ccefa4c8a2494d70b3e0c8d67d76fc', class: HTMLClasses.panelChildrenContainer }, h("calcite-label", { key: '5044f9cbad5ea9ad99ce6022f6664944a5f5a759' }, this.strings.label, h("calcite-input", { key: '13f390d339f2b0f7603a56e440cb3642bac63081', type: "text", value: this.selectedName, onCalciteInputInput: this.seriesNameChange }))), h("div", { key: '8a384652c750f78a25cc65862c69c0d5a87c37f3', class: HTMLClasses.panelChildrenContainer }, h("arcgis-charts-config-colors", { key: '5593755f47252b4f2bec85134b268101b4950fe6', heading: this.strings.symbolColor, rgba: rgba, onColorSelect: this.colorChange }))))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return getElement(this); }
    static get watchers() { return {
        "open": ["openChange"]
    }; }
};
ArcgisChartsConfigBarChartSeriesPopover.style = ArcgisChartsConfigBarChartSeriesPopoverStyle0;

export { ArcgisChartsConfigBarChartSeriesPopover as A };
