// @arcgis/charts-components/dist/esm/interfaces-f2a30723.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { R as RESTStatisticType } from './rest-js-object-literals-a5a8eefc.js';

const Slot = {
    actionBar: "action-bar",
    customActionTop: "custom-action-top",
    customActionBelow: "custom-action-below",
    expandTooltip: "expand-tooltip",
};
const HistogramOverlays = {
    Mean: "mean",
    Median: "median",
    StandardDev: "standardDev",
    NormalDist: "normalDist",
};
const HistogramStatistics = {
    Mean: "mean",
    Median: "median",
    StandardDev: "standardDev",
    Count: RESTStatisticType.Count,
    CountExcludingNulls: "countExcludingNulls",
    Min: RESTStatisticType.Minimum,
    Max: RESTStatisticType.Maximum,
    Sum: RESTStatisticType.Sum,
    Nulls: "nulls",
};
const ChartAuthoringPages = {
    Data: "data",
    Series: "series",
    Slices: "slices",
    Axes: "axes",
    Guides: "guides",
    Format: "format",
    General: "general",
};
const FormatIcons = {
    Title: "title",
    Text: "text",
    Label: "label",
    Palette: "palette",
};
const LineChartDataPopoverKinds = {
    xAxisField: "xAxisField",
    aggregation: "aggregation",
    numericFields: "numericField",
    splitByField: "splitByField",
    interval: "interval",
    intervalAlignment: "intervalAlignment",
};
const BarChartDataPopoverKinds = {
    category: "category",
    aggregation: "aggregation",
    numericFields: "numericField",
    splitByField: "splitByField",
    sortOrder: "sortOrder",
};
const PieChartDataPopoverKinds = {
    category: "category",
    aggregation: "aggregation",
    numericFields: "numericFields",
    sortOrder: "sortOrder",
};
const HistogramDataPopoverKinds = {
    numericFields: "numericField",
    transformation: "transformation",
    fillSymbol: "fillSymbol",
    lineSymbol: "lineSymbol",
};
const ScatterPlotDataPopoverKinds = {
    xAxisField: "xAxis",
    yAxisField: "yAxis",
    sizeVariable: "sizeVariable",
    symbol: "symbol",
    trendLine: "trendLine",
    none: "none",
};
const BoxPlotDataPopoverKinds = {
    category: "category",
    numericFields: "numericField",
    splitByField: "splitByField",
    sortOrder: "sortOrder",
};
const ScatterPlotSymbolSizeTypes = {
    Fixed: "fixed",
    Proportional: "proportional",
};
const ButtonDirections = {
    Up: "up",
    Down: "down",
};

export { BarChartDataPopoverKinds as B, ChartAuthoringPages as C, FormatIcons as F, HistogramDataPopoverKinds as H, LineChartDataPopoverKinds as L, PieChartDataPopoverKinds as P, Slot as S, BoxPlotDataPopoverKinds as a, ScatterPlotDataPopoverKinds as b, ButtonDirections as c, HistogramOverlays as d, HistogramStatistics as e, ScatterPlotSymbolSizeTypes as f };
