// @arcgis/charts-components/dist/esm/t9n-strings-a294f59f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { a2 as Dn } from './rest-js-object-literals-a5a8eefc.js';
import { g as getAssetPath } from './index-accb48c0.js';

/**
 * Fetches strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchStrings(props) {
    const { htmlElement, relativePath, filePrefix } = props;
    const assetsPath = getAssetPath(relativePath);
    return await Dn({
        element: htmlElement,
        baseURL: assetsPath,
        filePrefix,
    });
}
/**
 * Fetches the error strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchErrorsStrings(htmlElement) {
    return await fetchStrings({
        htmlElement,
        relativePath: "assets/errors",
        filePrefix: "errors",
    });
}
/**
 * Fetches the error strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchCommonStrings(htmlElement) {
    return await fetchStrings({
        htmlElement,
        relativePath: "assets/common",
        filePrefix: "common",
    });
}

export { fetchErrorsStrings as a, fetchCommonStrings as f };
