// @arcgis/charts-js-common/dist/commonjs/messages.cs-f7dfa0ea-a33d2aac.js
'use strict';

const a="Při načítání grafu došlo k chybě.",e="Nelze uplatnit logaritmickou transformaci na záporné nebo nulové hodnoty.",o="Nelze uplatnit transformaci odmocniny na záporné hodnoty.",t="Při načítání vrstvy došlo k chybě. URL = ${ url }. ID položky portálu = ${ portalItemId }.",n="Pole [${ fieldNames }] definovaná v konfiguraci nebyla nalezena ve vrstvě.",r="${ dataPath } musí být jedinečné. Řada s názvem ${ seriesName } má ID (${ seriesID }), které již používá jiná řada.",i="${ dataPath } nemůže provádět jiné seskupení než podle počtu na jiném než numerickém poli.",s="V ${ dataPath } chybí vlastnost s názvem ${ missingProperty }.",l="${ dataPath } nesmí být kratší než ${ limit } znaků.",d="${ dataPath } nesmí obsahovat méně než ${ limit } položek.",m="${ dataPath } nesmí obsahovat více než ${ limit } položek.",p="${ dataPath } musí obsahovat alespoň jeden znak jiný než mezera.",u="${ dataPath } nesmí obsahovat ${ additionalProperty }.",v="${ dataPath } se musí rovnat jedné z těchto povolených hodnot: ${ allowedValues }.",h="${ dataPath } se musí shodovat s jedním z těchto schémat: ${ schemaOptions }.",y="Rozptýlené bodové grafy s proporcionálními symboly nejsou podporovány. Byla použita výchozí velikost symbolu.",$="Vstupní data se nepodařilo přečíst.",c="Očekávaný typ řady s indexem ${ seriesIndex } je '${ expectedType }', ale obdrželi jsme '${ receivedType }'",b="nebo";var P={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:o,layerLoadFailure:t,layerFieldsNotFound:n,duplicateSeriesID:r,nonNumericAggregation:i,requiredProperty:s,minLength:l,minItems:d,maxItems:m,whiteSpacePattern:p,additionalProperty:u,enumValues:v,anyOfValues:h,bubbleChartValidateMsg:y,queryError:$,invalidSeriesType:c,or:b};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = y;
exports["default"] = P;
exports.defaultError = a;
exports.duplicateSeriesID = r;
exports.enumValues = v;
exports.invalidSeriesType = c;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = o;
exports.nonNumericAggregation = i;
exports.or = b;
exports.queryError = $;
exports.requiredProperty = s;
exports.whiteSpacePattern = p;
