// @arcgis/charts-js-common/dist/commonjs/messages.lv-90fae237-165bba33.js
'use strict';

const a="Ielādējot diagrammu, radās kļūda.",t="Žurnālfaila transformāciju nevar piemērot negatīvām vai nulles vērtībām.",e="Kvadrātsaknes transformāciju nevar lietot negatīvām vērtībām.",i="Ielādējot slāni, radās kļūda. Vietrādis (URL) = ${ url }. Portāla vienības ID = ${ portalItemId }.",r="Konfigurācijā definētie lauki [${ fieldNames }] slānī nav atrasti.",s="${ dataPath } jābūt unikālam. Sērijai ar nosaukumu ${ seriesName } ir ID (${ seriesID }), ko jau izmanto cita sērija.",n="${ dataPath } nevar veikt apkopošanu laukā, kurā nav skaitlisku vērtību.",m="${ dataPath } trūkst parametra ar nosaukumu ${ missingProperty }.",d="${ dataPath } nedrīkst būt īsāks par ${ limit } rakstzīmēm.",o="${ dataPath } nedrīkst būt mazāk par ${ limit } vienībām.",l="${ dataPath } nedrīkst būt vairāk par ${ limit } vienībām.",u="${ dataPath } jābūt vismaz vienai rakstzīmei, kas nav atstarpe.",v="${ dataPath } nedrīkst būt ${ additionalProperty }.",k="${ dataPath } jābūt vienādam ar kādu no šīm atļautajām vērtībām: ${ allowedValues }.",p="${ dataPath } jāatbilst kādai no šādām shēmām: ${ schemaOptions }.",$="Izkliedes diagrammas ar proporcionāliem simboliem netiek atbalstītas. Ticis piemērots noklusējuma simbola izmērs.",b="Neizdevās nolasīt ievades datus.",P="Indeksa ${ seriesIndex } prognozētais sērijas veids ir “${ expectedType }”, bet tā vietā tika saņemts “${ receivedType }”",g="vai";var h={defaultError:a,negativeValueInDataForLogTransformation:t,negativeValueInDataForSqrtTransformation:e,layerLoadFailure:i,layerFieldsNotFound:r,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:m,minLength:d,minItems:o,maxItems:l,whiteSpacePattern:u,additionalProperty:v,enumValues:k,anyOfValues:p,bubbleChartValidateMsg:$,queryError:b,invalidSeriesType:P,or:"vai"};

exports.additionalProperty = v;
exports.anyOfValues = p;
exports.bubbleChartValidateMsg = $;
exports["default"] = h;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = k;
exports.invalidSeriesType = P;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = i;
exports.maxItems = l;
exports.minItems = o;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = t;
exports.negativeValueInDataForSqrtTransformation = e;
exports.nonNumericAggregation = n;
exports.or = g;
exports.queryError = b;
exports.requiredProperty = m;
exports.whiteSpacePattern = u;
