// @arcgis/charts-js-common/dist/commonjs/messages.vi-7bf62f1e-99ad8fee.js
'use strict';

const t="Đã xảy ra lỗi khi tải biểu đồ.",n="Không thể áp dụng phép biến đổi log lên giá trị âm hoặc bằng không.",h="Không thể áp dụng phép biến đổi căn bậc hai lên giá trị âm.",a="Đã xảy ra lỗi khi tải lớp. URL = ${ url }. ID mục cổng thông tin = ${ portalItemId }.",i="Không tìm thấy trong lớp này các trường [${ fieldNames }] được xác định trong cấu hình.",e="${ dataPath } phải là duy nhất. Chuỗi có tên ${ seriesName } có một ID (${ seriesID }) đã được chuỗi khác sử dụng.",r="${ dataPath } không thể thực hiện tổng hợp không đếm trên trường không phải là số.",g="${ dataPath } đang thiếu một tên phù hợp ${ missingProperty }.",c="${ dataPath } không được ngắn hơn ${ limit } ký tự.",l="${ dataPath } không được chứa dưới ${ limit } mục.",o="${ dataPath } không được chứa nhiều hơn ${ limit } mục.",d="${ dataPath } phải chứa ít nhất một ký tự không phải khoảng trắng.",p="${ dataPath } không được chứa ${ additionalProperty }.",m="${ dataPath } phải bằng một trong những giá trị được phép này: ${ allowedValues }.",u="${ dataPath } phải khớp với lược đồ của một trong những thứ sau: ${ schemaOptions }.",s="Biểu đồ phân tán có ký hiệu tỷ lệ không được hỗ trợ. Đã áp dụng kích cỡ ký hiệu theo mặc định.",$="Không thể đọc dữ liệu đầu vào.",k="Loại chuỗi mong đợi tại chỉ mục ${ seriesIndex } là '${ expectedType }' nhưng thay vào đó lại nhận được '${ receivedType }'",y="hoặc";var P={defaultError:t,negativeValueInDataForLogTransformation:n,negativeValueInDataForSqrtTransformation:h,layerLoadFailure:a,layerFieldsNotFound:i,duplicateSeriesID:e,nonNumericAggregation:r,requiredProperty:g,minLength:c,minItems:l,maxItems:o,whiteSpacePattern:d,additionalProperty:p,enumValues:m,anyOfValues:u,bubbleChartValidateMsg:s,queryError:$,invalidSeriesType:k,or:y};

exports.additionalProperty = p;
exports.anyOfValues = u;
exports.bubbleChartValidateMsg = s;
exports["default"] = P;
exports.defaultError = t;
exports.duplicateSeriesID = e;
exports.enumValues = m;
exports.invalidSeriesType = k;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = a;
exports.maxItems = o;
exports.minItems = l;
exports.minLength = c;
exports.negativeValueInDataForLogTransformation = n;
exports.negativeValueInDataForSqrtTransformation = h;
exports.nonNumericAggregation = r;
exports.or = y;
exports.queryError = $;
exports.requiredProperty = g;
exports.whiteSpacePattern = d;
