// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart-data2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { proxyCustomElement, HTMLElement, createEvent, getAssetPath, h, Host } from '@stencil/core/internal/client';
import { a0 as wc, M as Mn, w as ta, A as Tl, a1 as Cl, a as ts, m as ml, Z as Zy, r as Qy, a2 as Jy } from './index2.js';
import { J as Jf } from './index3.js';
import { B as BarChartDataPopoverKinds, R as RESTStatisticType, f as ButtonDirections } from './interfaces.js';
import { d as deleteMinMaxAxisKey, s as state } from './store.js';
import { U as UIDefaults } from './chart-ui-utils.js';
import { d as defineCustomElement$1 } from './arcgis-charts-config-field-select2.js';
import { i as isEqual } from './isEqual.js';
import { d as defaultBarLineFromFieldsName } from './config-defaults.js';
import { W as WebChartTypes } from './chart-object-literals.js';

const HTMLClasses = {
    label: "label",
    text: "text",
    fieldList: "field-list",
    fab: "fab",
    disabledLabel: "disabled-label",
    rtl: "arcgis-charts-rtl",
    list: "list",
    customSortButtons: "custom-sort-buttons",
    dimText: "dim-text",
};

const arcgisChartsConfigBarChartDataCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin);margin-top:var(--arcgis-charts-top-section-margin)}.label,.disabled-label{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:var(--arcgis-charts-config-margin-minor);justify-content:space-between}.text{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);width:calc(100% - 1.5rem);height:2rem;margin-bottom:var(--arcgis-charts-config-margin-minor);justify-content:space-between}.field-list{max-height:50vh;overflow-y:auto}.disabled-label{pointer-events:none}.fab{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:1.5rem;justify-content:center}:host(.arcgis-chart-rtl) .text{margin-left:0rem;margin-right:var(--arcgis-charts-config-margin-minor)}.list{margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:var(--arcgis-charts-config-margin-major)}.custom-sort-buttons{display:flex;justify-content:space-between;align-items:center;font-weight:var(--calcite-font-weight-medium);font-size:var(--calcite-font-size-0)}";
const ArcgisChartsConfigBarChartDataStyle0 = arcgisChartsConfigBarChartDataCss;

const ArcGISChartsConfigBarChartData = /*@__PURE__*/ proxyCustomElement(class ArcGISChartsConfigBarChartData extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.arcgisChartsConfigBarChartDataNumericFieldValidate = createEvent(this, "arcgisChartsConfigBarChartDataNumericFieldValidate", 7);
        this.arcgisChartsConfigDataChangeError = createEvent(this, "arcgisChartsConfigDataChangeError", 7);
        /**
         * @category Private
         */
        this.barChartPopover = null;
        /**
         * @category Private
         * List of original custom sort labels
         * Used to reset the custom sort labels
         */
        this.originalCustomSortLabels = [];
        /**
         * @category Private
         * List of custom sort labels
         */
        this.customSortLabels = [];
        /**
         * @category Private
         *
         * Flag to disable the custom sort list.
         */
        this.isCustomSortDisabled = false;
        /**
         * @category Event handler
         */
        this.onCalciteListOrderChange = (e) => {
            if (e.detail !== null && e.detail !== undefined) {
                const { newIndex, oldIndex } = e.detail;
                this.model.moveCustomSortValuesInOrderByList(oldIndex, newIndex);
                this.customSortLabels = this.model.getOrderByList();
            }
        };
        /**
         * @category Event handler
         */
        this.onCustomSortItemSelected = (e) => {
            const { id } = e.target;
            // unselect the custom sort item in the custom sort list if it is already selected
            this.selectedCustomSortItemId = this.selectedCustomSortItemId === id ? "" : id;
        };
        /**
         * @category Event handler
         */
        this.onCustomSortItemMoved = (e) => {
            const { id } = e.target;
            let toIndex = this.customSortLabels.length;
            const fromIndex = this.customSortLabels.indexOf(this.selectedCustomSortItemId);
            // calculate the index to move the selected custom sort item to
            if (id === ButtonDirections.Up) {
                toIndex = fromIndex - 1;
            }
            else if (id === ButtonDirections.Down) {
                toIndex = fromIndex + 1;
            }
            // move the selected item to the calculated index only within the valid list bound
            if (fromIndex >= 0 && toIndex >= 0 && toIndex < this.customSortLabels.length) {
                this.model.moveCustomSortValuesInOrderByList(fromIndex, toIndex);
                this.customSortLabels = this.model.getOrderByList();
            }
        };
        /**
         * @category Event handler
         *
         * called when the aggregation type is clicked
         */
        this.onAggregationSelect = () => {
            this.contentKind = BarChartDataPopoverKinds.aggregation;
            this.setupPopover();
        };
        /**
         * @category Event handler
         *
         * called when the sort order is clicked
         */
        this.onSortOrderSelect = () => {
            this.contentKind = BarChartDataPopoverKinds.sortOrder;
            this.setupPopover();
        };
        /**
         * @category Event handler
         *
         * called when the category is clicked
         */
        this.onCategorySelect = () => {
            this.contentKind = BarChartDataPopoverKinds.category;
            this.setupPopover();
        };
        /**
         * @category Event handler
         *
         * called when the split-by is clicked
         */
        this.onSplitBySelect = () => {
            if (!this.splitByFieldElement.disabled) {
                this.contentKind = BarChartDataPopoverKinds.splitByField;
                this.setupPopover();
            }
        };
        /**
         * @category Event handler
         *
         * called when the add numeric fields button is clicked
         */
        this.onAddNumericFields = () => {
            this.contentKind = BarChartDataPopoverKinds.numericFields;
            this.setupPopover();
        };
        /**
         * @category Event handler
         *
         * Called when individual numeric field item is removed.
         */
        this.onRemoveNumericField = async (removedItemEvent) => {
            const removedField = removedItemEvent.target.value;
            await this.model.setNumericFields(this.model.getNumericFields().filter((field) => field !== removedField));
            // update custom sort's value list if the numeric fields are changed and sort order is custom sort
            if (this.model.getSortOrder() === Jf.customSort) {
                await this.updateCustomSortProps();
                this.model.setOrderByList(this.customSortLabels);
            }
        };
        /**
         * @category Event handler
         *
         * called in response to the arcgisChartsConfigBarChartDataPopoverChange event
         */
        this.onBarChartDataChange = async (e) => {
            try {
                if (e.detail !== null && e.detail?.value !== undefined) {
                    const { contentKind } = e.detail;
                    const newValue = [...e.detail.value];
                    // reset axes bounds for this chart in stencil store
                    deleteMinMaxAxisKey(state);
                    switch (contentKind) {
                        case BarChartDataPopoverKinds.aggregation:
                            await this.model.setAggregationType(newValue[0] !== RESTStatisticType.NoAggregation ? newValue[0] : RESTStatisticType.NoAggregation);
                            break;
                        case BarChartDataPopoverKinds.category: {
                            const newCategory = newValue[0]?.trim() || "";
                            await this.model.setXAxisField(newCategory);
                            // update custom sort's value list if the category field is changed
                            await this.updateCustomSortProps();
                            this.model.setOrderByList(this.customSortLabels);
                            break;
                        }
                        case BarChartDataPopoverKinds.numericFields:
                            await this.model.setNumericFields(newValue);
                            if (this.model.getNumericFields().length > 0 &&
                                this.model.getAggregationType() === RESTStatisticType.Count &&
                                this.model.getNumericFields()[0] !== wc(this.layerFieldsInfo)) {
                                await this.model.setAggregationType(RESTStatisticType.Sum);
                            }
                            // update custom sort's value list if the numeric fields are changed and sort order is custom sort
                            if (this.model.getSortOrder() === Jf.customSort) {
                                await this.updateCustomSortProps();
                                this.model.setOrderByList(this.customSortLabels);
                            }
                            break;
                        case BarChartDataPopoverKinds.splitByField: {
                            const newSplitByField = newValue[0]?.trim() || "";
                            await this.model.setSplitByField(newSplitByField);
                            this.model.setColorMatch(true);
                            break;
                        }
                        case BarChartDataPopoverKinds.sortOrder: {
                            const newSortOrder = newValue[0];
                            if (newSortOrder === Jf.customSort) {
                                await this.updateCustomSortProps();
                                this.model.setSortOrder(newSortOrder, this.customSortLabels);
                            }
                            else {
                                this.model.setSortOrder(newSortOrder);
                            }
                            break;
                        }
                    }
                }
            }
            catch (error) {
                this.arcgisChartsConfigDataChangeError.emit({ error, model: this.model });
                throw error;
            }
            this.removeBarChartDataPopover();
            const prevUpdates = this.updatesOccurred;
            this.updatesOccurred = prevUpdates + 1;
        };
        /**
         * @category Event handler
         */
        this.onShowDataLabelsChange = (e) => {
            this.model.setDataLabelsVisibility(e.target.checked);
        };
        /**
         * @category Private
         *
         * Update custom sort properties
         */
        this.updateCustomSortProps = async () => {
            const customSortValues = await this.model.getCustomSortValues();
            this.customSortLabels = customSortValues;
            this.originalCustomSortLabels = [...customSortValues];
            // enable/disable the custom sort list based on the number of custom sort labels
            const isCustomSortDisabled = customSortValues.length > UIDefaults.customSortLimit;
            if (isCustomSortDisabled) {
                this.sortOrderElement.label = this.strings.xAxisAsc;
                this.model.setSortOrder(Jf.xAxisAsc);
            }
            this.isCustomSortDisabled = isCustomSortDisabled;
        };
        /**
         * Event handler when model config changes.
         * @category Event handler
         */
        this.onModelConfigChange = () => {
            const prevUpdates = this.updatesOccurred;
            this.updatesOccurred = prevUpdates + 1;
        };
        /**
         * Close bar chart popover
         * @category Private
         */
        this.popoverClose = (e) => {
            if (this.barChartPopover === e?.target) {
                this.removeBarChartDataPopover();
            }
        };
        /**
         * @category Private
         */
        this.resetToOriginalCustomSortLabels = () => {
            this.selectedCustomSortItemId = "";
            this.model.setOrderByList([...this.originalCustomSortLabels]);
        };
        this.model = undefined;
        this.layerFieldsInfo = undefined;
        this.open = true;
        this.selectedCustomSortItemId = undefined;
        this.updatesOccurred = 0;
    }
    /**
     * @category Watch handler
     */
    modelChange(newModel, oldModel) {
        // config change handler for chart model
        oldModel?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
        newModel?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.removeBarChartDataPopover();
        this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings, direction: this.direction } = await Mn(this.hostElement, getAssetPath(`.`)));
        this.modelChange(this.model);
        this.onModelConfigChange();
        this.currentDataFilter = this.model.getDataFilter();
        this.updateCustomSortProps();
    }
    /**
     * @category Lifecycle
     */
    async componentWillRender() {
        this.aggregationLabel = await ta(this.model.getAggregationType());
        const savedDataFilter = this.model.getDataFilter();
        if (!isEqual(savedDataFilter, this.currentDataFilter)) {
            this.updateCustomSortProps();
            this.currentDataFilter = this.model.getDataFilter();
        }
        else {
            this.customSortLabels = this.model.getOrderByList();
        }
        this.disableSplitBy =
            this.model.getNumericFields().length > 1 ||
                this.model.getXAxisField() === "" ||
                this.model.getXAxisField() === defaultBarLineFromFieldsName;
    }
    /**
     * @category Render UX
     * @param field
     */
    buildNumericFieldItem(field) {
        return (h("calcite-list-item", { label: Tl(this.layerFieldsInfo, field), value: field, key: field, closable: true, onCalciteListItemClose: this.onRemoveNumericField }));
    }
    /**
     * @category Render UX
     */
    buildCustomSortListItems() {
        const categoryListItems = [];
        for (let index = 0; index < this.customSortLabels.length; index += 1) {
            const label = this.customSortLabels[index];
            const description = Cl(this.layerFieldsInfo, label);
            categoryListItems.push(h("calcite-list-item", { id: label, key: label, label: label, description: description, value: label, selected: this.selectedCustomSortItemId === label, onCalciteListItemSelect: this.onCustomSortItemSelected }, h("calcite-icon", { slot: "content-end" })));
        }
        return categoryListItems;
    }
    /**
     * @category Render UX
     */
    buildCustomSortList() {
        return (h("calcite-list", { label: "", class: HTMLClasses.list, dragEnabled: true, selectionMode: "single", selectionAppearance: "border", onCalciteListOrderChange: this.onCalciteListOrderChange }, this.buildCustomSortListItems()));
    }
    /**
     * @category Render UX
     */
    renderSorting() {
        return (h("arcgis-charts-config-field-select", { disabled: this.model.getXAxisValueFormat()?.type === WebChartTypes.DateAxisFormat, label: this.model.getXAxisValueFormat()?.type === WebChartTypes.DateAxisFormat
                ? this.strings.xAxisAsc
                : this.strings[this.model.getSortOrder()], ref: (e) => {
                this.sortOrderElement = e;
            }, onClick: this.onSortOrderSelect }));
    }
    /**
     * @category Render UX
     */
    renderAggregation() {
        return (h("arcgis-charts-config-field-select", { label: this.model.getAggregationType() === RESTStatisticType.NoAggregation
                ? this.strings.noAggregation
                : this.aggregationLabel, ref: (e) => {
                this.aggregationElement = e;
            }, onClick: this.onAggregationSelect }));
    }
    /**
     * @category Render UX
     */
    renderCategory() {
        const categoryString = this.getCorrectedCategoryValue();
        return (h("arcgis-charts-config-field-select", { label: categoryString, ref: (e) => {
                this.categoryElement = e;
            }, onClick: this.onCategorySelect }));
    }
    /**
     * @category Render UX
     */
    renderSplitBy() {
        const splitByString = this.getCorrectedSplitByValue();
        return (h("arcgis-charts-config-field-select", { label: splitByString, disabled: this.disableSplitBy, ref: (e) => {
                this.splitByFieldElement = e;
            }, onClick: this.onSplitBySelect }));
    }
    /**
     * @category Render UX
     */
    renderNumericFields(fields) {
        const objectIdField = wc(this.layerFieldsInfo);
        const fieldList = [];
        for (let i = 0; i < fields.length; i += 1) {
            if (fields[i] !== objectIdField) {
                fieldList.push(this.buildNumericFieldItem(fields[i]));
            }
        }
        return (h("calcite-list", { label: "", class: HTMLClasses.fieldList, ref: (e) => {
                this.numericFieldsElement = e;
            } }, fieldList));
    }
    /**
     * @category Render UX
     */
    renderCustomSortUI() {
        const sortOrder = this.model.getSortOrder();
        const valueType = this.model.getXAxisValueFormat()?.type;
        const index = this.customSortLabels.indexOf(this.selectedCustomSortItemId);
        const customSortListUI = valueType ===
            WebChartTypes.CategoryAxisFormat /*render the custom sort for the field with value format set to CategoryAxisFormat */ &&
            sortOrder === Jf.customSort &&
            !this.isCustomSortDisabled ? (h("calcite-label", { class: HTMLClasses.label }, h("div", { class: HTMLClasses.customSortButtons }, h("div", null, h("calcite-button", { appearance: "transparent", disabled: index === 0 || index === -1 || this.selectedCustomSortItemId.trim() === "", iconStart: "arrow-bold-up", id: ButtonDirections.Up, kind: "neutral", scale: "s", onClick: this.onCustomSortItemMoved }), h("calcite-button", { appearance: "transparent", disabled: index === -1 ||
                index === this.customSortLabels.length - 1 ||
                this.selectedCustomSortItemId.trim() === "", iconStart: "arrow-bold-down", id: ButtonDirections.Down, kind: "neutral", scale: "s", onClick: this.onCustomSortItemMoved })), h("calcite-button", { onClick: this.resetToOriginalCustomSortLabels, disabled: isEqual(this.customSortLabels, this.originalCustomSortLabels), appearance: "transparent", iconEnd: "reset" }, this.strings.reset)), ts(this.model?.getConfig().series ?? []) === ml.BarAndLineFromFields
            ? this.strings.values
            : this.strings.category, this.buildCustomSortList())) : null;
        return customSortListUI;
    }
    /**
     * @category Lifecycle
     */
    render() {
        const aggregationUI = this.renderAggregation();
        const categoryUI = this.renderCategory();
        const splitByUI = this.renderSplitBy();
        const sortingUI = this.renderSorting();
        const numericFieldsUI = this.renderNumericFields([...this.model.getNumericFields()]);
        const customSortUI = this.renderCustomSortUI();
        return (h(Host, { key: 'd520b608238fe23846bb5b787944a6cebed1e97a', class: { [HTMLClasses.rtl]: this.direction === "rtl" } }, h("calcite-label", { key: '9d87db83f1a18be036e0ec77960db5e02bdb521c', class: HTMLClasses.label }, this.strings.category, categoryUI), h("calcite-label", { key: '1d48bd128c8c1df87d363021fabfb389f3a20ee6', class: HTMLClasses.label }, this.strings.aggregation, aggregationUI), h("calcite-label", { key: 'cd6d29c45d415f6bdc2ebe8c19540bb2536c42a5', class: HTMLClasses.label }, this.strings.numericFields, numericFieldsUI), h("div", { key: '576fd1a43d78a937bcb1698ef3d5d1f393c119e8', class: HTMLClasses.fab }, h("calcite-fab", { key: '31389b002023db711664d1259908efa14ebac2ef', appearance: "outline-fill", kind: "neutral", scale: "m", label: this.strings.selectNumericFields, "text-enabled": true, text: this.strings.selectNumericFields, icon: "plus", onClick: this.onAddNumericFields })), h("calcite-label", { key: '56536812fd8ba83ab3560588c6e4fb5584fee414', class: HTMLClasses.label }, h("span", { key: '22802f0cb6286e8e4ff80feb6b90ed6604622670', class: this.disableSplitBy ? HTMLClasses.dimText : "" }, this.strings.splitBy), splitByUI), h("calcite-label", { key: 'be78b831308f8c819475749a0cb4bb6fcb17e2b0', class: HTMLClasses.label, layout: "inline-space-between" }, this.strings.showDataLabels, h("calcite-switch", { key: '3cbc95d9741ea3a1e18920e0c031cc792150e701', checked: this.model.getDataLabelsVisibility(), onCalciteSwitchChange: this.onShowDataLabelsChange })), h("calcite-label", { key: '5b6e3f2996f9946b7e92aa2707c872b57884c7c3', class: this.model.getXAxisValueFormat()?.type === WebChartTypes.DateAxisFormat
                ? HTMLClasses.disabledLabel
                : HTMLClasses.label }, this.strings.sortOrder, sortingUI), customSortUI));
    }
    /**
     * Get the split by value based on the current model's split by field.
     *
     * **NOTE**: If the split by field is empty or there are more than one numeric fields, return an empty string.
     * @category Private
     * @returns {string} - The corrected split by value.
     */
    getCorrectedSplitByValue() {
        return this.model.getSplitByField() === "" || this.model.getNumericFields().length > 1
            ? ""
            : Tl(this.layerFieldsInfo, this.model.getSplitByField());
    }
    /**
     * Get the category value based on the current model's category field.
     *
     * **NOTE**: If the category field is empty or set to the default value (`arcgis_charts_bar_line_from_fields_name`),
     * return an empty string.
     * @category Private
     * @returns {string} - The corrected category value.
     */
    getCorrectedCategoryValue() {
        return this.model.getXAxisField() === defaultBarLineFromFieldsName || this.model.getXAxisField().trim() === ""
            ? ""
            : Tl(this.layerFieldsInfo, this.model.getXAxisField());
    }
    /**
     * @category Private
     */
    setupPopover() {
        // remove already existing chart popover
        // TODO: this is a hotfix for https://devtopia.esri.com/webgis/arcgis-charts/issues/5366
        // have to find a better way to do this
        this.removeBarChartDataPopover();
        // barChartPopover config
        const popoverConfig = {
            layerFieldsInfo: this.layerFieldsInfo,
            onBarChartDataChange: this.onBarChartDataChange,
            popoverClose: this.popoverClose,
        };
        // create a new barChartPopover
        this.barChartPopover = Zy(Qy.BarChartDataPopover, popoverConfig);
        if (this.barChartPopover !== null) {
            // switch content of bar chart popover based on the provided content
            switch (this.contentKind) {
                case BarChartDataPopoverKinds.aggregation:
                    this.barChartPopover.referenceElement = this.aggregationElement;
                    this.barChartPopover.headingTitle = this.strings.aggregationHeader;
                    this.barChartPopover.isNoAggregationDisabled =
                        ts(this.model?.getConfig().series ?? []) === ml.BarAndLineFromFields;
                    this.barChartPopover.selectedContent = this.model.getAggregationType();
                    break;
                case BarChartDataPopoverKinds.category:
                    this.barChartPopover.referenceElement = this.categoryElement;
                    this.barChartPopover.headingTitle = this.strings.categoryHeader;
                    this.barChartPopover.selectedContent = this.model.getXAxisField();
                    break;
                case BarChartDataPopoverKinds.numericFields:
                    this.barChartPopover.referenceElement = this.numericFieldsElement;
                    this.barChartPopover.headingTitle = this.strings.numericFieldsHeader;
                    this.barChartPopover.selectedContent = this.model.getNumericFields();
                    break;
                case BarChartDataPopoverKinds.splitByField:
                    this.barChartPopover.referenceElement = this.splitByFieldElement;
                    this.barChartPopover.headingTitle = this.strings.splitByFieldHeader;
                    this.barChartPopover.selectedContent = this.model.getSplitByField();
                    break;
                case BarChartDataPopoverKinds.sortOrder:
                    this.barChartPopover.referenceElement = this.sortOrderElement;
                    this.barChartPopover.headingTitle = this.strings.sortOrderHeader;
                    this.barChartPopover.selectedContent = this.model.getSortOrder();
                    this.barChartPopover.isCustomSortDisabled = this.isCustomSortDisabled;
                    break;
            }
            this.barChartPopover.contentKind = this.contentKind;
            this.barChartPopover.open = true;
            this.addPopover();
        }
    }
    /**
     * @category Private
     */
    addPopover() {
        if (!document.body.contains(this.barChartPopover) && this.barChartPopover !== null) {
            document.body.appendChild(this.barChartPopover);
        }
    }
    /**
     * Remove bar chart data popover
     * @category Private
     */
    removeBarChartDataPopover() {
        if (this.barChartPopover !== null) {
            this.barChartPopover.open = false;
            this.barChartPopover = Jy(this.barChartPopover);
        }
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return this; }
    static get watchers() { return {
        "model": ["modelChange"]
    }; }
    static get style() { return ArcgisChartsConfigBarChartDataStyle0; }
}, [1, "arcgis-charts-config-bar-chart-data", {
        "model": [16],
        "layerFieldsInfo": [16],
        "open": [516],
        "selectedCustomSortItemId": [32],
        "updatesOccurred": [32]
    }, undefined, {
        "model": ["modelChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-config-bar-chart-data", "arcgis-charts-config-field-select"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-config-bar-chart-data":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcGISChartsConfigBarChartData);
            }
            break;
        case "arcgis-charts-config-field-select":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { ArcGISChartsConfigBarChartData as A, defineCustomElement as d };
