// @arcgis/charts-components/dist/components/arcgis-charts-config-color-style2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { proxyCustomElement, HTMLElement, createEvent, getAssetPath, h, Host } from '@stencil/core/internal/client';
import { M as Mn } from './index2.js';
import { d as defineCustomElement$1 } from './arcgis-charts-config-colors2.js';

const HTMLClasses = {
    colorStyleContainer: "color-style-container",
};

const arcgisChartsConfigColorStyleCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{overflow-y:auto;width:-moz-min-content;width:min-content}:host calcite-panel .color-style-container{display:flex;padding:0 1rem;margin-top:1rem}:host calcite-panel .color-style-container calcite-label{flex:1 1 0px;display:flex;width:0px}.color-styling-panel>.panel-body{padding:var(--arcgis-charts-panel-padding);margin-top:var(--arcgis-charts-inter-section-margin)}";
const ArcgisChartsConfigColorStyleStyle0 = arcgisChartsConfigColorStyleCss;

const ArcgisChartsConfigColorStyle = /*@__PURE__*/ proxyCustomElement(class ArcgisChartsConfigColorStyle extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.arcgisChartsConfigColorStyleChange = createEvent(this, "arcgisChartsConfigColorStyleChange", 7);
        this.arcgisChartsConfigColorStyleClose = createEvent(this, "arcgisChartsConfigColorStyleClose", 7);
        /**
         * @category Event handler
         */
        this.closePanel = (e) => {
            const panel = e.target;
            if (panel.closed && this.enableClose) {
                this.open = false;
                this.arcgisChartsConfigColorStyleClose.emit();
            }
        };
        /**
         * @category Event handler
         */
        this.colorChange = (e) => {
            this.selectedColor = e.detail;
            this.arcgisChartsConfigColorStyleChange.emit(this.selectedColor);
        };
        this.header = undefined;
        this.selectedColor = undefined;
        this.open = undefined;
        this.enableClose = false;
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await Mn(this.hostElement, getAssetPath(`.`)));
        // T.B.D issue with lang getting set correctly when this component used as a child
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (h(Host, { key: 'b50e22358bbb59a9ff3a2a891bb1ad7d3936f5f6' }, h("calcite-panel", { key: 'f4d233ed1ae74b6dfbfc01d0ae6a6306a3f603a5', heading: this.header ?? this.strings.header, closed: this.open === false, closable: this.enableClose, onCalcitePanelClose: this.closePanel }, h("div", { key: 'b7318596eff23fe15c1f778a83624ebe19783be9', class: HTMLClasses.colorStyleContainer }, h("arcgis-charts-config-colors", { key: 'e72288a17ccc4f2a8d7a5fcae244671f9be13c78', onColorSelect: this.colorChange, rgba: this.selectedColor })))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return this; }
    static get style() { return ArcgisChartsConfigColorStyleStyle0; }
}, [1, "arcgis-charts-config-color-style", {
        "header": [1],
        "selectedColor": [1040],
        "open": [1540],
        "enableClose": [516, "enable-close"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-config-color-style", "arcgis-charts-config-colors"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-config-color-style":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisChartsConfigColorStyle);
            }
            break;
        case "arcgis-charts-config-colors":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { ArcgisChartsConfigColorStyle as A, defineCustomElement as d };
