// @arcgis/charts-components/dist/components/arcgis-charts-config-format-popover2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { proxyCustomElement, HTMLElement, createEvent, getAssetPath, h, Host } from '@stencil/core/internal/client';
import { M as Mn, y as eu } from './index2.js';
import { d as defineCustomElement$4 } from './arcgis-charts-config-color-style2.js';
import { d as defineCustomElement$3 } from './arcgis-charts-config-colors2.js';
import { d as defineCustomElement$2 } from './arcgis-charts-config-font-style2.js';
import { d as defineCustomElement$1 } from './arcgis-charts-config-line-style2.js';
import { a as RESTSymbolType } from './interfaces.js';

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
};

const arcgisChartsConfigFormatPopoverCss = "";
const ArcgisChartsConfigFormatPopoverStyle0 = arcgisChartsConfigFormatPopoverCss;

const ArcgisChartsConfigFormatPopover = /*@__PURE__*/ proxyCustomElement(class ArcgisChartsConfigFormatPopover extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.arcgisChartsConfigFormatPopoverSymbolChange = createEvent(this, "arcgisChartsConfigFormatPopoverSymbolChange", 7);
        this.arcgisChartsConfigFormatPopoverTextChange = createEvent(this, "arcgisChartsConfigFormatPopoverTextChange", 7);
        this.arcgisChartsConfigFormatPopoverColorChange = createEvent(this, "arcgisChartsConfigFormatPopoverColorChange", 7);
        this.arcgisChartsConfigPopoverClose = createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        /**
         * @category Private
         */
        this.closePopover = () => {
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit();
        };
        /**
         * @category Event handler
         */
        this.updateSymbolProperties = (e) => {
            this.selectedElement = e.target.lineStyle;
            this.arcgisChartsConfigFormatPopoverSymbolChange.emit(e.detail);
        };
        /**
         * @category Event handler
         */
        this.updateTextProperties = (e) => {
            this.selectedElement = e.target.fontStyle;
            this.arcgisChartsConfigFormatPopoverTextChange.emit(e.detail);
        };
        /**
         * @category Event handler
         */
        this.updateColorProperties = (e) => {
            this.selectedElement = e.detail;
            this.arcgisChartsConfigFormatPopoverColorChange.emit(this.selectedElement);
        };
        this.selectedElement = undefined;
        this.referenceElement = undefined;
        this.placement = "leading";
        this.open = undefined;
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await Mn(this.hostElement, getAssetPath(`.`)));
        // T.B.D issue with lang getting set correctly when this component used as a child
    }
    /**
     * @category Lifecycle
     */
    componentDidRender() {
        eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    render() {
        const colorElement = this.selectedElement;
        const symbolElement = this.selectedElement;
        return (h(Host, { key: '3a48568d1634f1cfa44e6b9e83612543164526f2', class: HTMLClasses.jsAppFlyout }, h("calcite-popover", { key: '90af72c59e152de293f078b7d84cf5deff94102e', referenceElement: this.referenceElement, ref: (e) => {
                this.popoverElement = e;
            }, placement: this.placement, open: this.open, label: "", onCalcitePopoverClose: this.closePopover, pointerDisabled: true }, Array.isArray(colorElement) ? (h("arcgis-charts-config-color-style", { header: this.strings.backgroundElement, enableClose: true, open: this.open, selectedColor: colorElement, onArcgisChartsConfigColorStyleClose: this.closePopover, onArcgisChartsConfigColorStyleChange: this.updateColorProperties })) : null, symbolElement?.type === RESTSymbolType.SLS ? (h("arcgis-charts-config-line-style", { header: this.strings.lineStyle, open: this.open, enableClose: true, lineStyle: symbolElement, onArcgisChartsConfigLineStyleClose: this.closePopover, onArcgisChartsConfigLineStyleChange: this.updateSymbolProperties })) : null, symbolElement?.type === RESTSymbolType.TS ? (h("arcgis-charts-config-font-style", { header: this.strings.textStyle, open: this.open, enableClose: true, fontStyle: symbolElement, onArcgisChartsConfigFontStyleClose: this.closePopover, onArcgisChartsConfigFontStyleChange: this.updateTextProperties })) : null)));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return this; }
    static get style() { return ArcgisChartsConfigFormatPopoverStyle0; }
}, [1, "arcgis-charts-config-format-popover", {
        "selectedElement": [1040],
        "referenceElement": [16],
        "placement": [513],
        "open": [1540],
        "reposition": [64]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-config-format-popover", "arcgis-charts-config-color-style", "arcgis-charts-config-colors", "arcgis-charts-config-font-style", "arcgis-charts-config-line-style"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-config-format-popover":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcgisChartsConfigFormatPopover);
            }
            break;
        case "arcgis-charts-config-color-style":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "arcgis-charts-config-colors":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "arcgis-charts-config-font-style":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "arcgis-charts-config-line-style":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { ArcgisChartsConfigFormatPopover as A, defineCustomElement as d };
