// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart-series2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { proxyCustomElement, HTMLElement, createEvent, getAssetPath, h, Host } from '@stencil/core/internal/client';
import { a2 as Jy, M as Mn, Z as Zy, r as Qy, m as ml } from './index2.js';
import { c as createActionTooltip, d as destroyActionTooltip } from './chart-ui-utils.js';
import { e as WebChartStackedKinds } from './chart-object-literals.js';
import { i as isNil } from './isNil.js';
import { R as RESTStatisticType } from './interfaces.js';

const HTMLIDs = {
    stackedType: "stacked-type",
};
const HTMLClasses = {
    topSection: "top-section",
    sectionLabel: "section-label",
    list: "list",
    disabledInfo: "disabled-info",
    disabledStackingOption: "disabled-stacking-option",
};

const arcgisChartsConfigBarChartSeriesCss = ":host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin)}.top-section{margin-top:var(--arcgis-charts-config-margin-minor)}.section-label{display:flex;margin-bottom:var(--arcgis-charts-config-margin-major);margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);justify-content:space-between;align-items:center;font-weight:var(--calcite-font-weight-medium);font-size:var(--calcite-font-size-0)}.list{margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:var(--arcgis-charts-config-margin-major)}.disabled-info{margin:auto 1rem;pointer-events:auto}.disabled-stacking-option{pointer-events:none;opacity:var(--calcite-opacity-disabled)}";
const ArcgisChartsConfigBarChartSeriesStyle0 = arcgisChartsConfigBarChartSeriesCss;

const ArcGISChartsConfigBarChartSeries = /*@__PURE__*/ proxyCustomElement(class ArcGISChartsConfigBarChartSeries extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.arcgisChartsConfigBarChartSeriesPopoverChange = createEvent(this, "arcgisChartsConfigBarChartSeriesPopoverChange", 7);
        this.arcgisChartsConfigPopoverOpenChange = createEvent(this, "arcgisChartsConfigPopoverOpenChange", 7);
        /**
         * @category Private
         */
        this.resetSeriesOriginal = async () => {
            this.configSeriesOpen = false;
            this.selectedSeriesIndex = undefined;
            // reset series styling
            await this.model.resetSeriesStyling();
        };
        /**
         * @category Event handler
         */
        this.onMultipleBarSelect = async (event) => {
            if (!this.shouldDisableStacking()) {
                const target = event.target;
                await this.model.setStackedType(target.value ??
                    target.parentElement.value ??
                    "");
            }
        };
        /**
         * @category Event handler
         */
        this.seriesNameChange = (event) => {
            if (event.detail !== null || event.detail !== undefined) {
                const name = event.detail?.name;
                this.model.setSeriesName(name, this.selectedSeriesIndex ?? -1);
            }
        };
        /**
         * @category Event handler
         */
        this.seriesColorChange = (e) => {
            if (e.detail !== null && e.detail !== undefined) {
                const { color } = e.detail;
                this.model.setColorMatch(false);
                if (color !== undefined) {
                    this.model.setSeriesColor(color, this.selectedSeriesIndex ?? -1);
                }
            }
        };
        /**
         * @category Event handler
         */
        this.onCalciteListOrderChange = (e) => {
            if (e.detail !== null && e.detail !== undefined) {
                const { newIndex, oldIndex } = e.detail;
                this.model.moveSeries(oldIndex, newIndex);
            }
        };
        /**
         * @category Event handler
         */
        this.seriesPopoverClose = () => {
            this.selectedSeriesIndex = undefined;
            this.configSeriesOpen = false;
            this.seriesPopover = Jy(this.seriesPopover);
        };
        /**
         * @category Event handler
         */
        this.seriesPopoverChange = (open) => {
            this.arcgisChartsConfigBarChartSeriesPopoverChange.emit({
                open,
                element: this.seriesPopover,
            });
        };
        /**
         * @category Event handler
         */
        this.openSeriesPopOver = (event) => {
            // The target can be either the list item or the icon
            const target = event.target;
            const listItem = target.localName === "calcite-list-item"
                ? target
                : target.parentNode;
            const negateConfigSeriesOpen = !this.configSeriesOpen;
            const seriesIndex = this.model.getSeriesIndex(listItem.value);
            this.configSeriesOpen = this.selectedSeriesIndex === seriesIndex ? negateConfigSeriesOpen : true;
            if (this.configSeriesOpen) {
                this.selectedSeriesIndex = seriesIndex;
            }
            else {
                this.selectedSeriesIndex = undefined;
            }
        };
        /**
         * Creates a tooltip for the hovered action.
         * @category Event handler
         */
        this.createTooltip = (e) => {
            const iconElement = e.target;
            const text = this.model.getAggregationType() === RESTStatisticType.NoAggregation
                ? this.strings.noAggStackingDisabled
                : this.strings.singleStackingDisabled;
            this.tooltip = createActionTooltip(iconElement);
            this.tooltip.innerHTML = text;
            // append tooltip to the document body instead of action element
            // to avoid tooltip being cut off by calcite panel header element
            document.body.appendChild(this.tooltip);
        };
        /**
         * Destroys tooltip for the action that was being hovered.
         * @category Event handler
         */
        this.destroyTooltip = () => {
            destroyActionTooltip(this.tooltip);
        };
        /**
         * Event handler when model config changes.
         * @category Event handler
         */
        this.onModelConfigChange = () => {
            const prevUpdates = this.updatesOccurred;
            this.updatesOccurred = prevUpdates + 1;
        };
        /**
         * Emit event on open popover.
         * @param e
         * @category Private
         */
        this.openChange = (e) => {
            this.arcgisChartsConfigPopoverOpenChange.emit(e.detail);
        };
        this.model = undefined;
        this.configSeriesOpen = false;
        this.activeColorMatch = false;
        this.popoverPlacement = "leading";
        this.selectedSeriesIndex = undefined;
        this.updatesOccurred = 0;
        this.isResetAvailable = false;
    }
    /**
     * @category Watch handler
     */
    modelChange(newModel, oldModel) {
        // config change handler for chart model
        oldModel?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
        newModel?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    /**
     * @category Watch handler
     */
    selectedSeriesIndexChange() {
        if (this.selectedSeriesIndex !== undefined && this.configSeriesOpen) {
            this.setupSeriesPopover();
            if (this.seriesPopover !== null) {
                // sync back to seriesPopover
                this.seriesPopover.selectedName = this.model.getSeriesName(this.selectedSeriesIndex);
                this.seriesPopover.selectedColor = this.model.getSeriesColor(this.selectedSeriesIndex);
                if (!this.seriesPopover.open) {
                    this.seriesPopover.open = true;
                }
            }
            this.addSeriesPopover();
        }
        else {
            this.removeSeriesPopover();
        }
    }
    /**
     * @category Private
     */
    removeSeriesPopover() {
        if (this.seriesPopover !== null && this.seriesPopover !== undefined) {
            this.seriesPopover.open = false;
            this.seriesPopover = Jy(this.seriesPopover);
        }
        this.seriesPopoverChange(false);
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await Mn(this.hostElement, getAssetPath(`.`)));
        this.modelChange(this.model);
        this.onModelConfigChange();
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.removeSeriesPopover();
        this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    /**
     * @category Render UX
     */
    buildSeriesListItems() {
        const seriesLength = this.model.getSeriesLength();
        const seriesListItems = [];
        for (let index = 0; index < seriesLength; index += 1) {
            const id = this.model.getSeriesId(index);
            const color = this.model.getSeriesColor(index);
            const name = this.model.getSeriesName(index);
            const chartSubType = this.model.getChartSubType();
            const description = chartSubType === ml.BarAndLineFromFields ? this.strings.series : id;
            // If `undefined` is provided for the series, we default the color icon to a grey minus square
            let iconColor = "#CCCCCC";
            let icon = "minus-square";
            if ((seriesLength > 1 || !this.activeColorMatch) && color !== undefined) {
                icon = "square-f";
                const [r, g, b, a] = [...color];
                iconColor = `rgba(${r},${g},${b},${a / 255})`;
            }
            seriesListItems.push(h("calcite-list-item", { key: id, label: name ?? id ?? "", description: description, value: id, selected: this.selectedSeriesIndex === index, onClick: this.openSeriesPopOver }, h("calcite-icon", { icon: icon, style: { color: iconColor }, slot: "content-end" })));
        }
        return seriesListItems;
    }
    /**
     * @category Render UX
     */
    buildSeriesList() {
        return (h("calcite-list", { label: "", class: HTMLClasses.list, dragEnabled: true, selectionMode: "single", selectionAppearance: "border", onCalciteListOrderChange: this.onCalciteListOrderChange }, this.buildSeriesListItems()));
    }
    /**
     * @category Render UX
     */
    buildMultipleBarTypeList() {
        //  TODO: add the selected blue icon
        // icon = { this.stackedType === WebChartStackedKinds.Side ? ICON_TYPES.circle : null }
        // calcite-components build issue
        const stackingDisabled = this.shouldDisableStacking();
        if (stackingDisabled && this.model.getStackedType() !== WebChartStackedKinds.Side) {
            // no need to await here since `buildMultipleBarTypeList` is called by the render method, meaning it will be refreshed after the stacked type is set
            this.model.setStackedType(WebChartStackedKinds.Side);
        }
        return (h("calcite-list", { label: "", id: HTMLIDs.stackedType, class: HTMLClasses.list, selectionMode: "single-persist", selectionAppearance: "border", onClick: this.onMultipleBarSelect }, h("calcite-list-item", { label: this.strings.sideBySide, value: WebChartStackedKinds.Side, selected: this.model.getStackedType() === undefined || this.model.getStackedType() === WebChartStackedKinds.Side }, h("calcite-icon", { scale: "l", icon: "graph-bar-side-by-side", slot: "content-start" })), h("calcite-list-item", { class: stackingDisabled ? HTMLClasses.disabledStackingOption : "", label: this.strings.stacked, value: WebChartStackedKinds.Stacked, selected: this.model.getStackedType() === WebChartStackedKinds.Stacked }, h("calcite-icon", { scale: "l", icon: "graph-bar-stacked", slot: "content-start" }), stackingDisabled && (h("calcite-icon", { class: HTMLClasses.disabledInfo, slot: "content-end", icon: "information", scale: "s", flipRtl: true, onMouseOver: this.createTooltip, onMouseOut: this.destroyTooltip }))), h("calcite-list-item", { class: stackingDisabled ? HTMLClasses.disabledStackingOption : "", label: this.strings.stacked100, value: WebChartStackedKinds.Stacked100, selected: this.model.getStackedType() === WebChartStackedKinds.Stacked100 }, h("calcite-icon", { scale: "l", icon: "graph-bar-100-stacked", slot: "content-start" }), stackingDisabled && (h("calcite-icon", { class: HTMLClasses.disabledInfo, slot: "content-end", icon: "information", scale: "s", flipRtl: true, onMouseOver: this.createTooltip, onMouseOut: this.destroyTooltip })))));
    }
    async componentWillRender() {
        this.isResetAvailable = await this.model.resetAvailable();
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (h(Host, { key: 'c68c32d6706117802cdbf7064c6e7b98c6c8ee8e' }, h("div", { key: 'f0a0ea37114fe1e323aec655c569a21185028bbf', class: `${HTMLClasses.sectionLabel} ${HTMLClasses.topSection}` }, this.strings.displayMultiSeries), this.buildMultipleBarTypeList(), h("div", { key: 'f741d9f2a19e6bf86514f19eaddc02f863ff850d', class: HTMLClasses.sectionLabel }, this.strings.series, h("calcite-button", { key: 'f24295686f4299463b4607465e1403c1f996e5cd', onClick: this.resetSeriesOriginal, disabled: !this.isResetAvailable, appearance: "transparent", iconEnd: "reset" }, this.strings.reset)), this.buildSeriesList()));
    }
    /**
     * @category Private
     */
    setupSeriesPopover() {
        if (isNil(this.seriesPopover) && this.hostElement !== undefined) {
            // seriesPopover config
            const popoverConfig = {
                seriesNameChange: this.seriesNameChange,
                seriesColorChange: this.seriesColorChange,
                openChange: this.openChange,
                hostElement: this.hostElement,
                placement: this.popoverPlacement,
                popoverClose: this.seriesPopoverClose,
            };
            // create a new seriesPopover
            this.seriesPopover = Zy(Qy.BarChartSeriesPopover, popoverConfig);
        }
    }
    /**
     * @category Private
     */
    addSeriesPopover() {
        if (this.seriesPopover !== null && !document.body.contains(this.seriesPopover)) {
            document.body.appendChild(this.seriesPopover);
        }
        this.seriesPopoverChange(true);
    }
    /**
     * @category Private
     */
    shouldDisableStacking() {
        return this.model.getSeriesLength() <= 1 || this.model.getAggregationType() === RESTStatisticType.NoAggregation;
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return this; }
    static get watchers() { return {
        "model": ["modelChange"],
        "selectedSeriesIndex": ["selectedSeriesIndexChange"]
    }; }
    static get style() { return ArcgisChartsConfigBarChartSeriesStyle0; }
}, [1, "arcgis-charts-config-bar-chart-series", {
        "model": [16],
        "configSeriesOpen": [1028, "config-series-open"],
        "activeColorMatch": [4, "active-color-match"],
        "popoverPlacement": [1, "popover-placement"],
        "selectedSeriesIndex": [32],
        "updatesOccurred": [32],
        "isResetAvailable": [32]
    }, undefined, {
        "model": ["modelChange"],
        "selectedSeriesIndex": ["selectedSeriesIndexChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-config-bar-chart-series"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-config-bar-chart-series":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcGISChartsConfigBarChartSeries);
            }
            break;
    } });
}
defineCustomElement();

export { ArcGISChartsConfigBarChartSeries as A, defineCustomElement as d };
