// @arcgis/charts-components/dist/components/arcgis-charts-config-box-plot2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { s as bc, S as Sc, t as ty } from './index2.js';
import { C as ChartAuthoringPages } from './interfaces.js';
import { d as defineCustomElement$c } from './arcgis-charts-config-axes2.js';
import { d as defineCustomElement$b } from './arcgis-charts-config-axis2.js';
import { d as defineCustomElement$a } from './arcgis-charts-config-box-plot-data2.js';
import { d as defineCustomElement$9 } from './arcgis-chars-config-box-plot-series.js';
import { d as defineCustomElement$8 } from './arcgis-charts-config-category-format2.js';
import { d as defineCustomElement$7 } from './arcgis-charts-config-field-select2.js';
import { d as defineCustomElement$6 } from './arcgis-charts-config-format2.js';
import { d as defineCustomElement$5 } from './arcgis-charts-config-general2.js';
import { d as defineCustomElement$4 } from './arcgis-charts-config-guide2.js';
import { d as defineCustomElement$3 } from './arcgis-charts-config-guides2.js';
import { d as defineCustomElement$2 } from './arcgis-charts-config-invalid-input-message2.js';
import { d as defineCustomElement$1 } from './arcgis-charts-config-number-format2.js';
import { i as isEqual } from './isEqual.js';

const arcgisChartsConfigBoxPlotCss = ":host{overflow-y:auto;overflow-x:hidden;margin-bottom:var(--arcgis-charts-intra-section-margin);margin-top:var(--arcgis-charts-intra-section-margin);height:100%}";
const ArcgisChartsConfigBoxPlotStyle0 = arcgisChartsConfigBoxPlotCss;

const ArcGISChartsConfigBoxPlot = /*@__PURE__*/ proxyCustomElement(class ArcGISChartsConfigBoxPlot extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        /**
         * @category Private
         */
        this.popoverElement = null;
        /**
         * @category Event handler
         */
        this.popoverOpen = (e) => {
            const { open = false, element = null } = { ...e.detail };
            if (open === true && this.popoverElement !== element) {
                if (this.popoverElement !== null) {
                    this.popoverElement.open = false;
                }
                this.popoverElement = element;
            }
            else if (open === false && this.popoverElement === element) {
                this.popoverElement = null;
            }
        };
        /**
         * Event handler when model config changes.
         * @category Event handler
         */
        this.onModelConfigChange = async (e) => {
            const config = e?.detail.config ?? this.model.getConfig();
            if (this.layer === undefined) {
                await this.updateFetchLayerFieldsInfo(config);
            }
        };
        this.model = undefined;
        this.popoverPlacement = "leading";
        this.seriesColor = undefined;
        this.page = undefined;
        this.calculatedMinMaxBounds = undefined;
        this.layer = undefined;
        this.usePopupTemplateFieldsInfo = undefined;
        this.layerFieldsInfo = undefined;
        this.activeColorMatch = false;
    }
    /**
     * @category Watch handler
     */
    modelChange(newModel, oldModel) {
        // config change handler for chart model
        oldModel?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
        newModel.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    async layerChange(newValue, oldValue) {
        if (newValue !== undefined && !isEqual(oldValue, newValue)) {
            await this.updateFetchLayerFieldsInfo();
        }
    }
    async usePopupTemplateFieldsInfoChange(newValue, oldValue) {
        if (newValue !== undefined && newValue !== oldValue) {
            await this.updateFetchLayerFieldsInfo();
        }
    }
    /**
     * Assigns new colors to the series
     * @param seriesColor
     * @category Watch handler
     */
    seriesColorChange(seriesColor) {
        this.activeColorMatch = seriesColor?.get(0) === undefined;
        if (this.model.getColorMatch()) {
            seriesColor?.forEach((value, index) => {
                let color;
                if (value) {
                    if ("fillSymbol" in value) {
                        color = value.fillSymbol?.color;
                    }
                    else if ("lineSymbol" in value) {
                        color = value.lineSymbol.color;
                    }
                }
                if (seriesColor?.get(0) !== undefined &&
                    color !== undefined &&
                    !isEqual(this.model.getSeriesColor(index), color)) {
                    this.model.setSeriesColor(color, index);
                }
            });
        }
    }
    /**
     * update the data filter `where` clause with the layer definitionExpression if they are different
     */
    async updateDataFilters() {
        const savedDataFilter = this.model.getDataFilter();
        const whereFilter = savedDataFilter?.where;
        // NOTE: definitionalExpression is returned as null if it is not set
        const definitionExpressionFromLayer = this.layer?.definitionExpression ?? undefined;
        if (!isEqual(whereFilter, definitionExpressionFromLayer)) {
            await this.model.setDataFilter({
                ...savedDataFilter,
                where: definitionExpressionFromLayer ?? undefined,
            });
        }
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        if (this.layer !== undefined) {
            await this.layerChange(this.layer, undefined);
        }
        if (this.seriesColor !== undefined) {
            this.seriesColorChange(this.seriesColor);
        }
        if (this.usePopupTemplateFieldsInfo !== undefined) {
            await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, undefined);
        }
        this.modelChange(this.model);
        this.onModelConfigChange();
        this.updateDataFilters();
    }
    /**
     * @category Lifecycle
     */
    async componentWillRender() {
        // Ensure layerFieldsInfo is defined
        if (this.layerFieldsInfo === undefined) {
            await this.updateFetchLayerFieldsInfo();
        }
    }
    render() {
        let panel;
        switch (this.page) {
            case ChartAuthoringPages.Data:
                panel = (h(Host, { key: '090070d91c46b9690e2badef9cc9d3939dac7415' }, h("arcgis-charts-config-box-plot-data", { key: '0976cdba92052e3af8a9609463952a9544ec7b0e', model: this.model, layerFieldsInfo: this.layerFieldsInfo })));
                break;
            case ChartAuthoringPages.Series:
                panel = (h(Host, { key: '30f05ea39a30ada3ef720076d8833896c0099a38' }, h("arcgis-charts-config-box-plot-series", { key: '130e32d85130a94c213fc8e39a863e71a030799c', model: this.model, activeColorMatch: this.activeColorMatch, onArcgisChartsConfigBoxPlotSeriesPopoverChange: this.popoverOpen, popoverPlacement: this.popoverPlacement })));
                break;
            case ChartAuthoringPages.Axes:
                panel = (h(Host, { key: '84221ef0c0a3e0cb2f8f6fe983f96b061ce6ff8d' }, h("arcgis-charts-config-axes", { key: '1be58bd9e245ba2a3c087cfbc21f1efd78915276', model: this.model, showMinXBoundOption: false, showMaxXBoundOption: false, showMinYBoundOption: true, showMaxYBoundOption: true, showEnableXLogOption: false, showEnableYLogOption: false, calculatedMinMaxBounds: this.calculatedMinMaxBounds })));
                break;
            case ChartAuthoringPages.Guides:
                panel = (h(Host, { key: 'a752195850fdce70ee2be71d0ee5dc2b51bcb47a' }, h("arcgis-charts-config-guides", { key: '6acaf21fd03f09f17cb4722570b9f18b1cca0768', model: this.model, enableXAxisGuide: false, enableYAxisGuide: true, onArcgisChartsConfigGuidePopoverChange: this.popoverOpen })));
                break;
            case ChartAuthoringPages.Format:
                panel = (h(Host, { key: '16afe70fa2535a68e6f6e96b91b4e28e6661fc71' }, h("arcgis-charts-config-format", { key: '7ba59c9cdb58e981bc0c75a1a6b3cd39296f8ee9', model: this.model, onArcgisChartsConfigFormatPopoverChange: this.popoverOpen })));
                break;
            case ChartAuthoringPages.General:
                panel = (h(Host, { key: '943b6248b9e9845717de2a513aa233c5e27cdf1b' }, h("arcgis-charts-config-general", { key: 'e2948c75a57d1dc9915d3f51631cf5f59e414f2f', model: this.model })));
                break;
            default:
                panel = h(Host, { key: '5807d4f9f5ff47d827c2fa2edc1ab4403b9d7247' });
                break;
        }
        return panel;
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
    }
    /**
     * @category Private
     */
    async updateFetchLayerFieldsInfo(config) {
        // get feature layer from component prop if exists, otherwise get from current config iLayer
        const currentConfig = config ?? this.model.getConfig();
        let layer = this.layer;
        // if the layer doesn't exist we build one from the config data source
        if (!layer) {
            const iLayer = ty(currentConfig) ? currentConfig.iLayer : undefined;
            if (iLayer) {
                layer = await bc(iLayer);
            }
        }
        if (layer) {
            this.layerFieldsInfo = await Sc(layer, this.usePopupTemplateFieldsInfo);
        }
    }
    get hostElement() { return this; }
    static get watchers() { return {
        "model": ["modelChange"],
        "layer": ["layerChange"],
        "usePopupTemplateFieldsInfo": ["usePopupTemplateFieldsInfoChange"],
        "seriesColor": ["seriesColorChange"]
    }; }
    static get style() { return ArcgisChartsConfigBoxPlotStyle0; }
}, [1, "arcgis-charts-config-box-plot", {
        "model": [16],
        "popoverPlacement": [513, "popover-placement"],
        "seriesColor": [16],
        "page": [513],
        "calculatedMinMaxBounds": [16],
        "layer": [16],
        "usePopupTemplateFieldsInfo": [4, "use-popup-template-fields-info"],
        "layerFieldsInfo": [32],
        "activeColorMatch": [32]
    }, undefined, {
        "model": ["modelChange"],
        "layer": ["layerChange"],
        "usePopupTemplateFieldsInfo": ["usePopupTemplateFieldsInfoChange"],
        "seriesColor": ["seriesColorChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["arcgis-charts-config-box-plot", "arcgis-charts-config-axes", "arcgis-charts-config-axis", "arcgis-charts-config-box-plot-data", "arcgis-charts-config-box-plot-series", "arcgis-charts-config-category-format", "arcgis-charts-config-field-select", "arcgis-charts-config-format", "arcgis-charts-config-general", "arcgis-charts-config-guide", "arcgis-charts-config-guides", "arcgis-charts-config-invalid-input-message", "arcgis-charts-config-number-format"];
    components.forEach(tagName => { switch (tagName) {
        case "arcgis-charts-config-box-plot":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ArcGISChartsConfigBoxPlot);
            }
            break;
        case "arcgis-charts-config-axes":
            if (!customElements.get(tagName)) {
                defineCustomElement$c();
            }
            break;
        case "arcgis-charts-config-axis":
            if (!customElements.get(tagName)) {
                defineCustomElement$b();
            }
            break;
        case "arcgis-charts-config-box-plot-data":
            if (!customElements.get(tagName)) {
                defineCustomElement$a();
            }
            break;
        case "arcgis-charts-config-box-plot-series":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "arcgis-charts-config-category-format":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "arcgis-charts-config-field-select":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "arcgis-charts-config-format":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "arcgis-charts-config-general":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "arcgis-charts-config-guide":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "arcgis-charts-config-guides":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "arcgis-charts-config-invalid-input-message":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "arcgis-charts-config-number-format":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { ArcGISChartsConfigBoxPlot as A, defineCustomElement as d };
