// @arcgis/charts-components/dist/cjs/arcgis-charts-config-box-plot-series-popover-de6a6ade.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const chartUiUtils = require('./chart-ui-utils-ba499862.js');
const isEqual = require('./isEqual-7c4e90f5.js');

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    panelChildrenContainer: "panel-children-container",
};

const arcgisChartsConfigBoxPlotSeriesPopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.host{overflow-y:auto;width:-moz-min-content;width:min-content}.panel-children-container{display:flex;padding:0 1rem;margin-top:1rem}.panel-children-container calcite-label{flex:1 1 0px;display:flex;width:0px}";
const ArcgisChartsConfigBoxPlotSeriesPopoverStyle0 = arcgisChartsConfigBoxPlotSeriesPopoverCss;

const ArcgisChartsConfigBoxPlotSeriesPopover = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigSeriesPopoverNameChange = index.createEvent(this, "arcgisChartsConfigSeriesPopoverNameChange", 7);
        this.arcgisChartsConfigSeriesPopoverColorChange = index.createEvent(this, "arcgisChartsConfigSeriesPopoverColorChange", 7);
        this.arcgisChartsConfigSeriesPopoverLineStyleChange = index.createEvent(this, "arcgisChartsConfigSeriesPopoverLineStyleChange", 7);
        this.arcgisChartsConfigSeriesPopoverLineWidthChange = index.createEvent(this, "arcgisChartsConfigSeriesPopoverLineWidthChange", 7);
        this.arcgisChartsConfigPopoverClose = index.createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        this.arcgisChartsConfigPopoverOpenChange = index.createEvent(this, "arcgisChartsConfigPopoverOpenChange", 7);
        /**
         * @category Event handler
         */
        this.seriesNameChange = (event) => {
            const name = event.target.value;
            if (this.selectedName !== name) {
                this.selectedName = name;
                this.arcgisChartsConfigSeriesPopoverNameChange.emit({ name });
            }
        };
        /**
         * @category Event handler
         */
        this.closePopover = () => {
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit();
        };
        /**
         * @category Event handler
         */
        this.colorChange = (e) => {
            const color = e instanceof CustomEvent ? e.detail : e;
            const [r, g, b] = color;
            const a = restJsObjectLiterals.Hn * 255;
            this.selectedColor = [r, g, b, a];
            this.arcgisChartsConfigSeriesPopoverColorChange.emit({ color: [r, g, b, a] });
        };
        /**
         * @category Event handler
         */
        this.lineSymbolChange = (e) => {
            const { lineStyle } = e.target;
            const { color, style, width } = lineStyle;
            // Update the selected color if it has changed
            if (color !== undefined && !isEqual.isEqual(color, this.selectedColor)) {
                const [r, g, b] = color;
                const a = restJsObjectLiterals.Hn * 255;
                this.selectedColor = [r, g, b, a];
                this.arcgisChartsConfigSeriesPopoverColorChange.emit({ color: [r, g, b, a] });
            }
            // Update the selected line style if it has changed
            if (style !== undefined && style !== this.selectedLineStyle) {
                this.selectedLineStyle = style;
                this.arcgisChartsConfigSeriesPopoverLineStyleChange.emit({ style });
            }
            // Update the selected line width if it has changed
            if (width !== undefined && width !== this.selectedLineWidth) {
                this.selectedLineWidth = width;
                this.arcgisChartsConfigSeriesPopoverLineWidthChange.emit({ width });
            }
        };
        this.referenceElement = undefined;
        this.selectedName = undefined;
        this.selectedColor = undefined;
        this.selectedLineStyle = undefined;
        this.selectedLineWidth = undefined;
        this.isLineSymbol = false;
        this.placement = undefined;
        this.open = true;
        this.labelCanBeConfigured = true;
    }
    /**
     * Watch on open popover.
     * @category Private
     */
    openChange() {
        this.arcgisChartsConfigPopoverOpenChange.emit({ open: this.open });
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
    }
    /**
     * @category Lifecycle
     */
    componentDidLoad() {
        this.openChange();
    }
    /**
     * @category Lifecycle
     */
    componentDidUpdate() {
        setTimeout(() => {
            this.popoverElement?.reposition();
        }, chartUiUtils.UIDefaults.PopoverTimer);
    }
    componentDidRender() {
        restJsObjectLiterals.eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    render() {
        const rgba = this.selectedColor;
        return (index.h(index.Host, { key: '3e9bd455f4a358286ec10d025ab7f2999308f669', class: HTMLClasses.jsAppFlyout }, index.h("calcite-popover", { key: '9232c0f008a4b1752f7b6c41716148535d2b74b6', referenceElement: this.referenceElement, placement: this.placement, open: this.open, onCalcitePopoverClose: this.closePopover, ref: (e) => {
                this.popoverElement = e;
            }, label: "" }, index.h("calcite-panel", { key: '35402c619da7128053460bfd04ccff11f5f86aba', heading: this.strings.seriesElement, closable: true, closed: !this.open, onCalcitePanelClose: this.closePopover }, this.labelCanBeConfigured ? (index.h("div", { class: HTMLClasses.panelChildrenContainer }, index.h("calcite-label", null, this.strings.label, index.h("calcite-input", { type: "text", value: this.selectedName, onCalciteInputInput: this.seriesNameChange })))) : null, index.h("div", { key: 'c0db4454355f01ea1ee15fa42d864a15c4d64191', class: HTMLClasses.panelChildrenContainer }, this.isLineSymbol ? (index.h("arcgis-charts-config-line-style", { open: this.open, enableClose: true, usePanel: false, lineStyle: {
                type: restJsObjectLiterals.RESTSymbolType.SLS,
                style: this.selectedLineStyle,
                width: this.selectedLineWidth,
                color: rgba,
            }, onArcgisChartsConfigLineStyleClose: this.closePopover, onArcgisChartsConfigLineStyleChange: this.lineSymbolChange })) : (index.h("arcgis-charts-config-colors", { heading: this.strings.symbolColor, rgba: rgba, onColorSelect: this.colorChange })))))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "open": ["openChange"]
    }; }
};
ArcgisChartsConfigBoxPlotSeriesPopover.style = ArcgisChartsConfigBoxPlotSeriesPopoverStyle0;

exports.ArcgisChartsConfigBoxPlotSeriesPopover = ArcgisChartsConfigBoxPlotSeriesPopover;
