// @arcgis/charts-components/dist/cjs/arcgis-charts-config-box-plot-popover-3a32a096.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const index$1 = require('./index-e5a6a075.js');
const chartUiUtils = require('./chart-ui-utils-ba499862.js');
const interfaces = require('./interfaces-28e1772a.js');
const isEqual = require('./isEqual-7c4e90f5.js');

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    name: "bar-chart-data-popover",
    numericFieldPickList: "bar-chart-data-numeric-fields-pick-list",
    header: "header",
    fab: "fab",
    pickListItemLabel: "pick-list-item-label",
    sortMean: "sort-mean",
    sortMedian: "sort-median",
    sortOrderPickList: "sort-order-pick-list",
    panel: "panel",
};

const arcgisChartsConfigBoxPlotPopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.name{display:flex;flex-direction:column;margin:0.5rem;padding:0.5rem;background:white;min-width:250px}.header{margin:0;font-weight:bolder}.fab{display:flex;justify-content:stretch}.pick-list-item-label{font-size:small}.sort-icon-color{color:black}.sort-mean,.sort-median{transform:rotate(90deg)}.sort-order-pick-list{overflow-y:hidden}.panel{max-height:60vh}";
const ArcgisChartsConfigBoxPlotPopoverStyle0 = arcgisChartsConfigBoxPlotPopoverCss;

const ArcgisChartsConfigBoxPlotPopover = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigBoxPlotDataPopoverChange = index.createEvent(this, "arcgisChartsConfigBoxPlotDataPopoverChange", 7);
        this.arcgisChartsConfigPopoverClose = index.createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        /**
         * @category Private
         */
        this.placement = "leading";
        /**
         * @category Private
         */
        this.offsetDistance = -200;
        /**
         * @category Private
         */
        this.pickListHasChanged = false;
        /**
         * @category Event handler
         */
        this.closePopover = () => {
            if (this.contentKind === interfaces.BoxPlotDataPopoverKinds.numericFields && this.pickListHasChanged) {
                const numericFieldsChangeProps = {
                    pickListElement: this.picklist,
                    eventEmitter: this.arcgisChartsConfigBoxPlotDataPopoverChange,
                    contentKind: this.contentKind,
                };
                chartUiUtils.emitNumericFieldsChange(numericFieldsChangeProps);
            }
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit(this.pickListHasChanged);
        };
        /**
         * @category Event handler
         * @param e
         */
        this.onDataContentTypeChange = async (e) => {
            this.pickListHasChanged = true;
            const selectedItems = e.target.selectedItems;
            this.arcgisChartsConfigBoxPlotDataPopoverChange.emit({
                contentKind: this.contentKind,
                value: selectedItems.map((item) => item.value),
            });
            this.open = false;
        };
        /**
         * @category Event handler
         * @param e
         */
        this.onNumericFieldsChange = (e) => {
            if (e.target !== null) {
                this.selectedContent = e.target.selectedItems.map((item) => item.value);
            }
        };
        /**
         * @category Event handler
         */
        this.onNumericFieldSelectionDone = async () => {
            if (this.pickListHasChanged) {
                this.closePopover();
            }
        };
        /**
         * Creates a tooltip for the hovered action.
         * @category Event handler
         */
        this.createTooltip = (e) => {
            if (e.target?.label === this.strings.customSort && this.isCustomSortDisabled) {
                const itemElement = e.target;
                const { formatLocale } = restJsObjectLiterals.jn();
                const basicNumberFormatter = new Intl.NumberFormat(formatLocale);
                const formattedTotalLimit = basicNumberFormatter.format(chartUiUtils.UIDefaults.customSortLimit);
                const tooltipText = restJsObjectLiterals.On(this.strings.customSortLimit, {
                    totalLimit: formattedTotalLimit,
                });
                this.tooltip = chartUiUtils.createActionTooltip(itemElement);
                this.tooltip.innerHTML = tooltipText;
                // append tooltip to the document body instead of action element
                // to avoid tooltip being cut off by calcite panel header element
                document.body.appendChild(this.tooltip);
            }
        };
        /**
         * Destroys tooltip for the action that was being hovered.
         * @category Event handler
         */
        this.destroyTooltip = () => {
            chartUiUtils.destroyActionTooltip(this.tooltip);
        };
        this.headingTitle = undefined;
        this.referenceElement = undefined;
        this.open = undefined;
        this.contentKind = undefined;
        this.layerFieldsInfo = undefined;
        this.isCustomSortDisabled = false;
        this.selectedContent = undefined;
    }
    /**
     * Watches for change to selected content to record if there was a pick list change.
     * @param newSelectedContent
     * @param oldSelectedContent
     * @category Watch handler
     */
    selectedContentChange(newSelectedContent, oldSelectedContent) {
        this.pickListHasChanged = !isEqual.isEqual(newSelectedContent, oldSelectedContent);
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
    }
    /**
     * @category Lifestyle
     */
    componentDidRender() {
        setTimeout(async () => {
            if (typeof this.picklist?.setFocus === "function") {
                await this.picklist?.setFocus();
            }
        }, chartUiUtils.UIDefaults.PopoverTimer);
        restJsObjectLiterals.eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    componentDidUpdate() {
        setTimeout(() => {
            this.popoverElement?.reposition();
        }, chartUiUtils.UIDefaults.PopoverTimer);
    }
    /**
     * @param field
     * @category Render UX
     */
    buildPickListItem(field) {
        return (index.h("calcite-list-item", { key: field.name, label: restJsObjectLiterals.Tl(this.layerFieldsInfo, field.name), value: field.name, selected: chartUiUtils.itemMatchesSelectedContent(field.name, this.selectedContent) }));
    }
    /**
     * @param itemKind
     * @category Render UX
     */
    buildPickListItems(itemKind) {
        const fieldCount = this.layerFieldsInfo?.length ?? 0;
        const pickListItems = [];
        switch (itemKind) {
            case interfaces.BoxPlotDataPopoverKinds.numericFields: {
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Ic(field)) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
            case interfaces.BoxPlotDataPopoverKinds.category: {
                pickListItems.push(index.h("calcite-list-item", { label: "\u00A0", key: "\u00A0", value: "\u00A0", selected: chartUiUtils.itemMatchesSelectedContent("", this.selectedContent) }));
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Cc(field) || field.type === restJsObjectLiterals.RESTFieldType.String) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
            case interfaces.BoxPlotDataPopoverKinds.splitByField: {
                pickListItems.push(index.h("calcite-list-item", { label: "\u00A0", key: "\u00A0", value: "\u00A0", selected: chartUiUtils.itemMatchesSelectedContent("", this.selectedContent) }));
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Cc(field) || field.type === restJsObjectLiterals.RESTFieldType.String) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
        }
        return pickListItems;
    }
    /**
     * @category Render UX
     */
    renderCategory() {
        const items = this.buildPickListItems(interfaces.BoxPlotDataPopoverKinds.category);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.pickListItemLabel, selectionMode: "single", selectionAppearance: "border", filterEnabled: true, onCalciteListChange: this.onDataContentTypeChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     */
    renderNumericFields() {
        const items = this.buildPickListItems(interfaces.BoxPlotDataPopoverKinds.numericFields);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.numericFieldPickList, selectionMode: "multiple", filterEnabled: true, onCalciteListChange: this.onNumericFieldsChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     */
    renderSplitByField() {
        const items = this.buildPickListItems(interfaces.BoxPlotDataPopoverKinds.splitByField);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.pickListItemLabel, selectionMode: "single", selectionAppearance: "border", filterEnabled: true, onCalciteListChange: this.onDataContentTypeChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     * Unlike the other field-based render items, the sort orders
     * are static. So, a function separate from buildPickListItems()
     * needs to be used.
     */
    renderSortOrder() {
        const items = this.buildSortOrder();
        return (index.h("calcite-list", { label: "", class: HTMLClasses.sortOrderPickList, selectionMode: "single", selectionAppearance: "border", filterEnabled: false, onCalciteListChange: this.onDataContentTypeChange }, items));
    }
    /**
     * @category Render UX
     */
    buildSortOrder() {
        const items = Object.keys(index$1.Zf);
        let icon;
        return items.map((sortOrder) => {
            // y-axis icons need to be rotated 90 deg
            let rotateClass = "";
            switch (sortOrder) {
                case index$1.Zf.xAxisAsc: {
                    icon = "a-z-down";
                    break;
                }
                case index$1.Zf.xAxisDesc: {
                    icon = "a-z-up";
                    break;
                }
                case index$1.Zf.meanAsc:
                    {
                        // when rotated, will show as ascending
                        icon = "sort-descending";
                        rotateClass = HTMLClasses.sortMean;
                    }
                    break;
                case index$1.Zf.meanDesc: {
                    // when rotated, will show as descending
                    icon = "sort-ascending";
                    rotateClass = HTMLClasses.sortMean;
                    break;
                }
                case index$1.Zf.medianAsc: {
                    // when rotated, will show as ascending
                    icon = "sort-descending";
                    rotateClass = HTMLClasses.sortMedian;
                    break;
                }
                case index$1.Zf.medianDesc: {
                    // when rotated, will show as descending
                    icon = "sort-ascending";
                    rotateClass = HTMLClasses.sortMedian;
                    break;
                }
                case index$1.Zf.customSort: {
                    icon = "arrow-up-down";
                    break;
                }
            }
            return (index.h("calcite-list-item", { disabled: this.isCustomSortDisabled && sortOrder === index$1.Zf.customSort, key: sortOrder, label: this.strings[sortOrder], selected: this.selectedContent === sortOrder, value: index$1.Zf[sortOrder], onMouseOver: this.createTooltip, onMouseOut: this.destroyTooltip }, index.h("calcite-icon", { class: `sort-icon-color ${rotateClass}`, slot: "content-start", icon: icon })));
        });
    }
    /**
     * @category Render UX
     */
    renderPopoverInfo() {
        let content;
        switch (this.contentKind) {
            case interfaces.BoxPlotDataPopoverKinds.category:
                content = this.renderCategory();
                break;
            case interfaces.BoxPlotDataPopoverKinds.numericFields:
                content = this.renderNumericFields();
                break;
            case interfaces.BoxPlotDataPopoverKinds.splitByField:
                content = this.renderSplitByField();
                break;
            case interfaces.BoxPlotDataPopoverKinds.sortOrder:
                content = this.renderSortOrder();
                break;
        }
        return content;
    }
    /**
     * @category Lifecycle
     */
    render() {
        let footerButton;
        if (this.contentKind === interfaces.BoxPlotDataPopoverKinds.numericFields) {
            footerButton = (index.h("div", { key: '2ba8d441c029b5188c8e3da44dff415e5e6757e1', slot: "footer" }, index.h("calcite-fab", { key: '44fa1bbb636eb121709a7174cb224131ba6f07f8', class: HTMLClasses.fab, appearance: "outline-fill", kind: "neutral", icon: "", label: this.strings.selectionDone, scale: "s", "text-enabled": true, text: this.strings.selectionDone, onClick: this.onNumericFieldSelectionDone })));
        }
        return (index.h(index.Host, { key: 'd017ed79b794d79baabaa4066f7df1f38c01ec32', class: HTMLClasses.jsAppFlyout }, index.h("calcite-popover", { key: '04332f7cca5187dc61d4c095944eee56b7e9c0d8', class: HTMLClasses.name, referenceElement: this.referenceElement, placement: this.placement, offsetDistance: this.offsetDistance, open: this.open, onCalcitePopoverClose: this.closePopover, label: "", ref: (e) => {
                this.popoverElement = e;
            } }, index.h("calcite-panel", { key: 'b8f212fc5ea8a98aa0e1074574dd587b9f897755', class: HTMLClasses.panel, closable: true, closed: !this.open, onCalcitePanelClose: this.closePopover }, index.h("div", { key: '13d238aaf6bfa9dcbbbc302ddeece3d0905a4a57', slot: "header-content", class: HTMLClasses.header }, this.headingTitle ?? ""), this.renderPopoverInfo(), footerButton))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "selectedContent": ["selectedContentChange"]
    }; }
};
ArcgisChartsConfigBoxPlotPopover.style = ArcgisChartsConfigBoxPlotPopoverStyle0;

exports.ArcgisChartsConfigBoxPlotPopover = ArcgisChartsConfigBoxPlotPopover;
