// @arcgis/charts-components/dist/cjs/arcgis-charts-config-pie-chart-slices-popover-8c98dbfb.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const chartUiUtils = require('./chart-ui-utils-ba499862.js');

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    panelChildrenContainer: "panel-children-container",
};

const arcgisChartsConfigPieChartSlicesPopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.host{overflow-y:auto;width:-moz-min-content;width:min-content}.panel-children-container{display:flex;padding:0 1rem;margin-top:1rem}.panel-children-container calcite-label{flex:1 1 0px;display:flex;width:0px}";
const ArcgisChartsConfigPieChartSlicesPopoverStyle0 = arcgisChartsConfigPieChartSlicesPopoverCss;

const ArcgisChartsConfigPieChartSlicesPopover = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigSlicesPopoverNameChange = index.createEvent(this, "arcgisChartsConfigSlicesPopoverNameChange", 7);
        this.arcgisChartsConfigSlicesPopoverColorChange = index.createEvent(this, "arcgisChartsConfigSlicesPopoverColorChange", 7);
        this.arcgisChartsConfigPopoverClose = index.createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        this.arcgisChartsConfigPopoverOpenChange = index.createEvent(this, "arcgisChartsConfigPopoverOpenChange", 7);
        /**
         * @category Event handler
         */
        this.sliceNameChange = (event) => {
            const name = event.target.value;
            if (this.selectedName !== name) {
                this.selectedName = name;
                this.arcgisChartsConfigSlicesPopoverNameChange.emit({ name });
            }
        };
        /**
         * @category Event handler
         */
        this.closePopover = () => {
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit();
        };
        /**
         * @category Event handler
         */
        this.colorChange = (e) => {
            const color = e instanceof CustomEvent ? e.detail : e;
            const [r, g, b] = color;
            const a = restJsObjectLiterals.Hn * 255;
            this.selectedColor = [r, g, b, a];
            this.arcgisChartsConfigSlicesPopoverColorChange.emit({ color: [r, g, b, a] });
        };
        this.referenceElement = undefined;
        this.selectedName = undefined;
        this.selectedColor = undefined;
        this.placement = undefined;
        this.open = true;
    }
    /**
     * Watch on open popover.
     * @category Private
     */
    openChange() {
        this.arcgisChartsConfigPopoverOpenChange.emit({ open: this.open });
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
    }
    /**
     * @category Lifecycle
     */
    componentDidLoad() {
        this.openChange();
    }
    /**
     * @category Lifecycle
     */
    componentDidUpdate() {
        setTimeout(() => {
            this.popoverElement?.reposition();
        }, chartUiUtils.UIDefaults.PopoverTimer);
    }
    /**
     * @category Lifecycle
     */
    componentDidRender() {
        restJsObjectLiterals.eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    render() {
        const rgba = this.selectedColor !== undefined ? [...this.selectedColor] : undefined;
        return (index.h(index.Host, { key: 'c6b9a5f8263656735e233a7b25e48b93aadf0d64', class: HTMLClasses.jsAppFlyout }, index.h("calcite-popover", { key: '4ead8687b436fdaecab6bb434166a5b3ea4dce4c', referenceElement: this.referenceElement, placement: this.placement, onCalcitePopoverClose: this.closePopover, open: this.open, ref: (e) => {
                this.popoverElement = e;
            }, label: "" }, index.h("calcite-panel", { key: 'f6645beae8cd410c9ddd51a7278629647768c75f', heading: this.strings.sliceStyle, closable: true, closed: !this.open, onCalcitePanelClose: this.closePopover }, index.h("div", { key: '487270788405ba3ae79edce7e8e3b39ffa3050c7', class: HTMLClasses.panelChildrenContainer }, index.h("calcite-label", { key: 'd16da21861a570db59a26a4b8fbbd376d77d301f' }, this.strings.label, index.h("calcite-input", { key: 'f0feb0374ad42789888bbb7f39140587d8c5e775', type: "text", value: this.selectedName, onCalciteInputInput: this.sliceNameChange }))), index.h("div", { key: '74fdd08b09fe38592c322e3f52055c8c1c09dd3e', class: HTMLClasses.panelChildrenContainer }, index.h("arcgis-charts-config-colors", { key: '8714c3ee828def05976f32367db593c2e1e1beff', heading: this.strings.symbolColor, rgba: rgba, onColorSelect: this.colorChange }))))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "open": ["openChange"]
    }; }
};
ArcgisChartsConfigPieChartSlicesPopover.style = ArcgisChartsConfigPieChartSlicesPopoverStyle0;

exports.ArcgisChartsConfigPieChartSlicesPopover = ArcgisChartsConfigPieChartSlicesPopover;
