// @arcgis/charts-components/dist/cjs/arcgis-charts-config-bar-chart_7.cjs.entry.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const arcgisChartsConfigLineChart = require('./arcgis-charts-config-line-chart-0e02b94b.js');
const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
require('./interfaces-28e1772a.js');
require('./isEqual-7c4e90f5.js');
require('./_Uint8Array-9bd4a4da.js');
require('./_Map-d0ba2c83.js');
require('./isObject-b063d8a9.js');
require('./_getAllKeys-facd90a6.js');
require('./_getTag-66a55d52.js');
require('./cloneDeep-3503ada6.js');
require('./_initCloneObject-76ab9e25.js');

const HTMLClasses = {
    addChartContainer: "add-chart-container",
    viewChartComponent: "view-chart-component",
    header: "header",
    dropdownPopup: "dropdown-popup",
};

const arcgisChartsConfigViewChartsCss = ".view-chart-component{width:var(--arcgis-charts-config-panel-width);height:var(--arcgis-charts-config-panel-height)}.add-chart-container{display:flex;justify-content:center;padding:2rem 0;background-color:var(--calcite-color-background)}.header{margin:0}.dropdown-popup{position:relative}";
const ArcgisChartsConfigViewChartsStyle0 = arcgisChartsConfigViewChartsCss;

const ArcGISChartsConfigViewCharts = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigViewChartSelect = index.createEvent(this, "arcgisChartsConfigViewChartSelect", 7);
        this.arcgisChartsConfigViewChartDelete = index.createEvent(this, "arcgisChartsConfigViewChartDelete", 7);
        this.arcgisChartsConfigViewChartEdit = index.createEvent(this, "arcgisChartsConfigViewChartEdit", 7);
        this.arcgisChartsConfigViewChartAddChart = index.createEvent(this, "arcgisChartsConfigViewChartAddChart", 7);
        /**
         * rename popover
         * @category Private
         */
        this.chartRenamePopover = null;
        /**
         * @category Event handler
         */
        this.onAddChartClick = (e) => {
            this.arcgisChartsConfigViewChartAddChart.emit({
                referenceElement: e.target,
                layerId: this.layer?.id.toString(),
            });
        };
        /**
         * @category Event handler
         */
        this.onChartSelect = (event) => {
            const eventCurrentTarget = event.currentTarget;
            let selectedChartId;
            if (eventCurrentTarget?.localName === "calcite-list-item") {
                selectedChartId = eventCurrentTarget.value;
            }
            else if (eventCurrentTarget?.localName === "calcite-icon") {
                const parentNode = eventCurrentTarget.parentNode;
                selectedChartId = parentNode.value;
            }
            this.arcgisChartsConfigViewChartSelect.emit({ chartId: selectedChartId });
        };
        /**
         * @category Event handler
         */
        this.chartEditSelect = () => {
            if (!this.disableEditing) {
                this.arcgisChartsConfigViewChartEdit.emit({
                    chartId: this.selectedActionChartId,
                });
                this.chartActionClosed(this.selectedActionChartId);
            }
        };
        /**
         * @category Event handler
         */
        this.chartRenameSelect = (e) => {
            if (!this.disableEditing) {
                const dropdownItemRename = e.target;
                const chartListItem = dropdownItemRename.closest("calcite-list-item");
                const selectedModel = this.modelList?.find((model) => model.getChartId() === this.selectedActionChartId);
                if (chartListItem !== null) {
                    this.renameReferenceElement = chartListItem;
                }
                // Chart has to be selected before it can be renamed
                this.arcgisChartsConfigViewChartSelect.emit({
                    chartId: this.selectedActionChartId,
                });
                this.chartTitle = selectedModel?.getTitleText() ?? "";
                this.renameOpen = true;
                // popover config
                this.setupRenamePopover();
            }
        };
        /**
         * @category Event handler
         */
        this.chartDeleteSelect = () => {
            if (!this.disableEditing) {
                this.arcgisChartsConfigViewChartDelete.emit({
                    chartId: this.selectedActionChartId,
                });
                this.selectedActionChartId = "";
            }
        };
        /**
         * @category Event handler
         */
        this.chartActionClosed = (chartId) => {
            if (chartId === this.selectedActionChartId && !this.renameOpen) {
                this.selectedActionChartId = "";
            }
        };
        /**
         * @category Event handler
         */
        this.chartActionOpened = (chartId) => {
            this.selectedActionChartId = chartId;
        };
        /**
         * @category Event handler
         */
        this.popoverClose = () => {
            const selectedModel = this.modelList?.find((model) => model.getChartId() === this.selectedChartId);
            this.chartTitle = selectedModel?.getTitleText() ?? "";
            this.selectedActionChartId = "";
            this.renameOpen = false;
            this.removeChartRenamePopover();
        };
        /**
         * @category Event handler
         */
        this.onChartRename = (e) => {
            if (!this.disableEditing) {
                this.chartTitle = e.detail.name;
                const model = this.modelList?.find((model) => model.getChartId() === this.selectedActionChartId);
                model?.setTitleText(this.chartTitle);
                this.selectedActionChartId = "";
                this.renameOpen = false;
                this.removeChartRenamePopover();
            }
        };
        /**
         * Event handler when model config changes.
         * @category Event handler
         */
        this.onModelConfigChange = () => {
            const prevUpdates = this.updatesOccurred;
            this.updatesOccurred = prevUpdates + 1;
        };
        this.selectedChartId = undefined;
        this.layer = undefined;
        this.modelList = undefined;
        this.noGroup = false;
        this.disableEditing = false;
        this.chartTitle = undefined;
        this.renameOpen = undefined;
        this.selectedActionChartId = undefined;
        this.updatesOccurred = 0;
    }
    /**
     * Called every time modelList is updated
     * @category Watch handler
     */
    async modelListChange(newModelList, oldModelList) {
        // remove event listeners from old models
        oldModelList.forEach((model) => {
            model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
        });
        // add event listeners to new models
        newModelList.forEach((model) => {
            model.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
        });
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.removeChartRenamePopover();
    }
    /**
     * @category Lifecycle
     */
    render() {
        return index.h(index.Host, { key: 'ffcb1017f90767e00a29624a19765210cd13b802', class: HTMLClasses.viewChartComponent }, this.buildChartList());
    }
    /**
     * @category Render UX
     */
    buildChartList() {
        const chartListItems = this.modelList?.map((model, index$1) => {
            const id = model.getChartId();
            const title = model.getTitleText();
            const icon = restJsObjectLiterals.tu(model.getConfig());
            return (index.h("calcite-list-item", { label: title ?? id ?? restJsObjectLiterals.On(this.strings.chartTitle, { index: index$1.toString() }), selected: this.selectedChartId === (id?.toString() ?? index$1.toString()), value: id ?? index$1.toString(), onClick: this.onChartSelect }, index.h("calcite-icon", { slot: "content-start", icon: icon }), !this.disableEditing ? (index.h("calcite-dropdown", { slot: "actions-end", overlayPositioning: "fixed", placement: "bottom-end", onCalciteDropdownClose: () => this.chartActionClosed(id ?? ""), onCalciteDropdownOpen: () => this.chartActionOpened(id ?? ""), class: HTMLClasses.dropdownPopup }, index.h("calcite-action", { scale: "l", text: this.strings.chartActions, color: "dark", slot: "trigger", active: id === this.selectedActionChartId }, index.h("calcite-icon", { icon: "ellipsis" })), index.h("calcite-dropdown-group", { selectionMode: "none" }, index.h("calcite-dropdown-item", { accessKey: "edit", onCalciteDropdownItemSelect: this.chartEditSelect }, this.strings.edit), index.h("calcite-dropdown-item", { accessKey: "rename", onCalciteDropdownItemSelect: this.chartRenameSelect }, this.strings.rename), index.h("calcite-dropdown-item", { accessKey: "delete", onCalciteDropdownItemSelect: this.chartDeleteSelect }, this.strings.delete)))) : null));
        }) ?? null;
        return this.layer ? (index.h("calcite-list", { label: "", selectionMode: "single-persist", selectionAppearance: "border" }, this.noGroup ? (chartListItems) : (index.h("calcite-list-item-group", { key: this.layer.id, heading: this.layer.title ?? this.layer.id.toString() }, chartListItems)), !this.disableEditing ? (index.h("div", { class: HTMLClasses.addChartContainer }, index.h("calcite-fab", { slot: "fab", appearance: "outline-fill", kind: "neutral", hidden: false, textEnabled: true, text: this.strings.addChart, label: this.strings.addChart, onClick: this.onAddChartClick, scale: "s" }))) : null)) : null;
    }
    // setup rename popover
    setupRenamePopover() {
        if (this.chartRenamePopover === null) {
            // chartRenamePopover config
            const popoverConfig = {
                name: this.chartTitle,
                popoverReferenceElement: this.renameReferenceElement,
                onChartRename: this.onChartRename,
                popoverClose: this.popoverClose,
            };
            // create a new chartRenamePopover
            this.chartRenamePopover = restJsObjectLiterals.Zy(restJsObjectLiterals.Qy.ChartRenamePopover, popoverConfig);
            document.body.appendChild(this.chartRenamePopover);
        }
        if (this.chartRenamePopover !== null) {
            this.chartRenamePopover.open = this.renameOpen;
            this.chartRenamePopover.name = this.chartTitle;
            this.chartRenamePopover.popoverReferenceElement = this.renameReferenceElement;
        }
    }
    /**
     * Remove chart rename popover
     * @category Private
     */
    removeChartRenamePopover() {
        if (this.chartRenamePopover !== null) {
            this.chartRenamePopover.open = false;
            this.chartRenamePopover = restJsObjectLiterals.Jy(this.chartRenamePopover);
        }
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "modelList": ["modelListChange"]
    }; }
};
ArcGISChartsConfigViewCharts.style = ArcgisChartsConfigViewChartsStyle0;

exports.arcgis_charts_config_bar_chart = arcgisChartsConfigLineChart.ArcGISChartsConfigBarChart;
exports.arcgis_charts_config_box_plot = arcgisChartsConfigLineChart.ArcGISChartsConfigBoxPlot;
exports.arcgis_charts_config_histogram = arcgisChartsConfigLineChart.ArcGISChartsConfigHistogram;
exports.arcgis_charts_config_line_chart = arcgisChartsConfigLineChart.ArcGISChartsConfigLineChart;
exports.arcgis_charts_config_pie_chart = arcgisChartsConfigLineChart.ArcGISChartsConfigPieChart;
exports.arcgis_charts_config_scatter_plot = arcgisChartsConfigLineChart.ArcgisChartsConfigScatterPlot;
exports.arcgis_charts_config_view_charts = ArcGISChartsConfigViewCharts;
