// @arcgis/charts-components/dist/cjs/arcgis-charts-config-chart-rename-popover-1581be26.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const chartUiUtils = require('./chart-ui-utils-ba499862.js');
const isEmpty = require('./isEmpty-b0fd430d.js');

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    chartRenameComponent: "chart-rename-component",
    renameLabelContainer: "rename-label-container",
    renamePanel: "rename-panel",
};

const arcgisChartsConfigChartRenamePopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.chart-rename-component{width:var(--arcgis-charts-config-panel-width-medium);height:var(--arcgis-charts-config-panel-height)}.host{overflow-y:auto;width:-moz-min-content;width:min-content}.rename-label-container{display:flex;padding:0 1rem;margin-top:1rem}.rename-label-container calcite-label{flex:1 1 0px;display:flex;width:0px}.rename-panel{width:calc(var(--calcite-panel-width-multiplier) * 20vw);max-width:calc(var(--calcite-panel-width-multiplier) * 420px);min-width:calc(var(--calcite-panel-width-multiplier) * 240px)}";
const ArcgisChartsConfigChartRenamePopoverStyle0 = arcgisChartsConfigChartRenamePopoverCss;

const ArcgisChartsConfigChartRenamePopover = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigChartRenamePopoverChange = index.createEvent(this, "arcgisChartsConfigChartRenamePopoverChange", 7);
        this.arcgisChartsConfigPopoverClose = index.createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        /**
         * @category Event handler
         */
        this.onNameChange = (e) => {
            const chartTitle = e.target.value;
            this.changedName = chartTitle;
            this.isChartTitleInvalid = isEmpty.isEmpty(chartTitle.trim());
            chartUiUtils.modifyCalciteInputStatus({
                inputElement: this.characterLimitInputElement,
                status: this.isChartTitleInvalid ? chartUiUtils.UIInputStatus.INVALID : chartUiUtils.UIInputStatus.IDLE,
            });
        };
        /**
         * @category Event handler
         */
        this.onCancelClick = () => {
            this.changedName = this.name;
            this.arcgisChartsConfigPopoverClose.emit();
            this.open = false;
        };
        /**
         * @category Event handler
         */
        this.onSubmitClick = () => {
            if (!this.isChartTitleInvalid) {
                this.name = this.changedName;
                this.arcgisChartsConfigChartRenamePopoverChange.emit({ name: this.changedName });
                this.open = false;
            }
        };
        /**
         * @category Event handler
         */
        this.popoverClose = () => {
            this.changedName = this.name;
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit();
        };
        this.name = undefined;
        this.open = undefined;
        this.popoverReferenceElement = undefined;
        this.isChartTitleInvalid = false;
        this.changedName = undefined;
    }
    /**
     * @category Watch handler
     */
    changeName(newValue) {
        this.changedName = newValue;
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
        // set changedName default value to name
        this.changedName = this.name;
    }
    componentDidRender() {
        this.popoverElement?.reposition();
        restJsObjectLiterals.eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (index.h(index.Host, { key: 'a1631e64413969232b3ba9cddd11ee92053f35ea', class: `${HTMLClasses.chartRenameComponent} ${HTMLClasses.jsAppFlyout}` }, index.h("calcite-popover", { key: 'ee214851896443abcd5637db875c8c09c77b4e52', referenceElement: this.popoverReferenceElement, ref: (e) => {
                this.popoverElement = e;
            }, onCalcitePopoverClose: this.popoverClose, open: this.open, placement: "bottom", label: "" }, index.h("calcite-panel", { key: '6f6ca5f439e7a60ba8ad76b609a3d61bb0ff3916', class: HTMLClasses.renamePanel, closable: true, closed: !this.open, onCalcitePanelClose: this.popoverClose, heading: this.strings.renameChart }, index.h("div", { key: '05db16ee88391d6414ae20dfaffdf43588e86db5', class: HTMLClasses.renameLabelContainer }, index.h("calcite-label", { key: '2bf02bdfd4800a7d3bf61d820d185000495069e2' }, this.strings.chartName, index.h("calcite-input", { key: '50ed46890fda0be6bff788ab4e129d5470539fd5', required: true, type: "text", value: this.changedName, onCalciteInputInput: this.onNameChange, ref: (e) => {
                this.characterLimitInputElement = e;
            } }), index.h("arcgis-charts-config-invalid-input-message", { key: 'a8d6c829d0f27af1d75edecc3a569cb9b3ef5d3a', invalidInputMessage: this.strings.emptyChartTitle, isInvalid: this.isChartTitleInvalid }))), index.h("calcite-button", { key: 'e0410269488356ce566648501dd22a86096348f7', slot: "footer", appearance: "solid", scale: "m", width: "half", onClick: this.onSubmitClick }, this.strings.submit), index.h("calcite-button", { key: '9e2d33804654253574b7bad25cf1c8674176f2fc', slot: "footer", appearance: "outline-fill", scale: "m", width: "half", onClick: this.onCancelClick }, this.strings.cancel)))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "name": ["changeName"]
    }; }
};
ArcgisChartsConfigChartRenamePopover.style = ArcgisChartsConfigChartRenamePopoverStyle0;

exports.ArcgisChartsConfigChartRenamePopover = ArcgisChartsConfigChartRenamePopover;
