// @arcgis/charts-components/dist/cjs/arcgis-charts-config-bar-chart-popover-57a22d97.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const index$1 = require('./index-e5a6a075.js');
const interfaces = require('./interfaces-28e1772a.js');
const chartUiUtils = require('./chart-ui-utils-ba499862.js');
const isEqual = require('./isEqual-7c4e90f5.js');

const HTMLClasses = {
    jsAppFlyout: "js-app-flyout",
    name: "bar-chart-data-popover",
    numericFieldPickList: "bar-chart-data-numeric-fields-pick-list",
    header: "header",
    fab: "fab",
    pickListItemLabel: "pick-list-item-label",
    rotateClass: "sort-y-axis",
    sortOrderPickList: "sort-order-pick-list",
    panel: "panel",
};

const arcgisChartsConfigBarChartPopoverCss = ".hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0px;left:0px}.am5-modal-curtain{top:0px;left:0px;width:100%;height:100%;position:absolute;background:rgba(255, 255, 255, 0.5);z-index:100}.am5-modal-wrapper{top:0px;left:0px;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:rgba(255, 255, 255, 0.5);z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:left;white-space:normal;background:rgb(255, 255, 255);border-radius:4px;box-shadow:rgba(0, 0, 0, 0.45) 0px 0px 36px 0px;color:rgb(0, 0, 0)}.arcgis-charts-modal{box-shadow:none !important}.arcgis-charts-modal-header{background-color:rgba(0, 0, 0, 0.05);font-weight:bold;padding:4px 4px 4px 4px}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.name{display:flex;flex-direction:column;margin:0.5rem;padding:0.5rem;background:white;min-width:250px}.header{margin:0;font-weight:bolder}.fab{display:flex;justify-content:stretch}.pick-list-item-label{font-size:small}.sort-icon-color{color:black}.sort-y-axis{transform:rotate(90deg)}.sort-order-pick-list{overflow-y:hidden}.panel{max-height:60vh}";
const ArcgisChartsConfigBarChartPopoverStyle0 = arcgisChartsConfigBarChartPopoverCss;

const ArcgisChartsConfigBarChartPopover = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisChartsConfigBarChartDataPopoverChange = index.createEvent(this, "arcgisChartsConfigBarChartDataPopoverChange", 7);
        this.arcgisChartsConfigPopoverClose = index.createEvent(this, "arcgisChartsConfigPopoverClose", 7);
        /**
         * @category Private
         */
        this.placement = "leading";
        /**
         * @category Private
         */
        this.offsetDistance = -200;
        /**
         * @category Private
         */
        this.pickListHasChanged = false;
        /**
         * Creates a tooltip for the hovered action.
         * @category Event handler
         */
        this.createTooltip = (e) => {
            if (e.target?.label === this.strings.customSort && this.isCustomSortDisabled) {
                const itemElement = e.target;
                const { formatLocale } = restJsObjectLiterals.jn();
                const basicNumberFormatter = new Intl.NumberFormat(formatLocale);
                const formattedTotalLimit = basicNumberFormatter.format(chartUiUtils.UIDefaults.customSortLimit);
                const tooltipText = restJsObjectLiterals.On(this.strings.customSortLimit, {
                    totalLimit: formattedTotalLimit,
                });
                this.tooltip = chartUiUtils.createActionTooltip(itemElement);
                this.tooltip.innerHTML = tooltipText;
                // append tooltip to the document body instead of action element
                // to avoid tooltip being cut off by calcite panel header element
                document.body.appendChild(this.tooltip);
            }
        };
        /**
         * Destroys tooltip for the action that was being hovered.
         * @category Event handler
         */
        this.destroyTooltip = () => {
            chartUiUtils.destroyActionTooltip(this.tooltip);
        };
        /**
         * @category Event handler
         */
        this.closePopover = () => {
            if (this.contentKind === interfaces.BarChartDataPopoverKinds.numericFields && this.pickListHasChanged) {
                const numericFieldsChangeProps = {
                    pickListElement: this.picklist,
                    eventEmitter: this.arcgisChartsConfigBarChartDataPopoverChange,
                    contentKind: this.contentKind,
                };
                chartUiUtils.emitNumericFieldsChange(numericFieldsChangeProps);
            }
            this.open = false;
            this.arcgisChartsConfigPopoverClose.emit(this.pickListHasChanged);
        };
        /**
         * @category Event handler
         * @param e
         */
        this.onDataContentTypeChange = (e) => {
            this.pickListHasChanged = true;
            const selectedItems = e.target.selectedItems;
            this.arcgisChartsConfigBarChartDataPopoverChange.emit({
                contentKind: this.contentKind,
                value: selectedItems.map((item) => item.value),
            });
            this.open = false;
        };
        /**
         * @category Event handler
         * @param e
         */
        this.onNumericFieldsChange = (e) => {
            if (e.target !== null) {
                this.selectedContent = e.target.selectedItems.map((item) => item.value);
            }
        };
        /**
         * @category Event handler
         */
        this.onNumericFieldSelectionDone = () => {
            if (this.pickListHasChanged) {
                this.closePopover();
            }
        };
        this.headingTitle = undefined;
        this.referenceElement = undefined;
        this.open = undefined;
        this.contentKind = undefined;
        this.layerFieldsInfo = undefined;
        this.selectedContent = undefined;
        this.isCustomSortDisabled = false;
        this.isNoAggregationDisabled = false;
    }
    /**
     * Watches for change to selected content to record if there was a pick list change.
     * @param newSelectedContent
     * @param oldSelectedContent
     * @category Watch handler
     */
    selectedContentChange(newSelectedContent, oldSelectedContent) {
        this.pickListHasChanged = !isEqual.isEqual(newSelectedContent, oldSelectedContent);
    }
    /**
     * Calls `reposition()` method on popover element.
     * @category Public
     */
    async reposition() {
        await this.popoverElement?.reposition();
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.hostElement, index.getAssetPath(`.`)));
    }
    /**
     * @category Lifecycle
     */
    async componentWillRender() {
        const listPromises = chartUiUtils.aggregationList.map(async (item) => await restJsObjectLiterals.ta(item));
        this.aggregationLabels = await Promise.all(listPromises);
    }
    /**
     * @category Lifestyle
     */
    componentDidRender() {
        setTimeout(async () => {
            await this.picklist?.setFocus();
        }, chartUiUtils.UIDefaults.PopoverTimer);
        restJsObjectLiterals.eu(this.popoverElement, this.open);
    }
    /**
     * @category Lifecycle
     */
    componentDidUpdate() {
        setTimeout(() => {
            this.popoverElement?.reposition();
        }, chartUiUtils.UIDefaults.PopoverTimer);
    }
    /**
     * @category Render UX
     */
    renderAggregationType() {
        const items = this.buildAggregationListItems();
        return (index.h("calcite-list", { label: "", class: HTMLClasses.pickListItemLabel, filterEnabled: false, selectionMode: "single", selectionAppearance: "border", onCalciteListChange: this.onDataContentTypeChange }, items));
    }
    /**
     * @category Render UX
     */
    buildAggregationListItems() {
        const filteredAggregationTypes = this.isNoAggregationDisabled
            ? chartUiUtils.aggregationList.filter((aggregationType) => aggregationType !== restJsObjectLiterals.RESTStatisticType.NoAggregation)
            : chartUiUtils.aggregationList;
        return filteredAggregationTypes.map((item, index$1) => (index.h("calcite-list-item", { label: this.aggregationLabels[index$1], key: item, value: item, selected: chartUiUtils.itemMatchesSelectedContent(item, this.selectedContent) })));
    }
    /**
     * @category Render UX
     */
    renderCategory() {
        const items = this.buildPickListItems(interfaces.BarChartDataPopoverKinds.category);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.pickListItemLabel, selectionMode: "single", selectionAppearance: "border", filterEnabled: true, onCalciteListChange: this.onDataContentTypeChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     */
    renderNumericFields() {
        const items = this.buildPickListItems(interfaces.BarChartDataPopoverKinds.numericFields);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.numericFieldPickList, selectionMode: "multiple", filterEnabled: true, onCalciteListChange: this.onNumericFieldsChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     * Unlike the other field-based render items, the sort orders
     * are static. So, a function separate from buildPickListItems()
     * needs to be used.
     */
    renderSortOrder() {
        const items = this.buildSortOrder();
        return (index.h("calcite-list", { label: "", class: HTMLClasses.sortOrderPickList, selectionMode: "single", selectionAppearance: "border", filterEnabled: false, onCalciteListChange: this.onDataContentTypeChange }, items));
    }
    /**
     * @category Render UX
     */
    renderSplitByField() {
        const items = this.buildPickListItems(interfaces.BarChartDataPopoverKinds.splitByField);
        return (index.h("calcite-list", { label: "", class: HTMLClasses.pickListItemLabel, selectionMode: "single", selectionAppearance: "border", filterEnabled: true, onCalciteListChange: this.onDataContentTypeChange, ref: (e) => {
                this.picklist = e;
            } }, items));
    }
    /**
     * @category Render UX
     */
    buildSortOrder() {
        const items = Object.keys(index$1.Jf);
        let icon;
        return items.map((sortOrder) => {
            // y-axis icons need to be rotated 90 deg
            let rotateClass = "";
            switch (sortOrder) {
                case index$1.Jf.xAxisAsc: {
                    icon = "a-z-down";
                    break;
                }
                case index$1.Jf.xAxisDesc: {
                    icon = "a-z-up";
                    break;
                }
                case index$1.Jf.yAxisAsc: {
                    // when rotated, will show as ascending
                    icon = "sort-descending";
                    rotateClass = HTMLClasses.rotateClass;
                    break;
                }
                case index$1.Jf.yAxisDesc: {
                    // when rotated, will show as descending
                    icon = "sort-ascending";
                    rotateClass = HTMLClasses.rotateClass;
                    break;
                }
                case index$1.Jf.customSort: {
                    icon = "arrow-up-down";
                    break;
                }
            }
            return (index.h("calcite-list-item", { disabled: this.isCustomSortDisabled && sortOrder === index$1.Jf.customSort, key: sortOrder, label: this.strings[sortOrder], selected: this.selectedContent === sortOrder, value: index$1.Jf[sortOrder], onMouseOver: this.createTooltip, onMouseOut: this.destroyTooltip }, index.h("calcite-icon", { class: `sort-icon-color ${rotateClass}`, slot: "content-start", icon: icon })));
        });
    }
    /**
     * @param field
     * @category Render UX
     */
    buildPickListItem(field) {
        return (index.h("calcite-list-item", { key: field.name, label: restJsObjectLiterals.Tl(this.layerFieldsInfo, field.name), value: field.name, selected: chartUiUtils.itemMatchesSelectedContent(field.name, this.selectedContent) }));
    }
    /**
     * @param itemKind
     * @category Render UX
     */
    buildPickListItems(itemKind) {
        const fieldCount = this.layerFieldsInfo?.length ?? 0;
        const pickListItems = [];
        switch (itemKind) {
            case interfaces.BarChartDataPopoverKinds.numericFields: {
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Ic(field)) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
            case interfaces.BarChartDataPopoverKinds.category: {
                pickListItems.push(index.h("calcite-list-item", { label: "\u00A0", key: "\u00A0", value: "\u00A0", selected: chartUiUtils.itemMatchesSelectedContent("", this.selectedContent) }));
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Cc(field) || field.type === restJsObjectLiterals.RESTFieldType.String || field.type === restJsObjectLiterals.RESTFieldType.Date) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
            case interfaces.BarChartDataPopoverKinds.splitByField: {
                pickListItems.push(index.h("calcite-list-item", { label: "\u00A0", key: "\u00A0", value: "\u00A0", selected: chartUiUtils.itemMatchesSelectedContent("", this.selectedContent) }));
                for (let i = 0; i < fieldCount; i += 1) {
                    const field = this.layerFieldsInfo[i];
                    if (restJsObjectLiterals.Cc(field) || field.type === restJsObjectLiterals.RESTFieldType.String) {
                        pickListItems.push(this.buildPickListItem(field));
                    }
                }
                break;
            }
        }
        return pickListItems;
    }
    /**
     * @category Render UX
     */
    renderPopoverInfo() {
        let content;
        switch (this.contentKind) {
            case interfaces.BarChartDataPopoverKinds.aggregation:
                content = this.renderAggregationType();
                break;
            case interfaces.BarChartDataPopoverKinds.category:
                content = this.renderCategory();
                break;
            case interfaces.BarChartDataPopoverKinds.numericFields:
                content = this.renderNumericFields();
                break;
            case interfaces.BarChartDataPopoverKinds.splitByField:
                content = this.renderSplitByField();
                break;
            case interfaces.BarChartDataPopoverKinds.sortOrder:
                content = this.renderSortOrder();
                break;
        }
        return content;
    }
    /**
     * @category Lifecycle
     */
    render() {
        let footerButton;
        if (this.contentKind === interfaces.BarChartDataPopoverKinds.numericFields) {
            footerButton = (index.h("div", { key: '69b4e11d8b5b58db2fd66b490af834be3bde07f2', slot: "footer" }, index.h("calcite-fab", { key: '25c16bead2fcc1938ca60dd1f1f99cb11766b8e5', class: HTMLClasses.fab, appearance: "outline-fill", kind: "neutral", icon: "", label: this.strings.selectionDone, scale: "s", "text-enabled": true, text: this.strings.selectionDone, onClick: this.onNumericFieldSelectionDone })));
        }
        return (index.h(index.Host, { key: 'e11752f4a12ed99d65009ec64402333acb533ec3', class: HTMLClasses.jsAppFlyout }, index.h("calcite-popover", { key: '51dbaeb420b4ed674ad919029eb3cec6c405cc38', class: HTMLClasses.name, referenceElement: this.referenceElement, placement: this.placement, offsetDistance: this.offsetDistance, open: this.open, onCalcitePopoverClose: this.closePopover, label: "", ref: (e) => {
                this.popoverElement = e;
            } }, index.h("calcite-panel", { key: 'ff4f392843ccf06cd864d96c95e5edcf786bfc34', class: HTMLClasses.panel, closable: true, closed: !this.open, onCalcitePanelClose: this.closePopover }, index.h("div", { key: 'd9715b90dafd2a45ad6f2764a4648253be50d0d7', slot: "header-content", class: HTMLClasses.header }, this.headingTitle ?? ""), this.renderPopoverInfo(), footerButton))));
    }
    static get assetsDirs() { return ["assets"]; }
    get hostElement() { return index.getElement(this); }
    static get watchers() { return {
        "selectedContent": ["selectedContentChange"]
    }; }
};
ArcgisChartsConfigBarChartPopover.style = ArcgisChartsConfigBarChartPopoverStyle0;

exports.ArcgisChartsConfigBarChartPopover = ArcgisChartsConfigBarChartPopover;
