// @arcgis/charts-components/dist/cjs/charts-action-bar-1d680ddd.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const index = require('./index-2e64cfc1.js');
const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const interfaces = require('./interfaces-28e1772a.js');
const actionBarUtils = require('./action-bar-utils-893c6007.js');
const t9nStrings = require('./t9n-strings-ceef62aa.js');

const HTMLClasses = {
    actionBar: "action-bar",
};

function isChartElement(element) {
    return element?.tagName.toLowerCase() === "arcgis-chart";
}

const chartsActionBarCss = ".action-bar{display:flex;flex-direction:column;height:100%;border-right:1px solid var(--calcite-color-border-2);overflow-x:hidden}";
const ArcgisChartsActionBarStyle0 = chartsActionBarCss;

const ArcGISChartsActionBar = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.arcgisDefaultActionSelect = index.createEvent(this, "arcgisDefaultActionSelect", 7);
        this.arcgisActionBarToggle = index.createEvent(this, "arcgisActionBarToggle", 7);
        /**
         * Element reference for expand action in calciteActionBar
         * @type {HTMLCalciteActionElement | null}
         * @category Private
         * @default null
         */
        this.expandAction = null;
        /**
         * Handler for default action click event
         * @category Event handler
         * @param {MouseEvent} e Event payload from click
         */
        this.onDefaultActionClick = (e) => {
            const action = e.target;
            const payload = {
                actionId: action.id,
                actionActive: false,
            };
            switch (action.id) {
                case restJsObjectLiterals.Vy.EditChart:
                    // Edit chart toggle active state
                    this.editChartToggle =
                        this.editChartToggle === restJsObjectLiterals.Uy.Active ? restJsObjectLiterals.Uy.Inactive : restJsObjectLiterals.Uy.Active;
                    payload.actionActive = this.editChartToggle === restJsObjectLiterals.Uy.Active;
                    break;
                case restJsObjectLiterals.Vy.Legend:
                    // Legend toggle active state
                    this.legendToggle = this.legendToggle === restJsObjectLiterals.Uy.Active ? restJsObjectLiterals.Uy.Inactive : restJsObjectLiterals.Uy.Active;
                    payload.actionActive = this.legendToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.legendVisibility = payload.actionActive;
                    }
                    break;
                case restJsObjectLiterals.Vy.RotateChart:
                    // Rotate Chart toggle active state
                    this.rotateChartToggle =
                        this.rotateChartToggle === restJsObjectLiterals.Uy.Active ? restJsObjectLiterals.Uy.Inactive : restJsObjectLiterals.Uy.Active;
                    payload.actionActive = this.rotateChartToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.rotation = payload.actionActive;
                    }
                    break;
                case restJsObjectLiterals.Vy.FilterBySelection:
                    // Filter by Selection toggle active state
                    this.filterBySelectionToggle =
                        this.filterBySelectionToggle === restJsObjectLiterals.Uy.Active ? restJsObjectLiterals.Uy.Inactive : restJsObjectLiterals.Uy.Active;
                    payload.actionActive = this.filterBySelectionToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.filterBySelection = payload.actionActive;
                    }
                    break;
                case restJsObjectLiterals.Vy.FilterByExtent:
                    // Filter by Extent toggle active state
                    this.filterByExtentToggle =
                        this.filterByExtentToggle === restJsObjectLiterals.Uy.Active ? restJsObjectLiterals.Uy.Inactive : restJsObjectLiterals.Uy.Active;
                    payload.actionActive = this.filterByExtentToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.filterByExtent = payload.actionActive;
                    }
                    break;
                case restJsObjectLiterals.Vy.Selection:
                    // Activate Selection action and deactivate Zoom action
                    this.selectionToggle = restJsObjectLiterals.Uy.Active;
                    this.zoomToggle = restJsObjectLiterals.Uy.Inactive;
                    payload.actionActive = this.selectionToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.actionMode = this.actionBarSelectionMode;
                    }
                    break;
                case restJsObjectLiterals.Vy.Zoom:
                    // Activate Zoom action and deactivate Selection action
                    this.zoomToggle = restJsObjectLiterals.Uy.Active;
                    this.selectionToggle = restJsObjectLiterals.Uy.Inactive;
                    payload.actionActive = this.zoomToggle === restJsObjectLiterals.Uy.Active;
                    if (this.chartElement !== undefined) {
                        this.chartElement.actionMode = restJsObjectLiterals.zy.Zoom;
                    }
                    break;
                case restJsObjectLiterals.Vy.ClearSelection:
                    payload.actionActive = this.clearSelectionToggle === restJsObjectLiterals.Uy.Active;
                    this.chartElement?.clearSelection();
                    break;
                case restJsObjectLiterals.Vy.FullExtent:
                    payload.actionActive = this.fullExtentToggle === restJsObjectLiterals.Uy.Active;
                    this.chartElement?.resetZoom();
                    break;
                case restJsObjectLiterals.Vy.SwitchSelection:
                    payload.actionActive = this.switchSelectionToggle === restJsObjectLiterals.Uy.Active;
                    this.chartElement?.switchSelection();
                    break;
            }
            this.arcgisDefaultActionSelect.emit(payload);
        };
        /**
         * Handler for calciteActionBarToggle event
         * @category Event handler
         */
        this.handleCalciteActionBarToggle = () => {
            this.expanded = this.calciteActionBar.expanded;
            this.setCustomActionsExpanded();
            this.arcgisActionBarToggle.emit(this.expanded);
        };
        /**
         * Handler for mouseOver event that creates a tooltip for the hovered action
         * @category Event handler
         * @param {MouseEvent} e Event payload from mouse over
         */
        this.handleMouseOver = (e) => {
            // Only show tooltip if action bar in not expanded
            if (!this.calciteActionBar.expanded) {
                const element = e.target;
                if (this.actionTooltip !== undefined) {
                    this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                }
                // Find action from target element or its child
                let action;
                if (element?.tagName === "CALCITE-ACTION") {
                    action = element;
                }
                else if (element?.tagName === "DIV") {
                    action = element.firstChild;
                }
                if (action !== undefined) {
                    this.actionTooltip = actionBarUtils.createTooltip(action.text, action);
                    this.actionTooltip.open = true;
                }
            }
            else {
                this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for mouseOut event that destroys current action tooltip
         * @category Event handler
         */
        this.handleMouseOut = () => {
            if (this.actionTooltip !== null) {
                this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
            }
        };
        /**
         * Handler for `slotchange` event to set custom action event handlers dynamically
         * @category Event handler
         * @param {Event} e Event payload from slot change
         */
        this.handleSlotChange = (e) => {
            const slot = e.target;
            const customActions = slot.assignedElements();
            for (const action of customActions) {
                action.onmouseover = this.handleMouseOver;
            }
        };
        this.chartType = undefined;
        this.chartElement = undefined;
        this.forceDisableActions = false;
        this.editChartToggle = restJsObjectLiterals.Uy.Inactive;
        this.editChartState = restJsObjectLiterals.Gy.Disabled;
        this.editChartTitle = undefined;
        this.legendToggle = restJsObjectLiterals.Uy.Inactive;
        this.legendState = restJsObjectLiterals.Gy.Enabled;
        this.legendTitle = undefined;
        this.rotateChartToggle = restJsObjectLiterals.Uy.Inactive;
        this.rotateChartState = restJsObjectLiterals.Gy.Enabled;
        this.rotateChartTitle = undefined;
        this.filterBySelectionToggle = restJsObjectLiterals.Uy.Inactive;
        this.filterBySelectionState = restJsObjectLiterals.Gy.Enabled;
        this.filterBySelectionTitle = undefined;
        this.filterByExtentToggle = restJsObjectLiterals.Uy.Inactive;
        this.filterByExtentState = restJsObjectLiterals.Gy.Enabled;
        this.filterByExtentTitle = undefined;
        this.actionBarSelectionMode = restJsObjectLiterals.zy.MultiSelectionWithCtrlKey;
        this.selectionToggle = restJsObjectLiterals.Uy.Inactive;
        this.selectionState = restJsObjectLiterals.Gy.Enabled;
        this.selectionTitle = undefined;
        this.zoomToggle = restJsObjectLiterals.Uy.Inactive;
        this.zoomState = restJsObjectLiterals.Gy.Enabled;
        this.zoomTitle = undefined;
        this.fullExtentToggle = restJsObjectLiterals.Uy.Inactive;
        this.fullExtentState = restJsObjectLiterals.Gy.Enabled;
        this.fullExtentTitle = undefined;
        this.clearSelectionToggle = restJsObjectLiterals.Uy.Inactive;
        this.clearSelectionState = restJsObjectLiterals.Gy.Enabled;
        this.clearSelectionTitle = undefined;
        this.switchSelectionToggle = restJsObjectLiterals.Uy.Inactive;
        this.switchSelectionState = restJsObjectLiterals.Gy.Enabled;
        this.switchSelectionTitle = undefined;
        this.expanded = false;
        this.hiddenActions = [restJsObjectLiterals.Vy.EditChart];
        this.actionTooltip = null;
    }
    /**
     * Renders default actions
     * @category Private
     */
    // eslint-disable-next-line complexity
    renderDefaultActions() {
        const shownChartActions = actionBarUtils.getDefaultChartActions(this.chartType).filter((action) => !this.hiddenActions?.includes(action));
        const editChart = shownChartActions.includes(restJsObjectLiterals.Vy.EditChart);
        const legend = shownChartActions.includes(restJsObjectLiterals.Vy.Legend);
        const rotateChart = shownChartActions.includes(restJsObjectLiterals.Vy.RotateChart);
        const filterBySelection = shownChartActions.includes(restJsObjectLiterals.Vy.FilterBySelection);
        const filterByExtent = shownChartActions.includes(restJsObjectLiterals.Vy.FilterByExtent) && !this.chartElement?.layer?.isTable;
        const zoom = shownChartActions.includes(restJsObjectLiterals.Vy.Zoom);
        const selection = shownChartActions.includes(restJsObjectLiterals.Vy.Selection);
        const clearSelection = shownChartActions.includes(restJsObjectLiterals.Vy.ClearSelection);
        const fullExtent = shownChartActions.includes(restJsObjectLiterals.Vy.FullExtent);
        const switchSelection = shownChartActions.includes(restJsObjectLiterals.Vy.SwitchSelection);
        return [
            legend || rotateChart ? (index.h("calcite-action-group", null, editChart ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.EditChart, icon: "gear", text: this.commonStrings?.actions.editChart, active: this.editChartToggle === restJsObjectLiterals.Uy.Active, disabled: this.editChartState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, legend ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.Legend, text: this.legendTitle ?? this.strings?.legend, active: this.legendToggle === restJsObjectLiterals.Uy.Active, disabled: this.legendState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }, index.h("calcite-icon", { icon: "legend", scale: "s", flipRtl: true })))) : null, rotateChart ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.RotateChart, icon: "rotate", text: this.rotateChartTitle ?? this.strings?.rotateChart, active: this.rotateChartToggle === restJsObjectLiterals.Uy.Active, disabled: this.rotateChartState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null)) : null,
            filterBySelection || filterByExtent ? (index.h("calcite-action-group", null, filterBySelection ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.FilterBySelection, icon: "selection-filter", text: this.filterBySelectionTitle ?? this.strings?.filterBySelection, active: this.filterBySelectionToggle === restJsObjectLiterals.Uy.Active, disabled: this.filterBySelectionState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, filterByExtent ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.FilterByExtent, icon: "extent-filter", text: this.filterByExtentTitle ?? this.strings?.filterByExtent, active: this.filterByExtentToggle === restJsObjectLiterals.Uy.Active, disabled: this.filterByExtentState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null)) : null,
            selection || zoom || fullExtent || clearSelection || switchSelection ? (index.h("calcite-action-group", null, zoom ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.Zoom, icon: "magnifying-glass", text: this.zoomTitle ?? this.strings?.zoom, active: this.zoomToggle === restJsObjectLiterals.Uy.Active, disabled: this.zoomState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, selection ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.Selection, icon: "cursor-selection", text: this.selectionTitle ?? this.strings?.selection, active: this.selectionToggle === restJsObjectLiterals.Uy.Active, disabled: this.selectionState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, clearSelection ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.ClearSelection, icon: "erase", text: this.clearSelectionTitle ?? this.strings?.clearSelection, active: this.clearSelectionToggle === restJsObjectLiterals.Uy.Active, disabled: this.clearSelectionState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, fullExtent ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.FullExtent, icon: "zoom-out-fixed", text: this.fullExtentTitle ?? this.strings?.fullExtent, active: this.fullExtentToggle === restJsObjectLiterals.Uy.Active, disabled: this.fullExtentState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null, switchSelection ? (index.h("div", { onMouseOver: this.handleMouseOver }, index.h("calcite-action", { id: restJsObjectLiterals.Vy.SwitchSelection, icon: "arrow-right-left", text: this.switchSelectionTitle ?? this.strings?.switchSelection, active: this.switchSelectionToggle === restJsObjectLiterals.Uy.Active, disabled: this.switchSelectionState === restJsObjectLiterals.Gy.Disabled || this.forceDisableActions, onClick: this.onDefaultActionClick }))) : null)) : null,
        ];
    }
    /**
     * Sets whether text is expanded on all custom actions
     * @category Private
     */
    setCustomActionsExpanded() {
        this.el.querySelectorAll("calcite-action").forEach((action) => {
            const calciteAction = action;
            calciteAction.textEnabled = this.expanded;
        });
    }
    /**
     * @category Lifecycle
     */
    render() {
        return (index.h(index.Host, { key: '7866dc987a9a9c0418078e3768b7dcf73ec3ad30' }, index.h("calcite-action-bar", { key: '178e184337330b8b55f4bd2c9db4e44c2f451f4b', class: HTMLClasses.actionBar, onCalciteActionBarToggle: this.handleCalciteActionBarToggle, onMouseOut: this.handleMouseOut, expanded: this.expanded, ref: (ref) => {
                this.calciteActionBar = ref;
            }, messageOverrides: { expand: this.strings.expand, collapse: this.strings.collapse }, "overflow-actions-disabled": true }, index.h("slot", { key: 'a4bcab94d4c7065083a9007749d163eb93920bac', name: interfaces.Slot.customActionTop, onSlotchange: this.handleSlotChange }), this.renderDefaultActions(), index.h("slot", { key: '18170b43997487ca9e6c295b2556be1e442afd4f', name: interfaces.Slot.customActionBelow, onSlotchange: this.handleSlotChange }))));
    }
    /**
     * @category Lifecycle
     */
    async componentWillLoad() {
        this.commonStrings = await t9nStrings.fetchCommonStrings(this.el);
        ({ strings: this.strings } = await restJsObjectLiterals.Mn(this.el, index.getAssetPath(`.`), true));
        // Set chart element reference to its associated chart component.
        this.chartElement = isChartElement(this.el.parentElement) ? this.el.parentElement : undefined;
    }
    /**
     * @category Lifecycle
     */
    componentDidRender() {
        this.handleExpandTooltip();
    }
    /**
     * @category Lifecycle
     */
    disconnectedCallback() {
        this.actionTooltip?.remove();
    }
    /**
     * handle creation and deletion of the expand tooltip
     * @category Private
     */
    handleExpandTooltip() {
        if (this.expandAction === null) {
            this.expandAction = this.calciteActionBar.shadowRoot?.querySelector("calcite-action") ?? null;
            if (this.expandAction !== null) {
                this.expandAction.addEventListener("mouseenter", () => {
                    // add expand tooltip only iff
                    // a. calcite action bar is collapsed (not expanded)
                    // b. document body have calcite tooltip as its last child
                    //
                    if (!this.calciteActionBar?.expanded &&
                        this.expandAction !== null &&
                        !document.body.contains(this.actionTooltip)) {
                        this.actionTooltip = actionBarUtils.createTooltip(this.strings.expand, this.expandAction);
                        this.actionTooltip.open = true;
                    }
                });
                this.expandAction.addEventListener("mouseleave", () => {
                    if (!this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                    }
                });
                this.expandAction.addEventListener("click", () => {
                    // destroy tooltip when the user clicks on expand action
                    if (this.calciteActionBar?.expanded && this.actionTooltip !== null) {
                        this.actionTooltip = actionBarUtils.destroyTooltip(this.actionTooltip);
                    }
                });
            }
        }
    }
    static get assetsDirs() { return ["assets"]; }
    get el() { return index.getElement(this); }
};
ArcGISChartsActionBar.style = ArcgisChartsActionBarStyle0;

exports.ArcGISChartsActionBar = ArcGISChartsActionBar;
