// @arcgis/charts-components/dist/cjs/messages.da-90739473-d35fadd5.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Søjlediagram",lineChart:"Linjediagram",pieChart:"Cirkeldiagram",histogram:"Histogram",scatterPlot:"Punktdiagram",boxPlot:"Kassediagram"},t={mean:"Middel",median:"Median",sum:"Sum",discretePercentile:"Særskilt percentil",minimum:"Minimum",maximum:"Maksimum",variance:"Varians",count:"Tælling",aggregation:"${ statistics } for ${ fieldName }",noAggregation:"Ingen aggregering"},i="Diagram",e="Tællerække",r="X-akse",g="Ikke tilgængelig",n={start:"Fastgør til det første datapunkt",end:"Fastgør til det sidste datapunkt"};var s={chartType:a,statistics:t,defaultTitle:i,countSeries:e,xAxis:r,notAvailable:g,timeAggregationTypes:n};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = s;
exports.defaultTitle = i;
exports.notAvailable = g;
exports.statistics = t;
exports.timeAggregationTypes = n;
exports.xAxis = r;
