// @arcgis/charts-components/dist/cjs/messages.pt-BR-d65a99a9-7c1fef6a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Gráfico",barChart:"Gráfico de barra",lineChart:"Gráfico de linhas",pieChart:"Gráfico de pizza",histogram:"Histograma",scatterPlot:"Gráfico de dispersão",boxPlot:"Diagrama de caixa"},i={mean:"Médio",median:"Mediana",sum:"Soma",discretePercentile:"Percentil discreto",minimum:"Mínimo",maximum:"Máximo",variance:"Variância",count:"Contagem",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sem agregação"},e="Gráfico",o="Séries de contagem",t="Eixo X",r="Não Disponível",s={start:"Ajustar ao primeiro ponto de dados",end:"Ajustar ao útimo ponto de dados"};var n={chartType:a,statistics:i,defaultTitle:e,countSeries:o,xAxis:t,notAvailable:r,timeAggregationTypes:s};

exports.chartType = a;
exports.countSeries = o;
exports['default'] = n;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = i;
exports.timeAggregationTypes = s;
exports.xAxis = t;
