// @arcgis/charts-components/dist/cjs/messages.lv-57b9e23e-d1641f6a-2aae7a8a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagramma",barChart:"Joslu diagramma",lineChart:"Līniju diagramma",pieChart:"Sektoru diagramma",histogram:"Histogramma",scatterPlot:"Izkliedes diagramma",boxPlot:"Kastveida diagramma"},i={mean:"Vidējais",median:"Mediāna",sum:"Summa",discretePercentile:"Diskrētā procentīle",minimum:"Minimums",maximum:"Maksimums",variance:"Novirze",count:"Skaits",aggregation:"${ statistics } no ${ fieldName }",noAggregation:"Bez apkopošanas"},t="Diagramma",e="Sēriju uzskaite",s="X ass",m="Nav pieejams",r={start:"Pielipināt pie pirmā datu punkta",end:"Pielipināt pie pēdējā datu punkta"};var n={chartType:a,statistics:i,defaultTitle:t,countSeries:e,xAxis:s,notAvailable:m,timeAggregationTypes:r};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = n;
exports.defaultTitle = t;
exports.notAvailable = m;
exports.statistics = i;
exports.timeAggregationTypes = r;
exports.xAxis = s;
