// @arcgis/charts-components/dist/cjs/messages.sv-8a49b669-85a809cc.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Ett fel uppstod när diagrammet lästes in.",t="Det går inte att tillämpa loggomvandling på negativa värden eller nollvärden.",a="Det går inte att tillämpa kvadratrotomvandling på negativa värden.",r="Ett fel uppstod när lagret lästes in. URL = ${ url }. Portalobjekt-ID = ${ portalItemId }.",n="Fälten [${ fieldNames }] som är definierade i konfigurationen hittades inte i lagret.",i="${ dataPath } måste vara unik. Serien som heter ${ seriesName } har ett ID (${ seriesID }) som redan används av en annan serie.",s="${ dataPath } kan inte utföra icke-tidsmässig aggregering på ett icke-numeriskt fält.",l="${ dataPath } saknar egenskapen ${ missingProperty }.",d="${ dataPath } får inte vara kortare än ${ limit } tecken.",o="${ dataPath } får inte ha färre än ${ limit } objekt.",m="${ dataPath } får inte ha fler än ${ limit } objekt.",g="${ dataPath } måste ha minst ett tecken som inte är ett blanksteg.",p="${ dataPath } får inte innehålla ${ additionalProperty }.",$="${ dataPath } måste vara lika med ett av dessa tillåtna värden: ${ allowedValues }.",h="${ dataPath } måste matcha schemat för någon av dessa: ${ schemaOptions }.",u="Punktdiagram med proportionerliga symboler stöds inte. Standardstorleken för symboler har tillämpats.",v="Det gick inte att läsa indata.",f="Förväntad serietyp vid index ${ seriesIndex } är '${ expectedType }' men '${ receivedType }' togs emot istället",k="eller";var P={defaultError:e,negativeValueInDataForLogTransformation:t,negativeValueInDataForSqrtTransformation:a,layerLoadFailure:r,layerFieldsNotFound:n,duplicateSeriesID:i,nonNumericAggregation:s,requiredProperty:l,minLength:d,minItems:o,maxItems:m,whiteSpacePattern:g,additionalProperty:p,enumValues:$,anyOfValues:h,bubbleChartValidateMsg:u,queryError:v,invalidSeriesType:f,or:k};

exports.additionalProperty = p;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = u;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = i;
exports.enumValues = $;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = r;
exports.maxItems = m;
exports.minItems = o;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = t;
exports.negativeValueInDataForSqrtTransformation = a;
exports.nonNumericAggregation = s;
exports.or = k;
exports.queryError = v;
exports.requiredProperty = l;
exports.whiteSpacePattern = g;
