// @arcgis/charts-components/dist/cjs/messages.ca-21b03272-f358f0a6-f9533fff.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Gràfic",barChart:"Gràfic de barres",lineChart:"Gràfic de línies",pieChart:"Gràfic circular",histogram:"Histograma",scatterPlot:"Gràfic de dispersió",boxPlot:"Diagrama de caixa"},e={mean:"Mitjana",median:"Mediana",sum:"Suma",discretePercentile:"Percentil discret",minimum:"Mínim",maximum:"Màxim",variance:"Variància",count:"Recompte",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sense agregació"},i="Gràfic",t="Recompte de sèries",s="Eix X",r="No disponible",c={start:"Ajusta al primer punt de dades",end:"Ajusta a l'últim punt de dades"};var n={chartType:a,statistics:e,defaultTitle:i,countSeries:t,xAxis:s,notAvailable:r,timeAggregationTypes:c};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = r;
exports.statistics = e;
exports.timeAggregationTypes = c;
exports.xAxis = s;
