// @arcgis/charts-components/dist/cjs/messages.pl-1ed04c02-b3939b63-5934fff5.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Diagram słupkowy grupowy",lineChart:"Diagram liniowy",pieChart:"Diagram kołowy",histogram:"Histogram",scatterPlot:"Diagram punktowy",boxPlot:"Diagram pudełkowy"},i={mean:"Średnia",median:"Mediana",sum:"Suma",discretePercentile:"Percentyl dyskretny",minimum:"Minimum",maximum:"Maksimum",variance:"Odchylenie",count:"Liczba",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Bez agregacji"},e="Wykres",t="Zlicz serie",s="Oś X",r="Niedostępne",n={start:"Dociągnij do pierwszego punktu danych",end:"Dociągnij do ostatniego punktu danych"};var o={chartType:a,statistics:i,defaultTitle:e,countSeries:t,xAxis:s,notAvailable:r,timeAggregationTypes:n};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = i;
exports.timeAggregationTypes = n;
exports.xAxis = s;
