// @arcgis/charts-components/dist/cjs/messages.el-5582e711-ef34cef9-db584fac.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"Γράφημα",barChart:"Γράφημα ράβδων",lineChart:"Γράφημα γραμμών",pieChart:"Γράφημα πίτας",histogram:"Ιστόγραμμα",scatterPlot:"Διάγραμμα διασποράς",boxPlot:"Διάγραμμα πλαισίου"},a={mean:"Μέσος όρος",median:"Διάμεσο",sum:"Άθροισμα",discretePercentile:"Διακριτό ποσοστημόριο",minimum:"Ελάχιστο",maximum:"Μέγιστο",variance:"Διακύμανση",count:"Πλήθος",aggregation:"${ statistics } από ${ fieldName }",noAggregation:"Δεν υπάρχει συνάθροιση"},e="Γράφημα",i="Σειρά μέτρησης",s="Άξονας Χ",r="Μη διαθέσιμο",n={start:"Συνδεθείτε με το πρώτο σημείο δεδομένων",end:"Συνδεθείτε με το τελευταίο σημείο δεδομένων"};var o={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:s,notAvailable:r,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = s;
