// @arcgis/charts-components/dist/cjs/messages.th-63ead210-df1e4792.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="เกิดข้อผิดพลาดขึ้นขณะโหลดแผนภูมิ",e="ไม่สามารถใช้การแปลงค่า log กับค่าติดลบหรือศูนย์ได้",t="ไม่สามารถใช้การแปลงรากที่สองกับค่าติดลบได้",r="เกิดข้อผิดพลาดขึ้นขณะโหลดชั้นข้อมูล URL = ${ url } ID รายการพอร์ทัล = ${ portalItemId }",i="ไม่พบฟิลด์ [${ fieldNames }] ที่กำหนดค่าไว้ในคอนฟิกในชั้นข้อมูล",d="${ dataPath } ต้องไม่ซ้ำกัน ซีรีส์ที่ชื่อ ${ seriesName } มี ID (${ seriesID }) ที่ซีรีส์อื่นใช้งานอยู่แล้ว",o="${ dataPath } ไม่สามารถดำเนินการรวมแบบไม่นับกับฟิลด์ที่ไม่ใช่ตัวเลขได้",l="${ dataPath } ไม่มีชื่อคุณสมบัติ ${ missingProperty }",n="${ dataPath } ต้องมีความยาวมากกว่า ${ limit } อักขระ",$="${ dataPath } ต้องมีไม่น้อยกว่า ${ limit } รายการ",s="${ dataPath } ต้องมีไม่เกิน ${ limit } รายการ",m="${ dataPath } ต้องมีอย่างน้อยหนึ่งอักขระที่ไม่ใช่ไวท์สเปซ",u="${ dataPath } ต้องไม่มี ${ additionalProperty }",P="${ dataPath } ต้องเท่ากับหนึ่งในค่าที่อนุญาตเหล่านี้: ${ allowedValues }",h="${ dataPath } ต้องตรงกับรูปแบบของหนึ่งในรายการเหล่านี้: ${ schemaOptions }",p="ไม่รองรับพล็อตการกระจายที่มีสัญลักษณ์ตามสัดส่วน ใช้ขนาดสัญลักษณ์เริ่มต้นแล้ว",y="ไม่สามารถอ่านข้อมูลขาเข้า",I="ประเภทซีรีส์ที่คาดหวังที่ดัชนี ${ seriesIndex } คือ '${ expectedType }' แต่ได้รับ '${ receivedType }' แทน",g="หรือ";var c={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:d,nonNumericAggregation:o,requiredProperty:l,minLength:n,minItems:$,maxItems:s,whiteSpacePattern:m,additionalProperty:u,enumValues:P,anyOfValues:h,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:I,or:g};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = p;
exports['default'] = c;
exports.defaultError = a;
exports.duplicateSeriesID = d;
exports.enumValues = P;
exports.invalidSeriesType = I;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = s;
exports.minItems = $;
exports.minLength = n;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = o;
exports.or = g;
exports.queryError = y;
exports.requiredProperty = l;
exports.whiteSpacePattern = m;
