// @arcgis/charts-components/dist/cjs/messages.sk-7e8fa059-872eab16.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const i={chart:"Graf",barChart:"Pruhový graf",lineChart:"Čiarový graf",pieChart:"Koláčový graf",histogram:"Histogram",scatterPlot:"Korelačný graf",boxPlot:"Krabicový graf"},t={mean:"Priemer",median:"Medián",sum:"Súčet",discretePercentile:"Diskrétny percentil",minimum:"Minimum",maximum:"Maximum",variance:"Rozptyl",count:"Počet",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Bez agregácie"},e="Graf",a="Počet sérií",r="Os X",o="Nie je k dispozícii",m={start:"Prichytiť k prvému dátovému bodu",end:"Prichytiť k poslednému dátovému bodu"};var s={chartType:i,statistics:t,defaultTitle:e,countSeries:a,xAxis:r,notAvailable:o,timeAggregationTypes:m};

exports.chartType = i;
exports.countSeries = a;
exports['default'] = s;
exports.defaultTitle = e;
exports.notAvailable = o;
exports.statistics = t;
exports.timeAggregationTypes = m;
exports.xAxis = r;
