// @arcgis/charts-components/dist/cjs/messages.de-e2ce6e5c-901ed512.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Diagramm",barChart:"Balkendiagramm",lineChart:"Liniendiagramm",pieChart:"Kreisdiagramm",histogram:"Histogramm",scatterPlot:"Scatterplot",boxPlot:"Boxplot"},t={mean:"Mittelwert",median:"Medianwert",sum:"Summe",discretePercentile:"Diskontinuierliches Perzentil",minimum:"Minimum",maximum:"Maximum",variance:"Varianz",count:"Anzahl",aggregation:"${ statistics } von ${ fieldName }",noAggregation:"Keine Aggregation"},a="Diagramm",i="Mengenserie",n="X-Achse",r="Nicht verfügbar",m={start:"Am ersten Datenpunkt fangen",end:"Am letzten Datenpunkt fangen"};var g={chartType:e,statistics:t,defaultTitle:a,countSeries:i,xAxis:n,notAvailable:r,timeAggregationTypes:m};

exports.chartType = e;
exports.countSeries = i;
exports['default'] = g;
exports.defaultTitle = a;
exports.notAvailable = r;
exports.statistics = t;
exports.timeAggregationTypes = m;
exports.xAxis = n;
