// @arcgis/charts-components/dist/cjs/messages.hu-f801472e-8970409f-47065a28.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Oszlop-/sávdiagram",lineChart:"Vonaldiagram",pieChart:"Kördiagram",histogram:"Hisztogram",scatterPlot:"Szórásdiagram",boxPlot:"Boxdiagram"},t={mean:"Átlag",median:"Medián",sum:"Összeg",discretePercentile:"Diszkrét percentilis",minimum:"Minimum",maximum:"Maximum",variance:"Variancia",count:"Darabszám",aggregation:"${ statistics }/${ fieldName }",noAggregation:"Nincs összevonás"},i="Diagram",s="Számsorozat",e="X tengely",r="Nem érhető el",o={start:"Csatolás az első adatponthoz",end:"Csatolás az utolsó adatponthoz"};var m={chartType:a,statistics:t,defaultTitle:i,countSeries:s,xAxis:e,notAvailable:r,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = s;
exports['default'] = m;
exports.defaultTitle = i;
exports.notAvailable = r;
exports.statistics = t;
exports.timeAggregationTypes = o;
exports.xAxis = e;
