// @arcgis/charts-components/dist/cjs/messages.sl-fc1a9dc6-ece11f30.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Grafikon",barChart:"Palični grafikon",lineChart:"Linijski grafikon",pieChart:"Tortni grafikon",histogram:"Histogram",scatterPlot:"Raztreseni grafikon",boxPlot:"Škatlasti graf"},i={mean:"Srednja vrednost",median:"Mediana",sum:"Vsota",discretePercentile:"Diskretni percentil",minimum:"Minimum",maximum:"Maksimum",variance:"Varianca",count:"Število",aggregation:"${ statistics } od ${ fieldName }",noAggregation:"Brez agregacije"},t="Grafikon",e="Števec nizov",o="Os X",n="Ni na voljo",r={start:"Privlači na prvo podatkovno točko",end:"Privlači na zadnjo podatkovno točko"};var s={chartType:a,statistics:i,defaultTitle:t,countSeries:e,xAxis:o,notAvailable:n,timeAggregationTypes:r};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = s;
exports.defaultTitle = t;
exports.notAvailable = n;
exports.statistics = i;
exports.timeAggregationTypes = r;
exports.xAxis = o;
