// @arcgis/charts-components/dist/cjs/messages.et-79ae70c6-41a51a41-bb815c18.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagramm",barChart:"Lintdiagramm",lineChart:"Joondiagramm",pieChart:"Sektordiagramm",histogram:"Histogramm",scatterPlot:"Punktdiagramm",boxPlot:"Karpdiagramm"},e={mean:"Keskmine",median:"Keskmine",sum:"Summa",discretePercentile:"Diskreetne protsentiil",minimum:"Miinimum",maximum:"Maksimum",variance:"Dispersioon",count:"Kogus",aggregation:"${ statistics } /${ fieldName }",noAggregation:"Agregeerimist pole"},i="Diagramm",t="Sarjade arv",s="X-telg",m="Pole kättesaadav",r={start:"Haagi esimesse andmepunkti",end:"Haagi viimasesse andmepunkti"};var n={chartType:a,statistics:e,defaultTitle:i,countSeries:t,xAxis:s,notAvailable:m,timeAggregationTypes:r};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = m;
exports.statistics = e;
exports.timeAggregationTypes = r;
exports.xAxis = s;
