// @arcgis/charts-components/dist/cjs/messages.lt-437315dc-edd025ab.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagrama",barChart:"Stulpelinė diagrama",lineChart:"Linijinė diagrama",pieChart:"Skritulinė diagrama",histogram:"Histograma",scatterPlot:"Sklaidos diagrama",boxPlot:"Stačiakampė diagrama"},i={mean:"Vidurkis",median:"Mediana",sum:"Suma",discretePercentile:"Diskretinis procentilis",minimum:"Minimumas",maximum:"Maksimumas",variance:"Nuokrypis",count:"Bendras skaičius",aggregation:"${ statistics } iš ${ fieldName }",noAggregation:"Be agregavimo"},t="Diagrama",r="Skaičių serija",e="X ašis",s="Negalima",m={start:"Pritraukti pirmą duomenų tašką",end:"Pritraukti paskutinį duomenų tašką"};var n={chartType:a,statistics:i,defaultTitle:t,countSeries:r,xAxis:e,notAvailable:s,timeAggregationTypes:m};

exports.chartType = a;
exports.countSeries = r;
exports['default'] = n;
exports.defaultTitle = t;
exports.notAvailable = s;
exports.statistics = i;
exports.timeAggregationTypes = m;
exports.xAxis = e;
