// @arcgis/charts-components/dist/cjs/messages.fr-b992d801-89c0ef56-2a8f9d72-616d3477.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Une erreur est survenue lors du chargement du diagramme.",a="Impossible d'appliquer une transformation logarithmique à des valeurs négatives ou égales à zéro.",r="Impossible d'appliquer une transformation de racine carrée à des valeurs négatives.",t="Une erreur est survenue lors du chargement de la couche. URL = ${ url }. ID de l'élément du portail = ${ portalItemId }.",s="Les champs [${ fieldNames }] définis dans la configuration sont introuvables dans la couche.",n="${ dataPath } doit être unique. La série nommée ${ seriesName } comporte un ID (${ seriesID }) qui est déjà utilisé par une autre série.",i="${ dataPath } ne peut pas réaliser d'agrégation de type autre que Total sur un champ non numérique.",o="Propriété ${ missingProperty } manquante dans ${ dataPath }.",u="${ dataPath } ne doit pas comporter moins de ${ limit } caractères.",d="${ dataPath } ne doit pas comporter moins de ${ limit } éléments.",l="${ dataPath } ne doit pas comporter plus de ${ limit } éléments.",m="${ dataPath } doit contenir au moins un caractère autre qu'un espace.",p="${ dataPath } ne doit pas contenir ${ additionalProperty }.",c="${ dataPath } doit être égal à l'une de ces valeurs autorisées : ${ allowedValues }.",g="${ dataPath } doit correspondre à l'une des structures suivantes : ${ schemaOptions }.",h="Les nuages de points avec des symboles proportionnels ne sont pas pris en charge. La taille de symbole par défaut a été appliquée.",$="Échec de la lecture des données en entrée.",y="Le type de série attendu à l’index ${ seriesIndex } est '${ expectedType }', mais '${ receivedType }' a été reçu à la place",P="ou";var v={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:r,layerLoadFailure:t,layerFieldsNotFound:s,duplicateSeriesID:n,nonNumericAggregation:i,requiredProperty:o,minLength:u,minItems:d,maxItems:l,whiteSpacePattern:m,additionalProperty:p,enumValues:c,anyOfValues:g,bubbleChartValidateMsg:h,queryError:$,invalidSeriesType:y,or:"ou"};

exports.additionalProperty = p;
exports.anyOfValues = g;
exports.bubbleChartValidateMsg = h;
exports['default'] = v;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = c;
exports.invalidSeriesType = y;
exports.layerFieldsNotFound = s;
exports.layerLoadFailure = t;
exports.maxItems = l;
exports.minItems = d;
exports.minLength = u;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = r;
exports.nonNumericAggregation = i;
exports.or = P;
exports.queryError = $;
exports.requiredProperty = o;
exports.whiteSpacePattern = m;
