// @arcgis/charts-components/dist/cjs/messages.fi-cc657f3a-74b0cfc9-a829961e.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const i={chart:"Kaavio",barChart:"Palkkikaavio",lineChart:"Viivakaavio",pieChart:"Ympyräkaavio",histogram:"Histogrammi",scatterPlot:"Pistekaavio",boxPlot:"Laatikkojana"},a={mean:"Keskiarvo",median:"Mediaani",sum:"Summa",discretePercentile:"Irrallinen prosenttipiste",minimum:"Vähimmäisarvo",maximum:"Enimmäisarvo",variance:"Varianssi",count:"Määrä",aggregation:"${ statistics }/${ fieldName }",noAggregation:"Ei koostetta"},e="Kaavio",t="Laske sarjat",s="X-akseli",n="Ei käytettävissä",o={start:"Kiinnitä ensimmäiseen aineistopisteeseen",end:"Kiinnitä viimeiseen aineistopisteeseen"};var r={chartType:i,statistics:a,defaultTitle:e,countSeries:t,xAxis:s,notAvailable:n,timeAggregationTypes:o};

exports.chartType = i;
exports.countSeries = t;
exports['default'] = r;
exports.defaultTitle = e;
exports.notAvailable = n;
exports.statistics = a;
exports.timeAggregationTypes = o;
exports.xAxis = s;
