// @arcgis/charts-components/dist/cjs/messages.th-c923afca-522ce8df.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"แผนภูมิ",barChart:"แผนภูมิแท่ง",lineChart:"แผนภูมิเส้น",pieChart:"แผนภูมิทรงกลม",histogram:"ฮีสโตแกรม",scatterPlot:"พล็อตการกระจาย",boxPlot:"พล็อตแบบกล่อง"},a={mean:"ค่าเฉลี่ย",median:"ค่ามัธยฐาน",sum:"รวม",discretePercentile:"เปอร์เซนไทล์แบบไม่ต่อเนื่อง",minimum:"น้อยสุด",maximum:"สูงสุด",variance:"ความแปรปรวน",count:"นับ",aggregation:"${ statistics } จาก ${ fieldName }",noAggregation:"ไม่มีการรวม"},e="แผนภูมิ",i="นับซีรีส์",r="(แกน x)",s="ไม่สามารถใช้ได้",n={start:"สแน็ปเข้ากับจุดข้อมูลแรก",end:"สแน็ปเข้ากับจุดข้อมูลสุดท้าย"};var o={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:r,notAvailable:s,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = r;
