// @arcgis/charts-components/dist/cjs/messages.fi-c0c48057-bbece429-01ffd80f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Kaavion lataamisessa tapahtui virhe.",t="Logaritmimuunnosta ei voi käyttää negatiivisiin arvoihin tai nolla-arvoihin.",e="Neliöjuurimuunnosta ei voi käyttää negatiivisiin arvoihin.",i="Karttatason lataamisessa tapahtui virhe. URL = ${ url }. Portaalikohteen tunnus = ${ portalItemId }.",s="Määrityksessä määritettyjä kenttiä [${ fieldNames }] ei löytynyt karttatasolta.",n="Kohteen ${ dataPath } on oltava yksilöllinen. Sarjan ${ seriesName } tunnus (${ seriesID }) on jo toisen sarjan käytössä.",o="${ dataPath } ei saa suorittaa muuta kuin määrän koostamista muussa kuin numerokentässä.",r="Kohteesta ${ dataPath } puuttuu ominaisuus ${ missingProperty }.",l="${ dataPath } ei saa olla alle ${ limit } merkkiä pitkä.",u="Kohteessa ${ dataPath } ei saa olla alle ${ limit } kohdetta.",h="Kohteessa ${ dataPath } ei saa olla yli ${ limit } kohdetta.",m="Kohteessa ${ dataPath } on oltava vähintään yksi muu kuin välimerkki.",d="Kohteessa ${ dataPath } ei saa olla kohdetta ${ additionalProperty }.",k="Kohteen ${ dataPath } tulee olla yhtä suuri kuin jokin näistä sallituista arvoista: ${ allowedValues }.",y="Kohteen ${ dataPath } on vastattava jotakin näistä skeemoista: ${ schemaOptions }.",$="Pistekaavioita, joissa on suhteutettuja symboleja, ei tueta. Symbolin oletuskokoa on käytetty.",v="Lähtöaineiston lukeminen epäonnistui.",p='Odotettu sarjan tyyppi indeksissä ${ seriesIndex } on "${ expectedType }", mutta sen sijaan vastaanotettiin "${ receivedType }"',P="tai";var g={defaultError:a,negativeValueInDataForLogTransformation:t,negativeValueInDataForSqrtTransformation:e,layerLoadFailure:i,layerFieldsNotFound:s,duplicateSeriesID:n,nonNumericAggregation:o,requiredProperty:r,minLength:l,minItems:u,maxItems:h,whiteSpacePattern:m,additionalProperty:d,enumValues:k,anyOfValues:y,bubbleChartValidateMsg:$,queryError:v,invalidSeriesType:p,or:"tai"};

exports.additionalProperty = d;
exports.anyOfValues = y;
exports.bubbleChartValidateMsg = $;
exports['default'] = g;
exports.defaultError = a;
exports.duplicateSeriesID = n;
exports.enumValues = k;
exports.invalidSeriesType = p;
exports.layerFieldsNotFound = s;
exports.layerLoadFailure = i;
exports.maxItems = h;
exports.minItems = u;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = t;
exports.negativeValueInDataForSqrtTransformation = e;
exports.nonNumericAggregation = o;
exports.or = P;
exports.queryError = v;
exports.requiredProperty = r;
exports.whiteSpacePattern = m;
