// @arcgis/charts-components/dist/cjs/messages.sv-18b29f8a-32c2bb8c-a675630f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Liggande stapeldiagram",lineChart:"Linjediagram",pieChart:"Cirkeldiagram",histogram:"Histogram",scatterPlot:"Sambandsdiagram",boxPlot:"Låddiagram"},e={mean:"Medelvärde",median:"Median",sum:"Summa",discretePercentile:"Diskret percentil",minimum:"Minimum",maximum:"Maximum",variance:"Varians",count:"Antal",aggregation:"${ statistics } av ${ fieldName }",noAggregation:"Ingen aggregering"},t="Diagram",i="Antalsserie",n="X-axel",s="Ej tillgänglig",r={start:"Snappa till den första datapunkten",end:"Snappa till den sista datapunkten"};var g={chartType:a,statistics:e,defaultTitle:t,countSeries:i,xAxis:n,notAvailable:s,timeAggregationTypes:r};

exports.chartType = a;
exports.countSeries = i;
exports['default'] = g;
exports.defaultTitle = t;
exports.notAvailable = s;
exports.statistics = e;
exports.timeAggregationTypes = r;
exports.xAxis = n;
