// @arcgis/charts-components/dist/cjs/messages.es-907baf0a-96613cec-8925b54a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Gráfico",barChart:"Gráfico de barras",lineChart:"Gráfico de líneas",pieChart:"Gráfico circular",histogram:"Histograma",scatterPlot:"Gráfico de dispersión",boxPlot:"Diagrama de caja"},a={mean:"Valor medio",median:"Mediana",sum:"Suma",discretePercentile:"Percentil discreto",minimum:"Mínimo",maximum:"Máximo",variance:"Varianza",count:"Recuento",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sin agregación"},i="Gráfico",t="Recuento de series",r="Eje X",s="No disponible",o={start:"Alinear con el primer punto de datos",end:"Alinear con el último punto de datos"};var n={chartType:e,statistics:a,defaultTitle:i,countSeries:t,xAxis:r,notAvailable:s,timeAggregationTypes:o};

exports.chartType = e;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = o;
exports.xAxis = r;
