// @arcgis/charts-components/dist/cjs/messages.sr-b8d5fefe-188a6f3f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Grafikon",barChart:"Trakasti grafikon",lineChart:"Linijski grafikon",pieChart:"Kružni grafikon",histogram:"Histogram",scatterPlot:"Grafikon",boxPlot:"Kvadratni grafikon"},i={mean:"Sredina",median:"Medijana",sum:"Zbir",discretePercentile:"Diskretni percentil",minimum:"Minimum",maximum:"Maksimum",variance:"Odstupanje",count:"Broj",aggregation:"${ statistics } od ${ fieldName }",noAggregation:"Nema grupisanja"},t="Grafikon",r="Broj serija",e="X-osa",n="Nije dostupno",o={start:"Zakači za prvu tačku podataka",end:"Zakači za zadnju tačku podataka"};var s={chartType:a,statistics:i,defaultTitle:t,countSeries:r,xAxis:e,notAvailable:n,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = r;
exports['default'] = s;
exports.defaultTitle = t;
exports.notAvailable = n;
exports.statistics = i;
exports.timeAggregationTypes = o;
exports.xAxis = e;
