// @arcgis/charts-components/dist/cjs/messages.es-5242724e-fe3be8dc-3af212ef-c1e59caa.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Se ha producido un error al cargar el gráfico.",a="No se puede aplicar una transformación del registro a valores negativos o de cero.",r="No se puede aplicar una transformación de raíz cuadrada a valores negativos.",o="Se ha producido un error al cargar la capa. URL = ${ url }. Id. de elemento de portal = ${ portalItemId }.",n="Los campos [${ fieldNames }] definidos en la configuración no se han encontrado en la capa.",t="El elemento ${ dataPath } debe ser único. La serie denominada ${ seriesName } tiene un Id. (${ seriesID }) que ya se utiliza en otra serie.",i="${ dataPath } no puede realizar una agregación no numérica en un campo no numérico.",s="A ${ dataPath } le falta la propiedad denominada ${ missingProperty }.",d="${ dataPath } no debe tener menos de ${ limit } caracteres.",l="${ dataPath } no debe tener menos de ${ limit } elementos.",u="${ dataPath } no debe tener más de ${ limit } elementos.",m="${ dataPath } debe tener al menos un carácter que no sea un espacio en blanco.",c="${ dataPath } no debe tener ${ additionalProperty }",p="${ dataPath } debe ser igual a uno de los siguientes valores permitidos: ${ allowedValues }.",g="${ dataPath } debe coincidir con el esquema de uno de estos: ${ schemaOptions }.",$="No se admiten diagramas de dispersión con símbolos proporcionales. Se ha aplicado el tamaño de símbolo predeterminado.",h="Error al leer los datos de entrada.",y='El tipo de serie esperado en el índice ${ seriesIndex } es "${ expectedType }", pero en su lugar se ha recibido "${ receivedType }"',b="o";var P={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:r,layerLoadFailure:o,layerFieldsNotFound:n,duplicateSeriesID:t,nonNumericAggregation:i,requiredProperty:s,minLength:d,minItems:l,maxItems:u,whiteSpacePattern:m,additionalProperty:c,enumValues:p,anyOfValues:g,bubbleChartValidateMsg:$,queryError:h,invalidSeriesType:y,or:"o"};

exports.additionalProperty = c;
exports.anyOfValues = g;
exports.bubbleChartValidateMsg = $;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = t;
exports.enumValues = p;
exports.invalidSeriesType = y;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = o;
exports.maxItems = u;
exports.minItems = l;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = r;
exports.nonNumericAggregation = i;
exports.or = b;
exports.queryError = h;
exports.requiredProperty = s;
exports.whiteSpacePattern = m;
