// @arcgis/charts-components/dist/cjs/messages.he-70d41af0-2741f049-cbac9958.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="אירעה שגיאה במהלך טעינת התרשים.",e="לא ניתן לבצע טרנספורמציה לוגריתמית על ערכים שליליים או על אפס.",t="לא ניתן לבצע טרנספורמציה של שורש ריבועי על ערכים שליליים.",r="אירעה שגיאה במהלך טעינת השכבה. URL = ${ url }. מזהה פריט פורטל =${ portalItemId }.",i="השדות [${ fieldNames }] שהוגדרו ב-config לא נמצאו בשכבה.",s="${ dataPath } חייב/ת להיות ייחודי/ת. לסדרה בשם ${ seriesName } יש מזהה (${ seriesID }) שבו משתמשת כבר סדרה אחרת.",n="ל-${ dataPath } אסור לבצע צבירה שאינה ספירה על שדה לא מספרי.",o="ב-${ dataPath } חסר מאפיין בשם ${ missingProperty }.",d="${ dataPath } לא יכול להיות קצר יות מ-${ limit } תווים.",l="ל-${ dataPath } לא יכולים להיות פחות מ-${ limit } פריטים.",u="ל-${ dataPath } לא יכולים להיות יותר מ-${ limit } פריטים.",m="ל-${ dataPath } חייב להיות לפחות תו אחד שאינו רווח בלתי נראה.",$="ל-${ dataPath } אסור לכלול ${ additionalProperty }.",g="${ dataPath } חייב להיות שווה לאחד מהערכים המותרים הבאים: ${ allowedValues }.",p="${ dataPath } חייב להתאים לסכמה של אחד מאלה: ${ schemaOptions }.",y="תרשימי פיזור עם סמלים פרופורציונליים לא נתמכים. הוחל גודל סמל שמוגדר כברירת מחדל.",P="קריאת נתוני הקלט נכשלה.",h="סוג הסדרה הצפוי במדד ${ seriesIndex } הוא '${ expectedType }', אבל במקום זאת התקבל '${ receivedType }'.",I="או";var c={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:d,minItems:l,maxItems:u,whiteSpacePattern:m,additionalProperty:$,enumValues:g,anyOfValues:p,bubbleChartValidateMsg:y,queryError:P,invalidSeriesType:h,or:"או"};

exports.additionalProperty = $;
exports.anyOfValues = p;
exports.bubbleChartValidateMsg = y;
exports['default'] = c;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = g;
exports.invalidSeriesType = h;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = l;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = n;
exports.or = I;
exports.queryError = P;
exports.requiredProperty = o;
exports.whiteSpacePattern = m;
