// @arcgis/charts-components/dist/cjs/messages.en-US-d444015e-7fcc1245-9a621bd4-c7314dc7.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="An error occurred while loading the chart.",a="Cannot apply a log transformation to negative or zero values.",t="Cannot apply a square root transformation to negative values.",r="An error occurred while loading the layer. URL = ${ url }. Portal item ID = ${ portalItemId }.",o="The fields [${ fieldNames }] defined in the config were not found in the layer.",n="${ dataPath } must be unique. The series named ${ seriesName } has an ID (${ seriesID }) that is already used by another series.",s="${ dataPath } cannot perform non-count aggregation on non-numeric field.",i="${ dataPath } is missing a property named ${ missingProperty }.",l="${ dataPath } must not be shorter than ${ limit } characters.",d="${ dataPath } must not have fewer than ${ limit } items.",u="${ dataPath } must not have more than ${ limit } items.",m="${ dataPath } must have at least one non-whitespace character.",h="${ dataPath } must not have ${ additionalProperty }.",p="${ dataPath } must be equal to one of these allowed values: ${ allowedValues }.",c="${ dataPath } must match the schema of one of these: ${ schemaOptions }.",y="Scatterplots with proportional symbols are not supported. Default symbol size has been applied.",g="Failure to read input data.",$="The expected series type at index ${ seriesIndex } is '${ expectedType }' but '${ receivedType }' was received instead",f="or";var P={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:o,duplicateSeriesID:n,nonNumericAggregation:s,requiredProperty:i,minLength:l,minItems:d,maxItems:u,whiteSpacePattern:m,additionalProperty:h,enumValues:p,anyOfValues:c,bubbleChartValidateMsg:y,queryError:g,invalidSeriesType:$,or:"or"};

exports.additionalProperty = h;
exports.anyOfValues = c;
exports.bubbleChartValidateMsg = y;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = p;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = o;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = s;
exports.or = f;
exports.queryError = g;
exports.requiredProperty = i;
exports.whiteSpacePattern = m;
