// @arcgis/charts-components/dist/cjs/messages.hr-a0968519-b08fb091.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Grafikon",barChart:"Trakasti grafikon",lineChart:"Linijski grafikon",pieChart:"Tortni grafikon",histogram:"Histogram",scatterPlot:"Dijagram raspršivanja",boxPlot:"Dijagram pravokutnika"},i={mean:"Prosjek",median:"Srednja vrijednost",sum:"Zbroj",discretePercentile:"Diskretni postotak",minimum:"Minimum",maximum:"Maksimum",variance:"Varijanca",count:"Zbroj",aggregation:"${ statistics } od ${ fieldName }",noAggregation:"Nema agregacije"},t="Grafikon",r="Broj serija",e="Os x",o="Nije dostupno",n={start:"Prilijepi na prvu podatkovnu točku",end:"Prilijepi na posljednju podatkovnu točku"};var s={chartType:a,statistics:i,defaultTitle:t,countSeries:r,xAxis:e,notAvailable:o,timeAggregationTypes:n};

exports.chartType = a;
exports.countSeries = r;
exports['default'] = s;
exports.defaultTitle = t;
exports.notAvailable = o;
exports.statistics = i;
exports.timeAggregationTypes = n;
exports.xAxis = e;
