// @arcgis/charts-components/dist/cjs/messages.vi-087f5bf7-631dd062.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const i={chart:"Biểu đồ",barChart:"Biểu đồ cột",lineChart:"Biểu đồ dạng đường",pieChart:"Biểu đồ hình tròn",histogram:"Biểu đồ tần suất",scatterPlot:"Biểu đồ phân tán",boxPlot:"Biểu đồ hộp"},n={mean:"Trung bình",median:"Bình quân",sum:"Tổng",discretePercentile:"Phân vị rời rạc",minimum:"Tối thiểu",maximum:"Tối đa",variance:"Phương sai",count:"Số lượng",aggregation:"${ statistics } / ${ fieldName }",noAggregation:"Không tổng hợp"},t="Biểu đồ",a="Đếm Số chuỗi",e="Trục X",u="Không Khả dụng",g={start:"Chuyển sang điểm dữ liệu đầu tiên",end:"Chuyển sang điểm dữ liệu cuối cùng"};var r={chartType:i,statistics:n,defaultTitle:t,countSeries:a,xAxis:e,notAvailable:u,timeAggregationTypes:g};

exports.chartType = i;
exports.countSeries = a;
exports['default'] = r;
exports.defaultTitle = t;
exports.notAvailable = u;
exports.statistics = n;
exports.timeAggregationTypes = g;
exports.xAxis = e;
