// @arcgis/charts-components/dist/cjs/messages.tr-7495cb63-0f6ff45f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Çizge",barChart:"Çubuk grafik",lineChart:"Çizgi grafik",pieChart:"Pasta grafik",histogram:"Histogram",scatterPlot:"Dağılım grafiği",boxPlot:"Kutu düzeni"},i={mean:"Ortalama",median:"Medyan",sum:"Toplam",discretePercentile:"Ayrık yüzdelik",minimum:"Minimum",maximum:"Maksimum",variance:"Varyans",count:"Sayım",aggregation:"${ statistics } / ${ fieldName }",noAggregation:"Kümeleme yok"},e="Grafik",t="Serileri Say",r="X Ekseni",n="Kullanılabilir Değil",l={start:"İlk veri noktasına yerleştir",end:"Son veri noktasına yerleştir"};var m={chartType:a,statistics:i,defaultTitle:e,countSeries:t,xAxis:r,notAvailable:n,timeAggregationTypes:l};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = m;
exports.defaultTitle = e;
exports.notAvailable = n;
exports.statistics = i;
exports.timeAggregationTypes = l;
exports.xAxis = r;
