// @arcgis/charts-components/dist/cjs/messages.cs-f7dfa0ea-b1a79f2d-fec71da0-a1050c31.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Při načítání grafu došlo k chybě.",e="Nelze uplatnit logaritmickou transformaci na záporné nebo nulové hodnoty.",t="Nelze uplatnit transformaci odmocniny na záporné hodnoty.",o="Při načítání vrstvy došlo k chybě. URL = ${ url }. ID položky portálu = ${ portalItemId }.",n="Pole [${ fieldNames }] definovaná v konfiguraci nebyla nalezena ve vrstvě.",r="${ dataPath } musí být jedinečné. Řada s názvem ${ seriesName } má ID (${ seriesID }), které již používá jiná řada.",s="${ dataPath } nemůže provádět jiné seskupení než podle počtu na jiném než numerickém poli.",i="V ${ dataPath } chybí vlastnost s názvem ${ missingProperty }.",l="${ dataPath } nesmí být kratší než ${ limit } znaků.",d="${ dataPath } nesmí obsahovat méně než ${ limit } položek.",m="${ dataPath } nesmí obsahovat více než ${ limit } položek.",u="${ dataPath } musí obsahovat alespoň jeden znak jiný než mezera.",p="${ dataPath } nesmí obsahovat ${ additionalProperty }.",y="${ dataPath } se musí rovnat jedné z těchto povolených hodnot: ${ allowedValues }.",v="${ dataPath } se musí shodovat s jedním z těchto schémat: ${ schemaOptions }.",h="Rozptýlené bodové grafy s proporcionálními symboly nejsou podporovány. Byla použita výchozí velikost symbolu.",c="Vstupní data se nepodařilo přečíst.",$="Očekávaný typ řady s indexem ${ seriesIndex } je '${ expectedType }', ale obdrželi jsme '${ receivedType }'",b="nebo";var g={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:o,layerFieldsNotFound:n,duplicateSeriesID:r,nonNumericAggregation:s,requiredProperty:i,minLength:l,minItems:d,maxItems:m,whiteSpacePattern:u,additionalProperty:p,enumValues:y,anyOfValues:v,bubbleChartValidateMsg:h,queryError:c,invalidSeriesType:$,or:b};

exports.additionalProperty = p;
exports.anyOfValues = v;
exports.bubbleChartValidateMsg = h;
exports['default'] = g;
exports.defaultError = a;
exports.duplicateSeriesID = r;
exports.enumValues = y;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = o;
exports.maxItems = m;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = s;
exports.or = b;
exports.queryError = c;
exports.requiredProperty = i;
exports.whiteSpacePattern = u;
