// @arcgis/charts-components/dist/cjs/messages.et-62f107f7-84b2a372-5b84ea80-e13facf8.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Diagrammi laadimisel ilmnes viga.",e="Logi teisendamisel ei saa kasutada negatiivseid väärtusi ega nullväärtusi.",i="Ruutjuure teisendamist ei saa rakendada negatiivsetele väärtustele.",t="Kihi laadimisel tekkis viga. URL = ${ url }. Portaaliüksuse ID = ${ portalItemId }.",s="Konfiguratsioonis määratletud välju [${ fieldNames }] kihis ei leitud.",r="${ dataPath } peab olema kordumatu. Seerial nimega ${ seriesName } on ID (${ seriesID }), mida kasutatakse juba muudes seeriates.",l="${ dataPath } ei tohi mittenumbrilisel väljal teha loendamisega mitteseotud agregeerimist.",u="Omadusel ${ dataPath } puudub omadus nimega ${ missingProperty }.",o="${ dataPath } ei tohi olla lühem kui ${ limit } märki.",d="Muutujal ${ dataPath } ei tohi olla vähem kui ${ limit } üksust.",n="Muutujal ${ dataPath } ei tohi olla rohkem kui ${ limit } üksust.",m="${ dataPath } peab sisaldama vähemalt ühte märki, mis pole tühimärk.",h="Muutujal ${ dataPath } ei tohi olla: ${ additionalProperty }.",g="${ dataPath } peab võrduma ühega allpool nimetatud väärtustest: ${ allowedValues }.",p="${ dataPath } peab vastama ühele nendest skeemidest: ${ schemaOptions }.",v="Proportsionaalsete sümbolitega hajuvusdiagramme ei toetata. Rakendatud on sümboli vaikesuurus.",$="Sisendandmete lugemine nurjus.",k="Eeldatud seeriatüüp indeksis ${ seriesIndex } on '${ expectedType }', kuid saadi hoopis '${ receivedType }'",P="või";var y={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:t,layerFieldsNotFound:s,duplicateSeriesID:r,nonNumericAggregation:l,requiredProperty:u,minLength:o,minItems:d,maxItems:n,whiteSpacePattern:m,additionalProperty:h,enumValues:g,anyOfValues:p,bubbleChartValidateMsg:v,queryError:$,invalidSeriesType:k,or:"või"};

exports.additionalProperty = h;
exports.anyOfValues = p;
exports.bubbleChartValidateMsg = v;
exports['default'] = y;
exports.defaultError = a;
exports.duplicateSeriesID = r;
exports.enumValues = g;
exports.invalidSeriesType = k;
exports.layerFieldsNotFound = s;
exports.layerLoadFailure = t;
exports.maxItems = n;
exports.minItems = d;
exports.minLength = o;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = l;
exports.or = P;
exports.queryError = $;
exports.requiredProperty = u;
exports.whiteSpacePattern = m;
