// @arcgis/charts-components/dist/cjs/messages.no-950d514a-73cd9d94-c965944c.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Stolpediagram",lineChart:"Linjediagram",pieChart:"Sektordiagram",histogram:"Histogram",scatterPlot:"Punktdiagram",boxPlot:"Boksdiagram"},t={mean:"Gjennomsnitt",median:"Median",sum:"Sum",discretePercentile:"Diskret persentil",minimum:"Minimum",maximum:"Maksimum",variance:"Varians",count:"Antall",aggregation:"${ statistics } av ${ fieldName }",noAggregation:"Ingen sammenslåing"},e="Diagram",i="Tallserie",s="X-akse",n="Ikke tilgjengelig",r={start:"Fest til det første datapunktet",end:"Fest til det siste datapunktet"};var m={chartType:a,statistics:t,defaultTitle:e,countSeries:i,xAxis:s,notAvailable:n,timeAggregationTypes:r};

exports.chartType = a;
exports.countSeries = i;
exports['default'] = m;
exports.defaultTitle = e;
exports.notAvailable = n;
exports.statistics = t;
exports.timeAggregationTypes = r;
exports.xAxis = s;
