// @arcgis/charts-components/dist/cjs/messages.sv-8a49b669-29326c1b-4f678069.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Ett fel uppstod när diagrammet lästes in.",e="Det går inte att tillämpa loggomvandling på negativa värden eller nollvärden.",t="Det går inte att tillämpa kvadratrotomvandling på negativa värden.",r="Ett fel uppstod när lagret lästes in. URL = ${ url }. Portalobjekt-ID = ${ portalItemId }.",n="Fälten [${ fieldNames }] som är definierade i konfigurationen hittades inte i lagret.",i="${ dataPath } måste vara unik. Serien som heter ${ seriesName } har ett ID (${ seriesID }) som redan används av en annan serie.",s="${ dataPath } kan inte utföra icke-tidsmässig aggregering på ett icke-numeriskt fält.",l="${ dataPath } saknar egenskapen ${ missingProperty }.",o="${ dataPath } får inte vara kortare än ${ limit } tecken.",d="${ dataPath } får inte ha färre än ${ limit } objekt.",m="${ dataPath } får inte ha fler än ${ limit } objekt.",g="${ dataPath } måste ha minst ett tecken som inte är ett blanksteg.",u="${ dataPath } får inte innehålla ${ additionalProperty }.",p="${ dataPath } måste vara lika med ett av dessa tillåtna värden: ${ allowedValues }.",h="${ dataPath } måste matcha schemat för någon av dessa: ${ schemaOptions }.",v="Punktdiagram med proportionerliga symboler stöds inte. Standardstorleken för symboler har tillämpats.",$="Det gick inte att läsa indata.",f="Förväntad serietyp vid index ${ seriesIndex } är '${ expectedType }' men '${ receivedType }' togs emot istället",y="eller";var k={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:n,duplicateSeriesID:i,nonNumericAggregation:s,requiredProperty:l,minLength:o,minItems:d,maxItems:m,whiteSpacePattern:g,additionalProperty:u,enumValues:p,anyOfValues:h,bubbleChartValidateMsg:v,queryError:$,invalidSeriesType:f,or:y};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = v;
exports['default'] = k;
exports.defaultError = a;
exports.duplicateSeriesID = i;
exports.enumValues = p;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = r;
exports.maxItems = m;
exports.minItems = d;
exports.minLength = o;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = s;
exports.or = y;
exports.queryError = $;
exports.requiredProperty = l;
exports.whiteSpacePattern = g;
