// @arcgis/charts-components/dist/cjs/messages.he-70d41af0-2741f049-fdde32a3.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="אירעה שגיאה במהלך טעינת התרשים.",e="לא ניתן לבצע טרנספורמציה לוגריתמית על ערכים שליליים או על אפס.",t="לא ניתן לבצע טרנספורמציה של שורש ריבועי על ערכים שליליים.",r="אירעה שגיאה במהלך טעינת השכבה. URL = ${ url }. מזהה פריט פורטל =${ portalItemId }.",i="השדות [${ fieldNames }] שהוגדרו ב-config לא נמצאו בשכבה.",d="${ dataPath } חייב/ת להיות ייחודי/ת. לסדרה בשם ${ seriesName } יש מזהה (${ seriesID }) שבו משתמשת כבר סדרה אחרת.",o="ל-${ dataPath } אסור לבצע צבירה שאינה ספירה על שדה לא מספרי.",n="ב-${ dataPath } חסר מאפיין בשם ${ missingProperty }.",l="${ dataPath } לא יכול להיות קצר יות מ-${ limit } תווים.",$="ל-${ dataPath } לא יכולים להיות פחות מ-${ limit } פריטים.",s="ל-${ dataPath } לא יכולים להיות יותר מ-${ limit } פריטים.",m="ל-${ dataPath } חייב להיות לפחות תו אחד שאינו רווח בלתי נראה.",u="ל-${ dataPath } אסור לכלול ${ additionalProperty }.",P="${ dataPath } חייב להיות שווה לאחד מהערכים המותרים הבאים: ${ allowedValues }.",h="${ dataPath } חייב להתאים לסכמה של אחד מאלה: ${ schemaOptions }.",p="תרשימי פיזור עם סמלים פרופורציונליים לא נתמכים. הוחל גודל סמל שמוגדר כברירת מחדל.",y="קריאת נתוני הקלט נכשלה.",g="סוג הסדרה הצפוי במדד ${ seriesIndex } הוא '${ expectedType }', אבל במקום זאת התקבל '${ receivedType }'.",I="או";var c={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:d,nonNumericAggregation:o,requiredProperty:n,minLength:l,minItems:$,maxItems:s,whiteSpacePattern:m,additionalProperty:u,enumValues:P,anyOfValues:h,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:g,or:"או"};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = p;
exports['default'] = c;
exports.defaultError = a;
exports.duplicateSeriesID = d;
exports.enumValues = P;
exports.invalidSeriesType = g;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = s;
exports.minItems = $;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = o;
exports.or = I;
exports.queryError = y;
exports.requiredProperty = n;
exports.whiteSpacePattern = m;
