// @arcgis/charts-components/dist/cjs/messages.sv-18b29f8a-8e0ddce6.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Liggande stapeldiagram",lineChart:"Linjediagram",pieChart:"Cirkeldiagram",histogram:"Histogram",scatterPlot:"Sambandsdiagram",boxPlot:"Låddiagram"},e={mean:"Medelvärde",median:"Median",sum:"Summa",discretePercentile:"Diskret percentil",minimum:"Minimum",maximum:"Maximum",variance:"Varians",count:"Antal",aggregation:"${ statistics } av ${ fieldName }",noAggregation:"Ingen aggregering"},i="Diagram",t="Antalsserie",n="X-axel",r="Ej tillgänglig",g={start:"Snappa till den första datapunkten",end:"Snappa till den sista datapunkten"};var m={chartType:a,statistics:e,defaultTitle:i,countSeries:t,xAxis:n,notAvailable:r,timeAggregationTypes:g};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = m;
exports.defaultTitle = i;
exports.notAvailable = r;
exports.statistics = e;
exports.timeAggregationTypes = g;
exports.xAxis = n;
