// @arcgis/charts-components/dist/cjs/messages.pt-PT-67e158bf-3535c22d.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Ocorreu um erro ao carregar o gráfico.",a="Não é possível aplicar uma transformação de registo a valores negativos ou de zero.",o="Não é possível aplicar uma transformação de raiz quadrada a valores negativos.",r="Ocorreu um erro ao carregar a camada. URL = ${ url }. ID do item do portal = ${ portalItemId }.",t="Os campos [${ fieldNames }] definidos na configuração não foram encontrados na camada.",s="${ dataPath } tem de ser único. A série com o nome ${ seriesName } tem um ID (${ seriesID }) que já está a ser usado por outra série.",d="${ dataPath } não pode realizar agregação sem contagem num campo não numérico.",i="${ dataPath } tem uma propriedade com o nome ${ missingProperty } em falta.",n="${ dataPath } não pode ser inferior a ${ limit } caracteres.",m="${ dataPath } não pode conter menos do que ${ limit } itens.",u="${ dataPath } não pode conter mais do que ${ limit } itens.",l="${ dataPath } tem de ter, pelo menos, um caráter que não seja um espaço em branco.",p="${ dataPath } não pode conter ${ additionalProperty }.",c="${ dataPath } tem de ser igual a um destes valores permitidos: ${ allowedValues }.",$="${ dataPath } tem de corresponder ao esquema de um dos seguintes: ${ schemaOptions }.",g="Os gráficos de dispersão com símbolos proporcionais não são suportados. Foi aplicado o tamanho de símbolo predefinido.",f="Falha ao ler os dados de entrada.",h="O tipo de série esperado no índice ${ seriesIndex } é '${ expectedType }', mas '${ receivedType }' foi recebido em vez disso",P="ou";var v={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:o,layerLoadFailure:r,layerFieldsNotFound:t,duplicateSeriesID:s,nonNumericAggregation:d,requiredProperty:i,minLength:n,minItems:m,maxItems:u,whiteSpacePattern:l,additionalProperty:p,enumValues:c,anyOfValues:$,bubbleChartValidateMsg:g,queryError:f,invalidSeriesType:h,or:"ou"};

exports.additionalProperty = p;
exports.anyOfValues = $;
exports.bubbleChartValidateMsg = g;
exports['default'] = v;
exports.defaultError = e;
exports.duplicateSeriesID = s;
exports.enumValues = c;
exports.invalidSeriesType = h;
exports.layerFieldsNotFound = t;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = m;
exports.minLength = n;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = o;
exports.nonNumericAggregation = d;
exports.or = P;
exports.queryError = f;
exports.requiredProperty = i;
exports.whiteSpacePattern = l;
