// @arcgis/charts-components/dist/cjs/messages.da-90739473-58c40af2-28af0c0d.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Søjlediagram",lineChart:"Linjediagram",pieChart:"Cirkeldiagram",histogram:"Histogram",scatterPlot:"Punktdiagram",boxPlot:"Kassediagram"},t={mean:"Middel",median:"Median",sum:"Sum",discretePercentile:"Særskilt percentil",minimum:"Minimum",maximum:"Maksimum",variance:"Varians",count:"Tælling",aggregation:"${ statistics } for ${ fieldName }",noAggregation:"Ingen aggregering"},e="Diagram",i="Tællerække",s="X-akse",r="Ikke tilgængelig",g={start:"Fastgør til det første datapunkt",end:"Fastgør til det sidste datapunkt"};var n={chartType:a,statistics:t,defaultTitle:e,countSeries:i,xAxis:s,notAvailable:r,timeAggregationTypes:g};

exports.chartType = a;
exports.countSeries = i;
exports['default'] = n;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = t;
exports.timeAggregationTypes = g;
exports.xAxis = s;
