// @arcgis/charts-components/dist/cjs/messages.hu-f801472e-56cfa9f7.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Oszlop-/sávdiagram",lineChart:"Vonaldiagram",pieChart:"Kördiagram",histogram:"Hisztogram",scatterPlot:"Szórásdiagram",boxPlot:"Boxdiagram"},t={mean:"Átlag",median:"Medián",sum:"Összeg",discretePercentile:"Diszkrét percentilis",minimum:"Minimum",maximum:"Maximum",variance:"Variancia",count:"Darabszám",aggregation:"${ statistics }/${ fieldName }",noAggregation:"Nincs összevonás"},i="Diagram",e="Számsorozat",s="X tengely",r="Nem érhető el",m={start:"Csatolás az első adatponthoz",end:"Csatolás az utolsó adatponthoz"};var o={chartType:a,statistics:t,defaultTitle:i,countSeries:e,xAxis:s,notAvailable:r,timeAggregationTypes:m};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = o;
exports.defaultTitle = i;
exports.notAvailable = r;
exports.statistics = t;
exports.timeAggregationTypes = m;
exports.xAxis = s;
