// @arcgis/charts-components/dist/cjs/messages.nl-bf7731db-19e7cc57-fa8b698b.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Diagram",barChart:"Staafdiagram",lineChart:"Lijndiagram",pieChart:"Cirkeldiagram",histogram:"Histogram",scatterPlot:"Spreidingsdiagram",boxPlot:"Boxplot"},a={mean:"Gemiddelde",median:"Mediaan",sum:"Som",discretePercentile:"Discreet percentiel",minimum:"Minimum",maximum:"Maximum",variance:"Variantie",count:"Aantal",aggregation:"${ statistics } van ${ fieldName }",noAggregation:"Geen aggregatie"},t="Diagram",i="Tellingserie",s="X-as",r="Niet beschikbaar",n={start:"Koppelen met het eerste gegevenspunt",end:"Koppelen met het laatste gegevenspunt"};var g={chartType:e,statistics:a,defaultTitle:t,countSeries:i,xAxis:s,notAvailable:r,timeAggregationTypes:n};

exports.chartType = e;
exports.countSeries = i;
exports['default'] = g;
exports.defaultTitle = t;
exports.notAvailable = r;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = s;
