// @arcgis/charts-components/dist/cjs/messages.id-e0a138dd-47e29255.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Diagram batang",lineChart:"Diagram garis",pieChart:"Diagram pai",histogram:"Histogram",scatterPlot:"Scatter plot",boxPlot:"Plot kotak"},i={mean:"Rata-rata",median:"Median",sum:"Sum",discretePercentile:"Persentil diskrit",minimum:"Minimal",maximum:"Maksimal",variance:"Varian",count:"Count",aggregation:"${ statistics } dari ${ fieldName }",noAggregation:"Tidak ada agregasi"},t="Diagram",e="Jumlah Rangkaian",r="Sumbu X",s="Tidak Tersedia",n={start:"Posisikan ke titik data pertama",end:"Posisikan ke titik data terakhir"};var m={chartType:a,statistics:i,defaultTitle:t,countSeries:e,xAxis:r,notAvailable:s,timeAggregationTypes:n};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = m;
exports.defaultTitle = t;
exports.notAvailable = s;
exports.statistics = i;
exports.timeAggregationTypes = n;
exports.xAxis = r;
