// @arcgis/charts-components/dist/cjs/messages.id-e1e1d374-103bf642-91feab1c.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Terjadi kesalahan saat memuat grafik.",e="Tidak dapat menerapkan transformasi log ke nilai negatif atau nol.",i="Tidak dapat menerapkan transformasi akar pangkat dua ke nilai negatif.",t="Terjadi kesalahan saat memuat layer. URL = ${ url }. ID item portal = ${ portalItemId }.",r="Kolom [${ fieldNames }] yang ditentukan dalam konfigurasi tidak ditemukan di layer.",n="${ dataPath } harus unik. Rangkaian yang bernama ${ seriesName } memiliki ID (${ seriesID }) yang sudah digunakan oleh rangkaian lain.",s="${ dataPath } tidak dapat melakukan agregasi non-hitungan di kolom non-numerik.",d="${ dataPath } tidak memiliki properti bernama ${ missingProperty }.",l="${ dataPath } tidak boleh lebih pendek dari ${ limit } karakter.",o="${ dataPath } tidak boleh memiliki kurang dari ${ limit } item.",m="${ dataPath } tidak boleh memiliki lebih dari ${ limit } item.",u="${ dataPath } harus memiliki setidaknya satu karakter non-spasi.",k="${ dataPath } tidak boleh memiliki ${ additionalProperty }.",g="${ dataPath } harus sama dengan salah satu nilai yang diizinkan ini: ${ allowedValues }.",h="${ dataPath } harus cocok dengan skema salah satu dari ini: ${ schemaOptions }.",p="Scatter plot dengan simbol proporsional tidak didukung. Ukuran simbol default sudah diterapkan.",y="Gagal membaca input data.",$="Jenis seri yang diharapkan pada indeks ${ seriesIndex } adalah '${ expectedType }' tetapi '${ receivedType }' diterima sebagai gantinya",b="atau";var P={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:t,layerFieldsNotFound:r,duplicateSeriesID:n,nonNumericAggregation:s,requiredProperty:d,minLength:l,minItems:o,maxItems:m,whiteSpacePattern:u,additionalProperty:k,enumValues:g,anyOfValues:h,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:$,or:b};

exports.additionalProperty = k;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = p;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = n;
exports.enumValues = g;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = o;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = s;
exports.or = b;
exports.queryError = y;
exports.requiredProperty = d;
exports.whiteSpacePattern = u;
