// @arcgis/charts-components/dist/cjs/messages.cs-71896a8d-cfe54e1a-f84bfee8.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Graf",barChart:"Pruhový graf",lineChart:"Spojnicový graf",pieChart:"Výsečový graf",histogram:"Histogram",scatterPlot:"Bodový graf",boxPlot:"Krabicový graf"},t={mean:"Průměr",median:"Medián",sum:"Suma",discretePercentile:"Diskrétní percentil",minimum:"Minimum",maximum:"Maximum",variance:"Rozptyl",count:"Počet",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Žádná agregace"},i="Graf",e="Počet sérií",r="Osa X",s="Není k dispozici",o={start:"Přichytit k prvnímu datovému bodu",end:"Přichytit k poslednímu datovému bodu"};var n={chartType:a,statistics:t,defaultTitle:i,countSeries:e,xAxis:r,notAvailable:s,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = s;
exports.statistics = t;
exports.timeAggregationTypes = o;
exports.xAxis = r;
