// @arcgis/charts-components/dist/cjs/messages.nl-417c9242-82f1f562-9ad48549.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Er is een fout opgetreden bij het laden van de grafiek.",a="Kan logtransformatie niet toepassen op negatieve of nulwaarden.",n="Kan transformatie vierkantswortel niet toepassen op negatieve waarden.",t="Er is een fout opgetreden bij het laden van de laag. URL = ${ url }. Portal item ID = ${ portalItemId }.",r="De velden [${ fieldNames }] die in de config zijn gedefinieerd, zijn niet aangetroffen in de laag.",i="${ dataPath } moet uniek zijn. De serie genaamd ${ seriesName } heeft een ID (${ seriesID }) die al wordt gebruikt door een andere serie.",o="${ dataPath } kan geen niet-telbare aggregatie uitvoeren op een niet-numeriek veld.",s="${ dataPath } mist een eigenschap met de naam ${ missingProperty }.",d="${ dataPath } mag niet korter zijn dan ${ limit } tekens.",l="${ dataPath } mag niet minder dan ${ limit } items hebben.",m="${ dataPath } mag niet meer dan ${ limit } items hebben.",g="${ dataPath } moet minstens één non-whitespace teken hebben.",u="${ dataPath } mag geen ${ additionalProperty } hebben.",p="${ dataPath } moet gelijk zijn aan een van deze toegelaten waarden: ${ allowedValues }.",h="${ dataPath } moet overeenkomen met het schema van een van deze: ${ schemaOptions }.",v="Spreidingsdiagrammen met proportionele symbolen worden niet ondersteund. Standaardsymboolgrootte is toegepast.",$="Kan invoergegevens niet lezen.",f="Het verwachte serietype bij index ${ seriesIndex } is '${ expectedType }' maar '${ receivedType }' is in plaats daarvan ontvangen",b="of";var y={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:n,layerLoadFailure:t,layerFieldsNotFound:r,duplicateSeriesID:i,nonNumericAggregation:o,requiredProperty:s,minLength:d,minItems:l,maxItems:m,whiteSpacePattern:g,additionalProperty:u,enumValues:p,anyOfValues:h,bubbleChartValidateMsg:v,queryError:$,invalidSeriesType:f,or:"of"};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = v;
exports['default'] = y;
exports.defaultError = e;
exports.duplicateSeriesID = i;
exports.enumValues = p;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = l;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = n;
exports.nonNumericAggregation = o;
exports.or = b;
exports.queryError = $;
exports.requiredProperty = s;
exports.whiteSpacePattern = g;
