// @arcgis/charts-components/dist/cjs/messages.de-e2ce6e5c-5a57de61-ab5a2a29.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Diagramm",barChart:"Balkendiagramm",lineChart:"Liniendiagramm",pieChart:"Kreisdiagramm",histogram:"Histogramm",scatterPlot:"Scatterplot",boxPlot:"Boxplot"},a={mean:"Mittelwert",median:"Medianwert",sum:"Summe",discretePercentile:"Diskontinuierliches Perzentil",minimum:"Minimum",maximum:"Maximum",variance:"Varianz",count:"Anzahl",aggregation:"${ statistics } von ${ fieldName }",noAggregation:"Keine Aggregation"},t="Diagramm",i="Mengenserie",n="X-Achse",r="Nicht verfügbar",s={start:"Am ersten Datenpunkt fangen",end:"Am letzten Datenpunkt fangen"};var m={chartType:e,statistics:a,defaultTitle:t,countSeries:i,xAxis:n,notAvailable:r,timeAggregationTypes:s};

exports.chartType = e;
exports.countSeries = i;
exports['default'] = m;
exports.defaultTitle = t;
exports.notAvailable = r;
exports.statistics = a;
exports.timeAggregationTypes = s;
exports.xAxis = n;
