// @arcgis/charts-components/dist/cjs/messages.de-8ac83862-b326af6c-30acf87c.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Beim Laden des Diagramms ist ein Fehler aufgetreten.",n="Eine logarithmische Transformation kann nicht auf negative oder Nullwerte angewendet werden.",t="Eine Quadratwurzeltransformation kann nicht auf negative Werte angewendet werden.",a="Beim Laden des Layers ist ein Fehler aufgetreten. URL = ${ url } Portal-Element-ID = ${ portalItemId }.",i="Die in der Konfiguration definierten Felder [${ fieldNames }] wurden im Layer nicht gefunden.",r="${ dataPath } muss eindeutig sein. Die Reihe mit dem Namen ${ seriesName } hat eine ID (${ seriesID }), die schon von einer anderen Reihe verwendet wird.",d="${ dataPath } kann keine Aggregation, die nicht anzahlbasiert ist, für nicht numerische Felder durchführen.",s="${ dataPath } fehlt eine Eigenschaft mit dem Namen ${ missingProperty }.",l="${ dataPath } darf nicht kürzer als ${ limit } Zeichen sein.",m="${ dataPath } darf nicht weniger als ${ limit } Elemente enthalten.",h="${ dataPath } darf nicht mehr als ${ limit } Elemente enthalten.",o="${ dataPath } muss mindestens ein Zeichen enthalten, bei dem es sich nicht um ein Leerstellenzeichen handelt.",u="${ dataPath } darf nicht ${ additionalProperty } aufweisen.",g="${ dataPath } muss einem der folgenden zulässigen Werte entsprechen: ${ allowedValues }.",c="${ dataPath } muss mit einem der folgenden Schemas übereinstimmen: ${ schemaOptions }.",f="Scatterplots mit proportionalen Symbolen werden nicht unterstützt. Die standardmäßige Symbolgröße wurde angewendet.",$="Fehler beim Lesen der Eingabedaten.",p='Der erwartete Serientyp bei Index ${ seriesIndex } lautet "${ expectedType }", stattdessen wurde jedoch "${ receivedType }" empfangen.',w="oder";var y={defaultError:e,negativeValueInDataForLogTransformation:n,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:a,layerFieldsNotFound:i,duplicateSeriesID:r,nonNumericAggregation:d,requiredProperty:s,minLength:l,minItems:m,maxItems:h,whiteSpacePattern:o,additionalProperty:u,enumValues:g,anyOfValues:c,bubbleChartValidateMsg:f,queryError:$,invalidSeriesType:p,or:w};

exports.additionalProperty = u;
exports.anyOfValues = c;
exports.bubbleChartValidateMsg = f;
exports['default'] = y;
exports.defaultError = e;
exports.duplicateSeriesID = r;
exports.enumValues = g;
exports.invalidSeriesType = p;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = a;
exports.maxItems = h;
exports.minItems = m;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = n;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = d;
exports.or = w;
exports.queryError = $;
exports.requiredProperty = s;
exports.whiteSpacePattern = o;
