// @arcgis/charts-components/dist/cjs/messages.el-d39d2cdb-633ed8c5-2c52019f-fa02889a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Παρουσιάστηκε κάποιο σφάλμα κατά τη φόρτωση του γραφήματος.",e="Δεν είναι δυνατή η εφαρμογή του μετασχηματισμού αρχείου καταγραφής σε αρνητικές ή μηδενικές τιμές.",t="Δεν είναι δυνατή η εφαρμογή του μετασχηματισμού τετραγωνικής ρίζας σε αρνητικές τιμές.",r="Παρουσιάστηκε κάποιο σφάλμα κατά τη φόρτωση του θεματικού επιπέδου. Διεύθυνση URL = ${ url }. Αναγνωριστικό αντικειμένου του portal = ${ portalItemId }.",i="Τα πεδία [${ fieldNames }] που ορίστηκαν στη διαμόρφωση δεν βρέθηκαν στο θεματικό επίπεδο.",s="Το ${ dataPath } πρέπει να είναι μοναδικό. Η σειρά με το όνομα ${ seriesName } έχει αναγνωριστικό (${ seriesID }) που ήδη χρησιμοποιείται από κάποια άλλη σειρά.",n="Το ${ dataPath } δεν μπορεί να εκτελέσει συνάθροιση χωρίς μέτρηση σε μη αριθμητικό πεδίο.",o="Από το ${ dataPath } λείπει μια ιδιότητα με το όνομα ${ missingProperty }.",l="Το ${ dataPath } δεν πρέπει να έχει λιγότερους από ${ limit } χαρακτήρες.",d="Το ${ dataPath } δεν πρέπει να έχει λιγότερα από ${ limit } στοιχεία.",u="Το ${ dataPath } δεν πρέπει να έχει περισσότερα από ${ limit } στοιχεία.",m="Το ${ dataPath } πρέπει να έχει τουλάχιστον έναν χαρακτήρα που να μην είναι κενό διάστημα.",$="Το ${ dataPath } δεν πρέπει να έχει ${ additionalProperty }.",p="Το ${ dataPath } πρέπει να ισούται με μία από τις εξής επιτρεπόμενες τιμές: ${ allowedValues }.",y="Το ${ dataPath } πρέπει να συμφωνεί με τη δομή ενός από τα εξής: ${ schemaOptions }.",P="Δεν υποστηρίζονται γραφήματα διασποράς με αναλογικά σύμβολα. Έχει εφαρμοστεί το προεπιλεγμένο μέγεθος συμβόλου.",g="Αποτυχία ανάγνωσης δεδομένων εισόδου",h="Ο αναμενόμενος τύπος σειράς στο ευρετήριο ${ seriesIndex } είναι '${ expectedType }', αλλά λήφθηκε '${ receivedType }' αντί γι' αυτό",I="ή";var V={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:l,minItems:d,maxItems:u,whiteSpacePattern:m,additionalProperty:$,enumValues:p,anyOfValues:y,bubbleChartValidateMsg:P,queryError:g,invalidSeriesType:h,or:"ή"};

exports.additionalProperty = $;
exports.anyOfValues = y;
exports.bubbleChartValidateMsg = P;
exports['default'] = V;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = p;
exports.invalidSeriesType = h;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = n;
exports.or = I;
exports.queryError = g;
exports.requiredProperty = o;
exports.whiteSpacePattern = m;
