// @arcgis/charts-components/dist/cjs/messages.en-US-d444015e-7fcc1245-b262f73f.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="An error occurred while loading the chart.",a="Cannot apply a log transformation to negative or zero values.",t="Cannot apply a square root transformation to negative values.",r="An error occurred while loading the layer. URL = ${ url }. Portal item ID = ${ portalItemId }.",o="The fields [${ fieldNames }] defined in the config were not found in the layer.",n="${ dataPath } must be unique. The series named ${ seriesName } has an ID (${ seriesID }) that is already used by another series.",i="${ dataPath } cannot perform non-count aggregation on non-numeric field.",s="${ dataPath } is missing a property named ${ missingProperty }.",l="${ dataPath } must not be shorter than ${ limit } characters.",d="${ dataPath } must not have fewer than ${ limit } items.",h="${ dataPath } must not have more than ${ limit } items.",u="${ dataPath } must have at least one non-whitespace character.",m="${ dataPath } must not have ${ additionalProperty }.",p="${ dataPath } must be equal to one of these allowed values: ${ allowedValues }.",c="${ dataPath } must match the schema of one of these: ${ schemaOptions }.",$="Scatterplots with proportional symbols are not supported. Default symbol size has been applied.",y="Failure to read input data.",f="The expected series type at index ${ seriesIndex } is '${ expectedType }' but '${ receivedType }' was received instead",g="or";var P={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:o,duplicateSeriesID:n,nonNumericAggregation:i,requiredProperty:s,minLength:l,minItems:d,maxItems:h,whiteSpacePattern:u,additionalProperty:m,enumValues:p,anyOfValues:c,bubbleChartValidateMsg:$,queryError:y,invalidSeriesType:f,or:"or"};

exports.additionalProperty = m;
exports.anyOfValues = c;
exports.bubbleChartValidateMsg = $;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = p;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = o;
exports.layerLoadFailure = r;
exports.maxItems = h;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = i;
exports.or = g;
exports.queryError = y;
exports.requiredProperty = s;
exports.whiteSpacePattern = u;
