// @arcgis/charts-components/dist/cjs/messages.no-950d514a-96417035.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"Diagram",barChart:"Stolpediagram",lineChart:"Linjediagram",pieChart:"Sektordiagram",histogram:"Histogram",scatterPlot:"Punktdiagram",boxPlot:"Boksdiagram"},a={mean:"Gjennomsnitt",median:"Median",sum:"Sum",discretePercentile:"Diskret persentil",minimum:"Minimum",maximum:"Maksimum",variance:"Varians",count:"Antall",aggregation:"${ statistics } av ${ fieldName }",noAggregation:"Ingen sammenslåing"},e="Diagram",i="Tallserie",n="X-akse",r="Ikke tilgjengelig",s={start:"Fest til det første datapunktet",end:"Fest til det siste datapunktet"};var m={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:n,notAvailable:r,timeAggregationTypes:s};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = m;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = a;
exports.timeAggregationTypes = s;
exports.xAxis = n;
