// @arcgis/charts-components/dist/cjs/messages.cs-71896a8d-b4113870.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"Graf",barChart:"Pruhový graf",lineChart:"Spojnicový graf",pieChart:"Výsečový graf",histogram:"Histogram",scatterPlot:"Bodový graf",boxPlot:"Krabicový graf"},a={mean:"Průměr",median:"Medián",sum:"Suma",discretePercentile:"Diskrétní percentil",minimum:"Minimum",maximum:"Maximum",variance:"Rozptyl",count:"Počet",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Žádná agregace"},i="Graf",e="Počet sérií",r="Osa X",o="Není k dispozici",n={start:"Přichytit k prvnímu datovému bodu",end:"Přichytit k poslednímu datovému bodu"};var m={chartType:t,statistics:a,defaultTitle:i,countSeries:e,xAxis:r,notAvailable:o,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = e;
exports['default'] = m;
exports.defaultTitle = i;
exports.notAvailable = o;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = r;
