// @arcgis/charts-components/dist/cjs/messages.bg-32846a04-b7b7b291.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Възникна грешка при зареждането на диаграмата.",e="Не може да се приложи логаритмична трансформация към отрицателни или нулеви стойности.",t="Не може да се приложи трансформация с квадратен корен към отрицателни стойности.",r="Възникна грешка при зареждането на слоя. URL = ${ url }. ИД на елемент на портал = ${ portalItemId }.",i="Полетата [${ fieldNames }], които са дефинирани в конфигурацията, не са намерени в слоя.",d="${ dataPath } трябва да бъде унукално. Серията с името ${ seriesName } има ИД (${ seriesID }), който вече се използва от друга серия.",o="${ dataPath } не може да се извършва обобщение, различно от обобщение на брой, за нецифрово поле.",l="${ dataPath } няма име на свойство ${ missingProperty }.",n="${ dataPath } не трябва да има по-малко от ${ limit } символи.",$="${ dataPath } не трябва да има по-малко от ${ limit } елементи.",s="${ dataPath } не трябва да има повече от ${ limit } елементи.",m="${ dataPath } трябва да има поне един символ, който не е интервал.",u="${ dataPath } не трябва да бъде ${ additionalProperty }.",P="${ dataPath } трябва да бъде равна на една от тези позволени стойности: ${ allowedValues }.",h="${ dataPath }трябва да съответства на схемата на една от тези: ${ schemaOptions }.",p="Не се поддържат диаграми на разсейване с пропорционални символи. Използван е размер на символа по подразбиране.",y="Входните данни не могат да се прочетат.",g="Очакваният тип серия в индекс${ seriesIndex } е „${ expectedType }“, но „${ receivedType }“ е получен вместо това",I="или";var c={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:d,nonNumericAggregation:o,requiredProperty:l,minLength:n,minItems:$,maxItems:s,whiteSpacePattern:m,additionalProperty:u,enumValues:P,anyOfValues:h,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:g,or:"или"};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = p;
exports['default'] = c;
exports.defaultError = a;
exports.duplicateSeriesID = d;
exports.enumValues = P;
exports.invalidSeriesType = g;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = s;
exports.minItems = $;
exports.minLength = n;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = o;
exports.or = I;
exports.queryError = y;
exports.requiredProperty = l;
exports.whiteSpacePattern = m;
