// @arcgis/charts-components/dist/cjs/messages.vi-087f5bf7-20bdf5b5-9baa765d.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const i={chart:"Biểu đồ",barChart:"Biểu đồ cột",lineChart:"Biểu đồ dạng đường",pieChart:"Biểu đồ hình tròn",histogram:"Biểu đồ tần suất",scatterPlot:"Biểu đồ phân tán",boxPlot:"Biểu đồ hộp"},t={mean:"Trung bình",median:"Bình quân",sum:"Tổng",discretePercentile:"Phân vị rời rạc",minimum:"Tối thiểu",maximum:"Tối đa",variance:"Phương sai",count:"Số lượng",aggregation:"${ statistics } / ${ fieldName }",noAggregation:"Không tổng hợp"},a="Biểu đồ",n="Đếm Số chuỗi",e="Trục X",s="Không Khả dụng",u={start:"Chuyển sang điểm dữ liệu đầu tiên",end:"Chuyển sang điểm dữ liệu cuối cùng"};var g={chartType:i,statistics:t,defaultTitle:a,countSeries:n,xAxis:e,notAvailable:s,timeAggregationTypes:u};

exports.chartType = i;
exports.countSeries = n;
exports['default'] = g;
exports.defaultTitle = a;
exports.notAvailable = s;
exports.statistics = t;
exports.timeAggregationTypes = u;
exports.xAxis = e;
