// @arcgis/charts-components/dist/cjs/messages.hu-5bc9a465-774671cb.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Hiba történt a diagram betöltése közben.",a="A log transzformáció nem alkalmazható negatív vagy nulla értékre.",t="A négyzetgyök transzformáció nem alkalmazható negatív értékre.",l="Hiba történt a réteg betöltés közben. URL = ${ url }. Portál elemazonosítója = ${ portalItemId }.",r="A konfigurációban definiált [${ fieldNames }] mezők nem találhatók a rétegben.",n="Egyedinek kell lennie: ${ dataPath } A(z) ${ seriesName } nevű sorozat azonosítóját (${ seriesID }) egy másik sorozat már használja.",i="A(z) ${ dataPath } nem hajthat végre nem megszámlálható összevonást nem numerikus mezőn.",m="A(z) ${ dataPath } esetében hiányzik a(z) ${ missingProperty } nevű tulajdonság.",s="A(z) ${ dataPath } hossza legalább ${ limit } karakter kell legyen.",k="A(z) ${ dataPath } elemeinek száma legalább ${ limit } kell legyen.",o="A(z) ${ dataPath } elemeinek száma legfeljebb ${ limit } lehet.",z="A következőnek legalább egy nem-whitespace karaktert kell tartalmaznia: ${ dataPath }",d="${ dataPath } nem rendelkezhet ezzel: ${ additionalProperty }.",g="${ dataPath } értékének egyeznie kell a következő megengedett értékek egyikével: ${ allowedValues }.",h="A(z) ${ dataPath } értékének egyeznie kell a következők egyikének sémájával: ${ schemaOptions }.",y="A méretarányos szimbólummal rendelkező szórásdiagramok nem támogatottak. Az alapértelmezett szimbólum méret lesz alkalmazva.",$="Nem sikerült beolvasni a bemeneti adatokat.",b="A várt sorozattípus a(z) ${ seriesIndex } indexnél '${ expectedType }', de helyette '${ receivedType }' érkezett.",u="vagy";var v={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:l,layerFieldsNotFound:r,duplicateSeriesID:n,nonNumericAggregation:i,requiredProperty:m,minLength:s,minItems:k,maxItems:o,whiteSpacePattern:z,additionalProperty:d,enumValues:g,anyOfValues:h,bubbleChartValidateMsg:y,queryError:$,invalidSeriesType:b,or:u};

exports.additionalProperty = d;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = y;
exports['default'] = v;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = g;
exports.invalidSeriesType = b;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = l;
exports.maxItems = o;
exports.minItems = k;
exports.minLength = s;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = i;
exports.or = u;
exports.queryError = $;
exports.requiredProperty = m;
exports.whiteSpacePattern = z;
