// @arcgis/charts-components/dist/cjs/messages.da-12df1d76-e0cbd6ce-6bf23637.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Der opstod en fejl under indlæsning af diagrammet.",a="Kan ikke anvende log-transformation på negative værdier eller nulværdier.",r="Kan ikke anvende kvadratrod-transformation på negative værdier.",t="Der opstod en fejl under indlæsning af laget. URL = ${ url }. Portalelement-id = ${ portalItemId }.",n="De felter [${ fieldNames }], der er defineret i konfigurationen, blev ikke fundet i laget.",i="${ dataPath } skal være entydig. Serien med navnet ${ seriesName } har et id (${ seriesID }), der allerede bruges i en anden serie.",d="${ dataPath } kan ikke udføre ikke-antal-aggregering på ikke-numeriske felter.",l="${ dataPath } mangler en egenskab med navnet ${ missingProperty }.",s="${ dataPath } skal ikke være kortere end ${ limit } tegn.",o="${ dataPath } må ikke have færre end ${ limit } elementer.",m="${ dataPath } må ikke have flere end ${ limit } elementer.",k="${ dataPath } skal have mindst et tegn, der ikke er et mellemrumstegn.",g="${ dataPath } må ikke have ${ additionalProperty }.",u="${ dataPath } skal være lig med en af disse tilladte værdier: ${ allowedValues }.",v="${ dataPath } skal matche skemaet fra en af disse: ${ schemaOptions }.",f="Punktdiagrammer med proportionale symboler understøttes ikke. Der er blevet anvendt standardstørrelse på symboler.",$="Kunne ikke læse inputdata.",p="Den forventede serietype ved indekset ${ seriesIndex } er '${ expectedType }', men '${ receivedType }' blev modtaget i stedet for",h="eller";var P={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:r,layerLoadFailure:t,layerFieldsNotFound:n,duplicateSeriesID:i,nonNumericAggregation:d,requiredProperty:l,minLength:s,minItems:o,maxItems:m,whiteSpacePattern:k,additionalProperty:g,enumValues:u,anyOfValues:v,bubbleChartValidateMsg:f,queryError:$,invalidSeriesType:p,or:h};

exports.additionalProperty = g;
exports.anyOfValues = v;
exports.bubbleChartValidateMsg = f;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = i;
exports.enumValues = u;
exports.invalidSeriesType = p;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = o;
exports.minLength = s;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = r;
exports.nonNumericAggregation = d;
exports.or = h;
exports.queryError = $;
exports.requiredProperty = l;
exports.whiteSpacePattern = k;
