// @arcgis/charts-components/dist/cjs/messages.pl-2dec17dd-b41cd665-f512dc32-45d248bd.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Wystąpił błąd podczas wczytywania diagramu.",e="Nie można zastosować transformacji Logarytm do wartości ujemnych lub równych zero.",i="Nie można zastosować transformacji Pierwiastek kwadratowy do wartości ujemnych.",n="Wystąpił błąd podczas wczytywania warstwy. URL = ${ url }. Identyfikator elementu portalu = ${ portalItemId }.",t="Zdefiniowane w konfiguracji pola [${ fieldNames }] nie zostały znalezione w warstwie.",r="Nazwa ${ dataPath } musi być unikalna. Seria o nazwie ${ seriesName } ma identyfikator (${ seriesID }), który jest już używany przez inną serię.",o="Element ${ dataPath } nie może wykonywać agregacji innej niż zliczenie na polu innym niż liczbowe.",s="W elemencie ${ dataPath } brak właściwości o nazwie ${ missingProperty }.",m="Element ${ dataPath } nie może być krótszy niż następująca liczba znaków: ${ limit }.",l="Element ${ dataPath } nie może mieć mniej elementów niż ${ limit }.",y="Element ${ dataPath } nie może mieć więcej elementów niż ${ limit }.",d="Element ${ dataPath } musi zawierać co najmniej jeden znak inny niż biały znak.",u="Element ${ dataPath } nie może zawierać ${ additionalProperty }.",w="Element ${ dataPath } musi mieć wartość równą jednej z tych dozwolonych wartości: ${ allowedValues }.",c="Element ${ dataPath } musi być zgodny z jednym z tych schematów: ${ schemaOptions }.",z="Diagramy punktowe z symbolami proporcjonalnymi nie są obsługiwane. Zastosowano domyślną wielkość symbolu.",p="Nie powiódł się odczyt danych wejściowych.",g="Oczekiwany typ serii z indeksem ${ seriesIndex } to '${ expectedType }', ale zamiast niego otrzymano '${ receivedType }'",h="lub";var $={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:n,layerFieldsNotFound:t,duplicateSeriesID:r,nonNumericAggregation:o,requiredProperty:s,minLength:m,minItems:l,maxItems:y,whiteSpacePattern:d,additionalProperty:u,enumValues:w,anyOfValues:c,bubbleChartValidateMsg:z,queryError:p,invalidSeriesType:g,or:"lub"};

exports.additionalProperty = u;
exports.anyOfValues = c;
exports.bubbleChartValidateMsg = z;
exports['default'] = $;
exports.defaultError = a;
exports.duplicateSeriesID = r;
exports.enumValues = w;
exports.invalidSeriesType = g;
exports.layerFieldsNotFound = t;
exports.layerLoadFailure = n;
exports.maxItems = y;
exports.minItems = l;
exports.minLength = m;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = o;
exports.or = h;
exports.queryError = p;
exports.requiredProperty = s;
exports.whiteSpacePattern = d;
