// @arcgis/charts-components/dist/cjs/messages.fi-c0c48057-bbece429-1dfeae75.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Kaavion lataamisessa tapahtui virhe.",t="Logaritmimuunnosta ei voi käyttää negatiivisiin arvoihin tai nolla-arvoihin.",e="Neliöjuurimuunnosta ei voi käyttää negatiivisiin arvoihin.",i="Karttatason lataamisessa tapahtui virhe. URL = ${ url }. Portaalikohteen tunnus = ${ portalItemId }.",s="Määrityksessä määritettyjä kenttiä [${ fieldNames }] ei löytynyt karttatasolta.",n="Kohteen ${ dataPath } on oltava yksilöllinen. Sarjan ${ seriesName } tunnus (${ seriesID }) on jo toisen sarjan käytössä.",o="${ dataPath } ei saa suorittaa muuta kuin määrän koostamista muussa kuin numerokentässä.",r="Kohteesta ${ dataPath } puuttuu ominaisuus ${ missingProperty }.",l="${ dataPath } ei saa olla alle ${ limit } merkkiä pitkä.",u="Kohteessa ${ dataPath } ei saa olla alle ${ limit } kohdetta.",m="Kohteessa ${ dataPath } ei saa olla yli ${ limit } kohdetta.",d="Kohteessa ${ dataPath } on oltava vähintään yksi muu kuin välimerkki.",h="Kohteessa ${ dataPath } ei saa olla kohdetta ${ additionalProperty }.",y="Kohteen ${ dataPath } tulee olla yhtä suuri kuin jokin näistä sallituista arvoista: ${ allowedValues }.",k="Kohteen ${ dataPath } on vastattava jotakin näistä skeemoista: ${ schemaOptions }.",v="Pistekaavioita, joissa on suhteutettuja symboleja, ei tueta. Symbolin oletuskokoa on käytetty.",p="Lähtöaineiston lukeminen epäonnistui.",$='Odotettu sarjan tyyppi indeksissä ${ seriesIndex } on "${ expectedType }", mutta sen sijaan vastaanotettiin "${ receivedType }"',g="tai";var P={defaultError:a,negativeValueInDataForLogTransformation:t,negativeValueInDataForSqrtTransformation:e,layerLoadFailure:i,layerFieldsNotFound:s,duplicateSeriesID:n,nonNumericAggregation:o,requiredProperty:r,minLength:l,minItems:u,maxItems:m,whiteSpacePattern:d,additionalProperty:h,enumValues:y,anyOfValues:k,bubbleChartValidateMsg:v,queryError:p,invalidSeriesType:$,or:"tai"};

exports.additionalProperty = h;
exports.anyOfValues = k;
exports.bubbleChartValidateMsg = v;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = n;
exports.enumValues = y;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = s;
exports.layerLoadFailure = i;
exports.maxItems = m;
exports.minItems = u;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = t;
exports.negativeValueInDataForSqrtTransformation = e;
exports.nonNumericAggregation = o;
exports.or = g;
exports.queryError = p;
exports.requiredProperty = r;
exports.whiteSpacePattern = d;
