// @arcgis/charts-components/dist/cjs/messages.tr-7c612534-d6e20594-d9f71974.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Çizge yüklenirken bir hata oluştu.",e="Negatif veya sıfır değerlere işlem kaydı dönüştürme uygulanamaz.",r="Negatif değerlere kare kök dönüştürme uygulanamaz.",i="Katman yüklenirken bir hata oluştu. URL = ${ url }. Portal ögesi kimliği = ${ portalItemId }.",l="Yapılandırmada tanımlanan [${ fieldNames }] alanları katmanda bulunamadı.",t="${ dataPath } benzersiz olmalıdır. ${ seriesName } adlı serinin başka bir seri tarafından zaten kullanılan bir kimliği (${ seriesID }) var.",n="${ dataPath } sayısal olmayan alanda sayım dışı kümeleme gerçekleştiremez.",d="${ dataPath }'de ${ missingProperty } adlı özellik eksik.",s="${ dataPath }, ${ limit } karakterden daha kısa olmamalıdır.",m="${ dataPath }, ${ limit }'den az ögeye sahip olmamalıdır.",o="${ dataPath }, ${ limit }'den fazla ögeye sahip olmamalıdır.",u="${ dataPath }, en az bir adet boşluk olmayan karaktere sahip olmalıdır.",y="${ dataPath }'de ${ additionalProperty } olmamalıdır.",k="${ dataPath }, izin verilen şu değerlerden birine eşit olmalıdır: ${ allowedValues }.",g="${ dataPath }, şunlardan birinin şemasıyla eşleşmelidir: ${ schemaOptions }.",$="Orantısal sembollere sahip dağılım grafikleri desteklenmez. Varsayılan sembol boyutu uygulandı.",h="Girdi verileri okunamadı.",p="${ seriesIndex } dizininde beklenen seri tipi '${ expectedType }' ancak '${ receivedType }' alındı",b="veya";var P={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:r,layerLoadFailure:i,layerFieldsNotFound:l,duplicateSeriesID:t,nonNumericAggregation:n,requiredProperty:d,minLength:s,minItems:m,maxItems:o,whiteSpacePattern:u,additionalProperty:y,enumValues:k,anyOfValues:g,bubbleChartValidateMsg:$,queryError:h,invalidSeriesType:p,or:b};

exports.additionalProperty = y;
exports.anyOfValues = g;
exports.bubbleChartValidateMsg = $;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = t;
exports.enumValues = k;
exports.invalidSeriesType = p;
exports.layerFieldsNotFound = l;
exports.layerLoadFailure = i;
exports.maxItems = o;
exports.minItems = m;
exports.minLength = s;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = r;
exports.nonNumericAggregation = n;
exports.or = b;
exports.queryError = h;
exports.requiredProperty = d;
exports.whiteSpacePattern = u;
