// @arcgis/charts-components/dist/cjs/messages.bs-26dfc274-27769d05-bf12bb3b.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Grafikon",barChart:"Trakasti grafikon",lineChart:"Linijski grafikon",pieChart:"Tortni grafikon",histogram:"Histogram",scatterPlot:"Dijagram raspršivanja",boxPlot:"Dijagram pravokutnika"},i={mean:"Prosjek",median:"Srednja vrijednost",sum:"Zbroj",discretePercentile:"Diskretni postotak",minimum:"Minimum",maximum:"Maksimum",variance:"Varijanca",count:"Zbroj",aggregation:"${ statistics } od ${ fieldName }",noAggregation:"Nema agregacije"},t="Grafikon",e="Broj serija",r="Os x",s="Nije dostupno",o={start:"Prilijepi na prvu podatkovnu točku",end:"Prilijepi na posljednju podatkovnu točku"};var n={chartType:a,statistics:i,defaultTitle:t,countSeries:e,xAxis:r,notAvailable:s,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = e;
exports['default'] = n;
exports.defaultTitle = t;
exports.notAvailable = s;
exports.statistics = i;
exports.timeAggregationTypes = o;
exports.xAxis = r;
