// @arcgis/charts-components/dist/cjs/messages.ro-0cc0459a-be967c67-53eb1fcd.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagramă",barChart:"Diagramă cu bare",lineChart:"Diagramă liniară",pieChart:"Diagramă radială",histogram:"Histogramă",scatterPlot:"Diagramă de dispersie",boxPlot:"Diagramă tip casetă"},i={mean:"Medie",median:"Mediană",sum:"Sumă",discretePercentile:"Percentilă discretă",minimum:"Minim",maximum:"Maxim",variance:"Variaţie",count:"Număr",aggregation:"${ statistics } din ${ fieldName }",noAggregation:"Fără însumare"},e="Diagramă",t="Serie sumă",r="Axa X",s="Indisponibil",m={start:"Fixare la primul punct de date",end:"Fixare la ultimul punct de date"};var n={chartType:a,statistics:i,defaultTitle:e,countSeries:t,xAxis:r,notAvailable:s,timeAggregationTypes:m};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = e;
exports.notAvailable = s;
exports.statistics = i;
exports.timeAggregationTypes = m;
exports.xAxis = r;
