// @arcgis/charts-components/dist/cjs/messages.fr-fb0dd043-2f029c2d-5ed9b657.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Diagramme",barChart:"Diagramme en barres",lineChart:"Diagramme linéaire",pieChart:"Diagramme à secteurs",histogram:"Histogramme",scatterPlot:"Nuage de points",boxPlot:"Boîte à moustaches"},a={mean:"Moyenne",median:"Médiane",sum:"Somme",discretePercentile:"Centile discret",minimum:"Minimum",maximum:"Maximum",variance:"Variance",count:"Total",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Pas d’agrégation"},i="Diagramme",t="Série de nombres",r="Axe X",s="Non disponible",n={start:"Capturer sur le premier point de données",end:"Capturer sur le dernier point de données"};var m={chartType:e,statistics:a,defaultTitle:i,countSeries:t,xAxis:r,notAvailable:s,timeAggregationTypes:n};

exports.chartType = e;
exports.countSeries = t;
exports['default'] = m;
exports.defaultTitle = i;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = r;
