// @arcgis/charts-components/dist/cjs/messages.ja-4264bfb4-0db24230-e8e70439-ebd179a4.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="チャートの読み込み中にエラーが発生しました。",e="対数変換は負またはゼロの値に適用できません。",t="平方根変換は負の値に適用できません。",r="レイヤーの読み込み中にエラーが発生しました。 URL = ${ url } ポータル アイテム ID = ${ portalItemId }",i="構成に定義されたフィールド [${ fieldNames }] がレイヤー内に見つかりませんでした。",s="${ dataPath } は一意である必要があります。 ${ seriesName } という系列の ID (${ seriesID }) は、別の系列ですでに使用されています。",n="${ dataPath } は、数値以外のフィールドで、個数以外の集約を実行することはできません。",o="${ dataPath } に ${ missingProperty } というプロパティがありません。",d="${ dataPath } は ${ limit } 文字以上である必要があります。",l="${ dataPath } のアイテムは ${ limit } 以上にする必要があります。",u="${ dataPath } のアイテムは ${ limit } 以下にする必要があります。",m="${ dataPath } には空白以外の文字が 1 つ以上含まれている必要があります。",$="${ dataPath } が ${ additionalProperty } を持つことはできません。",p="${ dataPath } は、次のいずれかの許可された値と等しい必要があります: ${ allowedValues }",y="${ dataPath } は、次のいずれかのスキーマと一致する必要があります: ${ schemaOptions }",P="比例シンボルを使用した散布図はサポートされていません。 デフォルトのシンボル サイズが適用されています。",g="入力データを読み取れません。",h="インデックス ${ seriesIndex } で必要な系列タイプは ${ expectedType } ですが、代わりに ${ receivedType } が指定されました",I="または";var V={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:d,minItems:l,maxItems:u,whiteSpacePattern:m,additionalProperty:$,enumValues:p,anyOfValues:y,bubbleChartValidateMsg:P,queryError:g,invalidSeriesType:h,or:"または"};

exports.additionalProperty = $;
exports.anyOfValues = y;
exports.bubbleChartValidateMsg = P;
exports['default'] = V;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = p;
exports.invalidSeriesType = h;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = l;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = n;
exports.or = I;
exports.queryError = g;
exports.requiredProperty = o;
exports.whiteSpacePattern = m;
