// @arcgis/charts-components/dist/cjs/messages.lv-90fae237-122c75a6-38f55f37.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Ielādējot diagrammu, radās kļūda.",e="Žurnālfaila transformāciju nevar piemērot negatīvām vai nulles vērtībām.",t="Kvadrātsaknes transformāciju nevar lietot negatīvām vērtībām.",r="Ielādējot slāni, radās kļūda. Vietrādis (URL) = ${ url }. Portāla vienības ID = ${ portalItemId }.",i="Konfigurācijā definētie lauki [${ fieldNames }] slānī nav atrasti.",s="${ dataPath } jābūt unikālam. Sērijai ar nosaukumu ${ seriesName } ir ID (${ seriesID }), ko jau izmanto cita sērija.",n="${ dataPath } nevar veikt apkopošanu laukā, kurā nav skaitlisku vērtību.",o="${ dataPath } trūkst parametra ar nosaukumu ${ missingProperty }.",d="${ dataPath } nedrīkst būt īsāks par ${ limit } rakstzīmēm.",m="${ dataPath } nedrīkst būt mazāk par ${ limit } vienībām.",l="${ dataPath } nedrīkst būt vairāk par ${ limit } vienībām.",u="${ dataPath } jābūt vismaz vienai rakstzīmei, kas nav atstarpe.",v="${ dataPath } nedrīkst būt ${ additionalProperty }.",k="${ dataPath } jābūt vienādam ar kādu no šīm atļautajām vērtībām: ${ allowedValues }.",p="${ dataPath } jāatbilst kādai no šādām shēmām: ${ schemaOptions }.",b="Izkliedes diagrammas ar proporcionāliem simboliem netiek atbalstītas. Ticis piemērots noklusējuma simbola izmērs.",$="Neizdevās nolasīt ievades datus.",g="Indeksa ${ seriesIndex } prognozētais sērijas veids ir “${ expectedType }”, bet tā vietā tika saņemts “${ receivedType }”",I="vai";var P={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:d,minItems:m,maxItems:l,whiteSpacePattern:u,additionalProperty:v,enumValues:k,anyOfValues:p,bubbleChartValidateMsg:b,queryError:$,invalidSeriesType:g,or:"vai"};

exports.additionalProperty = v;
exports.anyOfValues = p;
exports.bubbleChartValidateMsg = b;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = k;
exports.invalidSeriesType = g;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = l;
exports.minItems = m;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = n;
exports.or = I;
exports.queryError = $;
exports.requiredProperty = o;
exports.whiteSpacePattern = u;
