// @arcgis/charts-components/dist/cjs/messages.sl-407c0dc1-5a9ccb44-20bea3a1-43a221e0.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Med nalaganjem grafikona je prišlo do napake.",e="Pretvorbe dnevnika ni mogoče uporabiti za negativne ali ničelne vrednosti.",i="Pretvorbe kvadratnega korena ni mogoče uporabiti za negativne vrednosti.",t="Med nalaganjem sloja je prišlo do napake. URL = ${ url }. ID elementa portala = ${ portalItemId }.",n="Polja [${ fieldNames }] opredeljena v konfiguraciji niso bila najdena v sloju.",r="${ dataPath } mora biti edinstveno. Serija z imenom ${ seriesName } ima ID (${ seriesID }), ki ga že uporablja druga serija.",o="${ dataPath } ne more izvajati združevanja brez števila na neštevilskem polju.",s="${ dataPath } manjka lastnost imenovana ${ missingProperty }.",l="${ dataPath } ne sme biti krajše od ${ limit } znakov.",d="${ dataPath } ne sme imeti manj kot ${ limit } elementov.",m="${ dataPath } ne sme imeti več kot ${ limit } elementov.",p="${ dataPath } mora vsebovati vsaj en znak, ki ni presledek.",u="${ dataPath } ne sme imeti ${ additionalProperty }.",v="${ dataPath } mora biti enak eni od teh dovoljenih vrednosti: ${ allowedValues }.",g="${ dataPath } se mora ujemati s shemo ene od teh: ${ schemaOptions }.",j="Raztreseni grafikoni s proporcionalnimi simboli niso podprti. Uporabljena je bila privzeta velikost simbolov.",$="Branje vhodnih podatkov ni uspelo.",k="Pričakovani tip niza z indeksom ${ seriesIndex } je '${ expectedType }', ampak '${ receivedType }' je bil prejet namesto",h="ali";var b={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:t,layerFieldsNotFound:n,duplicateSeriesID:r,nonNumericAggregation:o,requiredProperty:s,minLength:l,minItems:d,maxItems:m,whiteSpacePattern:p,additionalProperty:u,enumValues:v,anyOfValues:g,bubbleChartValidateMsg:j,queryError:$,invalidSeriesType:k,or:"ali"};

exports.additionalProperty = u;
exports.anyOfValues = g;
exports.bubbleChartValidateMsg = j;
exports['default'] = b;
exports.defaultError = a;
exports.duplicateSeriesID = r;
exports.enumValues = v;
exports.invalidSeriesType = k;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = d;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = o;
exports.or = h;
exports.queryError = $;
exports.requiredProperty = s;
exports.whiteSpacePattern = p;
