// @arcgis/charts-components/dist/cjs/messages.ro-fbedb46a-ed25b115.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="A survenit o eroare în timpul încărcării diagramei.",a="Nu se poate aplica o transformare de jurnal în valori negative sau zero.",i="Nu se poate aplica o transformare a rădăcinii pătrate în valori negative.",t="A survenit o eroare în timpul încărcării stratului tematic. URL = ${ url }. ID-ul elementului de portal = ${ portalItemId }.",r="Câmpurile (${ fieldNames }) definite în configurație nu au fost găsite în stratul tematic.",n="${ dataPath } trebuie să fie unic. Seria denumită ${ seriesName } are un ID (${ seriesID }) utilizat deja de către o altă serie.",u="${ dataPath } nu poate efectua agregarea non-numărare pe un câmp non-numeric.",s="${ dataPath } are lipsă o proprietate denumită ${ missingProperty }.",o="${ dataPath } nu trebuie să fie mai scurt de ${ limit } caractere.",l="${ dataPath } nu trebuie să aibă mai puțin de ${ limit } articole.",d="${ dataPath } nu trebuie să aibă mai mult de ${ limit } articole.",m="${ dataPath } trebuie să aibă cel puțin un caracter non-spațiu gol.",p="${ dataPath } nu trebuie să aibă ${ additionalProperty }.",c="${ dataPath } trebuie să fie egal cu una dintre aceste valori permise: ${ allowedValues }.",$="${ dataPath } trebuie să corespundă schemei uneia dintre acestea: ${ schemaOptions }.",g="Proprietățile diagramei de dispersie cu simbolurile proporționale nu sunt acceptate. A fost aplicată dimensiunea implicită a simbolului.",b="A eșuat citirea datelor de intrare.",f="Tipul așteptat al serie la indicele ${ seriesIndex } este „${ expectedType }”, dar a fost primit „${ receivedType }”",P="sau";var h={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:t,layerFieldsNotFound:r,duplicateSeriesID:n,nonNumericAggregation:u,requiredProperty:s,minLength:o,minItems:l,maxItems:d,whiteSpacePattern:m,additionalProperty:p,enumValues:c,anyOfValues:$,bubbleChartValidateMsg:g,queryError:b,invalidSeriesType:f,or:"sau"};

exports.additionalProperty = p;
exports.anyOfValues = $;
exports.bubbleChartValidateMsg = g;
exports['default'] = h;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = c;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = t;
exports.maxItems = d;
exports.minItems = l;
exports.minLength = o;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = u;
exports.or = P;
exports.queryError = b;
exports.requiredProperty = s;
exports.whiteSpacePattern = m;
