// @arcgis/charts-components/dist/cjs/messages.uk-4083ab16-04101efe.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"Діаграма",barChart:"Гістограма",lineChart:"Лінійна діаграма",pieChart:"Кругова діаграма",histogram:"Гістограма",scatterPlot:"Точкова діаграма",boxPlot:"Стовпчикова діаграма"},a={mean:"Середнє",median:"Медіанне",sum:"Сума",discretePercentile:"Дискретний процентиль",minimum:"Мінімум",maximum:"Максимум",variance:"Дисперсія",count:"Лічильник",aggregation:"${ statistics } з ${ fieldName }",noAggregation:"Немає агрегування"},e="Діаграма",i="Кількість серій",r="Вісь X",s="Недоступно",n={start:"Прив'язка до першої точки даних",end:"Прив'язка до останньої точки даних"};var o={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:r,notAvailable:s,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = r;
