// @arcgis/charts-components/dist/cjs/messages.pt-PT-e9eb0f86-16cb5e4b.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Gráfico",barChart:"Gráfico de barras",lineChart:"Gráfico de linhas",pieChart:"Gráfico circular",histogram:"Histograma",scatterPlot:"Gráfico de dispersão",boxPlot:"Gráfico de caixas"},i={mean:"Média",median:"Mediana",sum:"Soma",discretePercentile:"Percentil discreto",minimum:"Mínimo",maximum:"Máximo",variance:"Variação",count:"Contagem",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sem agregação"},e="Gráfico",o="Séries de Contagem",t="Eixo X",r="Não disponível",s={start:"Ajustar ao primeiro ponto de dados",end:"Ajustar ao último ponto de dados"};var n={chartType:a,statistics:i,defaultTitle:e,countSeries:o,xAxis:t,notAvailable:r,timeAggregationTypes:s};

exports.chartType = a;
exports.countSeries = o;
exports['default'] = n;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = i;
exports.timeAggregationTypes = s;
exports.xAxis = t;
