// @arcgis/charts-components/dist/cjs/messages.bg-32cb6752-cb1579df.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"Диаграма",barChart:"Диаграма във форма на лента",lineChart:"Диаграма във форма на линия",pieChart:"Диаграма в кръгла форма",histogram:"Хистограма",scatterPlot:"Точкова диаграма",boxPlot:"Каре диаграма"},a={mean:"Осреднена стойност",median:"Медиана",sum:"Сума",discretePercentile:"Дискретен перцентил",minimum:"Минимални",maximum:"Максимални",variance:"Отклонение",count:"Брой",aggregation:"${ statistics } от ${ fieldName }",noAggregation:"Няма агрегиране"},e="Диаграма",i="Серия за броене",r="Ос Х",s="Няма налични",n={start:"Притеглете до първата точка от данни",end:"Притеглете до последната точка от данни"};var o={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:r,notAvailable:s,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = r;
