// @arcgis/charts-components/dist/cjs/messages.lt-21cd5fa8-372b323d-2ba99974.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Įkeliant diagramą įvyko klaida.",i="Negalima pritaikyti log. transformacijos neigiamoms arba nulinėms reikšmėms.",e="Negalima pritaikyti kvadratinės šaknies transformacijos neigiamoms reikšmėms.",t="Įkeliant sluoksnį įvyko klaida. URL = ${ url }. Portalo elemento ID = ${ portalItemId }.",r="Sluoksnyje nepavyko rasti konfigūracijoje apibrėžtų laukų [${ fieldNames }].",s="${ dataPath } turi būti unikalus. Serijos, kurios pavadinimas ${ seriesName }, ID (${ seriesID }) jau naudojamas kitai serijai.",n="${ dataPath } negali atikti neskaičiavimo agregavimo ne skaičių lauke.",o="${ dataPath } trūksta savybės pavadinimu ${ missingProperty }.",l="${ dataPath } negali būti trumpesnis negu ${ limit } simb.",u="${ dataPath } negali turėti mažiau negu ${ limit } elementų.",m="${ dataPath } negali turėti daugiau negu ${ limit } elementų.",d="${ dataPath } turi sudaryti bent vienas simbolis, kuris nėra tarpas.",g="${ dataPath } negali sudaryti ${ additionalProperty }.",y="${ dataPath } turi būti lygus vienai iš šių leidžiamų reikšmių: ${ allowedValues }.",k="${ dataPath } turi sutapti su viena iš šių schemų: ${ schemaOptions }.",p="Sklaidos diagramos su proporciniais simboliais nėra palaikomos. Pritaikytas numatytasis simbolių dydis.",$="Nepavyko nuskaityti įvesties duomenų.",v="Numatytas serijos tipas indeksui ${ seriesIndex } yra ${ expectedType }, bet vietoj to gauta ${ receivedType }",P="arba";var b={defaultError:a,negativeValueInDataForLogTransformation:i,negativeValueInDataForSqrtTransformation:e,layerLoadFailure:t,layerFieldsNotFound:r,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:l,minItems:u,maxItems:m,whiteSpacePattern:d,additionalProperty:g,enumValues:y,anyOfValues:k,bubbleChartValidateMsg:p,queryError:$,invalidSeriesType:v,or:P};

exports.additionalProperty = g;
exports.anyOfValues = k;
exports.bubbleChartValidateMsg = p;
exports['default'] = b;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = y;
exports.invalidSeriesType = v;
exports.layerFieldsNotFound = r;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = u;
exports.minLength = l;
exports.negativeValueInDataForLogTransformation = i;
exports.negativeValueInDataForSqrtTransformation = e;
exports.nonNumericAggregation = n;
exports.or = P;
exports.queryError = $;
exports.requiredProperty = o;
exports.whiteSpacePattern = d;
