// @arcgis/charts-components/dist/cjs/messages.fr-b992d801-89c0ef56-be4405ba.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Une erreur est survenue lors du chargement du diagramme.",a="Impossible d'appliquer une transformation logarithmique à des valeurs négatives ou égales à zéro.",t="Impossible d'appliquer une transformation de racine carrée à des valeurs négatives.",r="Une erreur est survenue lors du chargement de la couche. URL = ${ url }. ID de l'élément du portail = ${ portalItemId }.",s="Les champs [${ fieldNames }] définis dans la configuration sont introuvables dans la couche.",n="${ dataPath } doit être unique. La série nommée ${ seriesName } comporte un ID (${ seriesID }) qui est déjà utilisé par une autre série.",i="${ dataPath } ne peut pas réaliser d'agrégation de type autre que Total sur un champ non numérique.",o="Propriété ${ missingProperty } manquante dans ${ dataPath }.",u="${ dataPath } ne doit pas comporter moins de ${ limit } caractères.",d="${ dataPath } ne doit pas comporter moins de ${ limit } éléments.",l="${ dataPath } ne doit pas comporter plus de ${ limit } éléments.",p="${ dataPath } doit contenir au moins un caractère autre qu'un espace.",m="${ dataPath } ne doit pas contenir ${ additionalProperty }.",c="${ dataPath } doit être égal à l'une de ces valeurs autorisées : ${ allowedValues }.",$="${ dataPath } doit correspondre à l'une des structures suivantes : ${ schemaOptions }.",h="Les nuages de points avec des symboles proportionnels ne sont pas pris en charge. La taille de symbole par défaut a été appliquée.",g="Échec de la lecture des données en entrée.",P="Le type de série attendu à l’index ${ seriesIndex } est '${ expectedType }', mais '${ receivedType }' a été reçu à la place",v="ou";var y={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:s,duplicateSeriesID:n,nonNumericAggregation:i,requiredProperty:o,minLength:u,minItems:d,maxItems:l,whiteSpacePattern:p,additionalProperty:m,enumValues:c,anyOfValues:$,bubbleChartValidateMsg:h,queryError:g,invalidSeriesType:P,or:"ou"};

exports.additionalProperty = m;
exports.anyOfValues = $;
exports.bubbleChartValidateMsg = h;
exports['default'] = y;
exports.defaultError = e;
exports.duplicateSeriesID = n;
exports.enumValues = c;
exports.invalidSeriesType = P;
exports.layerFieldsNotFound = s;
exports.layerLoadFailure = r;
exports.maxItems = l;
exports.minItems = d;
exports.minLength = u;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = i;
exports.or = v;
exports.queryError = g;
exports.requiredProperty = o;
exports.whiteSpacePattern = p;
