// @arcgis/charts-components/dist/cjs/messages.zh-TW-53165175-f2d2d8c2.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="載入圖表時出錯。",e="無法將對數轉換套用於負值或零值。",t="無法將平方根轉換套用於負值。",r="載入圖層時出錯。 URL = ${ url }。 入口網站項目 ID = ${ portalItemId }。",i="在圖層中找不到配置中定義的欄位 [${ fieldNames }]。",d="${ dataPath } 必須唯一。 名為 ${ seriesName } 的數列具有其他數列已在使用的 ID (${ seriesID })。",o="${ dataPath } 無法在非數值欄位上執行非計數匯聚。",l="${ dataPath } 缺少名為 ${ missingProperty } 的屬性。",n="${ dataPath } 不應短於 ${ limit } 個字元。",$="${ dataPath } 的項目數不應少於 ${ limit } 個。",s="${ dataPath } 的項目數不應多於 ${ limit } 個。",m="${ dataPath } 至少必須有一個非空白字元。",u="${ dataPath } 不能有 ${ additionalProperty }。",P="${ dataPath } 必須等於下列其中一個允許值: ${ allowedValues }。",h="${ dataPath } 必須符合下列其中一項結構描述: ${ schemaOptions }。",p="不支援包含比例符號的散點圖。 已套用預設符號大小。",y="無法讀取輸入資料。",I="索引 ${ seriesIndex } 上預期的數列類型為 '${ expectedType }'，但改為接收 '${ receivedType }'",g="或";var c={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:d,nonNumericAggregation:o,requiredProperty:l,minLength:n,minItems:$,maxItems:s,whiteSpacePattern:m,additionalProperty:u,enumValues:P,anyOfValues:h,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:I,or:"或"};

exports.additionalProperty = u;
exports.anyOfValues = h;
exports.bubbleChartValidateMsg = p;
exports['default'] = c;
exports.defaultError = a;
exports.duplicateSeriesID = d;
exports.enumValues = P;
exports.invalidSeriesType = I;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = s;
exports.minItems = $;
exports.minLength = n;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = o;
exports.or = g;
exports.queryError = y;
exports.requiredProperty = l;
exports.whiteSpacePattern = m;
