// @arcgis/charts-components/dist/cjs/messages.ca-21b03272-307f9ab0.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e={chart:"Gràfic",barChart:"Gràfic de barres",lineChart:"Gràfic de línies",pieChart:"Gràfic circular",histogram:"Histograma",scatterPlot:"Gràfic de dispersió",boxPlot:"Diagrama de caixa"},a={mean:"Mitjana",median:"Mediana",sum:"Suma",discretePercentile:"Percentil discret",minimum:"Mínim",maximum:"Màxim",variance:"Variància",count:"Recompte",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sense agregació"},i="Gràfic",t="Recompte de sèries",r="Eix X",s="No disponible",c={start:"Ajusta al primer punt de dades",end:"Ajusta a l'últim punt de dades"};var n={chartType:e,statistics:a,defaultTitle:i,countSeries:t,xAxis:r,notAvailable:s,timeAggregationTypes:c};

exports.chartType = e;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = s;
exports.statistics = a;
exports.timeAggregationTypes = c;
exports.xAxis = r;
