// @arcgis/charts-components/dist/cjs/messages.sk-7e8fa059-ad0402e1-71317954.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Graf",barChart:"Pruhový graf",lineChart:"Čiarový graf",pieChart:"Koláčový graf",histogram:"Histogram",scatterPlot:"Korelačný graf",boxPlot:"Krabicový graf"},t={mean:"Priemer",median:"Medián",sum:"Súčet",discretePercentile:"Diskrétny percentil",minimum:"Minimum",maximum:"Maximum",variance:"Rozptyl",count:"Počet",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Bez agregácie"},e="Graf",i="Počet sérií",r="Os X",s="Nie je k dispozícii",o={start:"Prichytiť k prvému dátovému bodu",end:"Prichytiť k poslednému dátovému bodu"};var n={chartType:a,statistics:t,defaultTitle:e,countSeries:i,xAxis:r,notAvailable:s,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = i;
exports['default'] = n;
exports.defaultTitle = e;
exports.notAvailable = s;
exports.statistics = t;
exports.timeAggregationTypes = o;
exports.xAxis = r;
