// @arcgis/charts-components/dist/cjs/messages.pl-1ed04c02-9de7edd4.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Diagram",barChart:"Diagram słupkowy grupowy",lineChart:"Diagram liniowy",pieChart:"Diagram kołowy",histogram:"Histogram",scatterPlot:"Diagram punktowy",boxPlot:"Diagram pudełkowy"},i={mean:"Średnia",median:"Mediana",sum:"Suma",discretePercentile:"Percentyl dyskretny",minimum:"Minimum",maximum:"Maksimum",variance:"Odchylenie",count:"Liczba",aggregation:"${ statistics } z ${ fieldName }",noAggregation:"Bez agregacji"},e="Wykres",t="Zlicz serie",r="Oś X",n="Niedostępne",o={start:"Dociągnij do pierwszego punktu danych",end:"Dociągnij do ostatniego punktu danych"};var s={chartType:a,statistics:i,defaultTitle:e,countSeries:t,xAxis:r,notAvailable:n,timeAggregationTypes:o};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = s;
exports.defaultTitle = e;
exports.notAvailable = n;
exports.statistics = i;
exports.timeAggregationTypes = o;
exports.xAxis = r;
