// @arcgis/charts-components/dist/cjs/messages.hr-5657371a-33d2d2e4-606f877a.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Došlo je do pogreške tijekom učitavanja grafikona.",e="Ne može se primijeniti transformacija dnevnika na negativne ili nulte vrijednosti.",i="Ne može se primijeniti transformacija kvadratnog korijena na negativne vrijednosti.",r="Došlo je do pogreške tijekom učitavanja sloja. URL = ${ url }. Id stavke portala = ${ portalItemId }.",o="Polja [${ fieldNames }] koja su definirana u konfiguraciji nisu pronađena u sloju.",t="${ dataPath } mora biti jedinstveno. Serija naziva ${ seriesName } ima ID (${ seriesID }) koji već koristi druga serija.",n="${ dataPath } ne može izvoditi nebrojeno agregiranje na ne-numeričkom polju.",s="${ dataPath } nedostaje svojstvo naziva ${ missingProperty }.",d="${ dataPath } mora biti kraće od ${ limit } znakova.",m="${ dataPath } ne smije imati manje od ${ limit } stavki.",l="${ dataPath } ne smije imati više od ${ limit } stavki.",j="${ dataPath } mora imati barem jedan znak koji nije razmak.",u="${ dataPath } ne smije imati ${ additionalProperty }.",v="${ dataPath } mora biti jednak jednoj od ovih dopuštenih vrijednosti: ${ allowedValues }.",p="${ dataPath } mora odgovarati shemi jednog od ovih: ${ schemaOptions }.",g="Raspršeni grafikoni s proporcionalnim simbolima nisu podržani. Primijenjena je zadana veličina simbola.",k="Nije uspjelo čitanje ulaznih podataka.",$="Očekivana vrsta serije po indeksu je ${ seriesIndex } je '${ expectedType }' ali primljeno je '${ receivedType }' umjesto toga",h="ili";var P={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:r,layerFieldsNotFound:o,duplicateSeriesID:t,nonNumericAggregation:n,requiredProperty:s,minLength:d,minItems:m,maxItems:l,whiteSpacePattern:j,additionalProperty:u,enumValues:v,anyOfValues:p,bubbleChartValidateMsg:g,queryError:k,invalidSeriesType:$,or:"ili"};

exports.additionalProperty = u;
exports.anyOfValues = p;
exports.bubbleChartValidateMsg = g;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = t;
exports.enumValues = v;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = o;
exports.layerLoadFailure = r;
exports.maxItems = l;
exports.minItems = m;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = n;
exports.or = h;
exports.queryError = k;
exports.requiredProperty = s;
exports.whiteSpacePattern = j;
