// @arcgis/charts-components/dist/cjs/messages.uk-7a3e2f41-284ded4d-0a801942.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Під час завантаження діаграми виникла помилка.",e="Неможливо застосувати перетворення логарифму до від’ємних або нульових значень.",t="Неможливо застосувати перетворення квадратного кореня до від’ємних значень.",r="Під час завантаження шару виникла помилка. URL = ${ url }. ІД елемента порталу = ${ portalItemId }.",i="Визначені у конфігурації поля [${ fieldNames }] не знайдено на шарі.",s="${ dataPath } повинен бути унікальним. Серія з назвою ${ seriesName } має ідентифікатор (${ seriesID }), який вже використовується іншою серією.",n="${ dataPath } не може виконувати агрегування без підрахунку в нечисловому полі.",o="${ dataPath } не має властивості з назвою ${ missingProperty }.",d="${ dataPath } не може мати менше ${ limit } символів.",l="${ dataPath } не може мати менше ${ limit } елементів.",u="${ dataPath } не може мати більше ${ limit } елементів.",m="${ dataPath } повинен мати принаймні один символ, який не є пробілом.",$="${ dataPath } не може мати ${ additionalProperty }.",p="${ dataPath } має дорівнювати одному з цих допустимих значень: ${ allowedValues }.",y="${ dataPath } має збігатися зі схемою одного з цих: ${ schemaOptions }.",P="Діаграми розсіювання з пропорційними символами не підтримуються. Застосовано розмір символу за замовчуванням.",g="Помилка читання вхідних даних.",h="Очікуваним типом серії з індексом ${ seriesIndex } є '${ expectedType }', але замість нього було отримано '${ receivedType }'",I="або";var V={defaultError:a,negativeValueInDataForLogTransformation:e,negativeValueInDataForSqrtTransformation:t,layerLoadFailure:r,layerFieldsNotFound:i,duplicateSeriesID:s,nonNumericAggregation:n,requiredProperty:o,minLength:d,minItems:l,maxItems:u,whiteSpacePattern:m,additionalProperty:$,enumValues:p,anyOfValues:y,bubbleChartValidateMsg:P,queryError:g,invalidSeriesType:h,or:"або"};

exports.additionalProperty = $;
exports.anyOfValues = y;
exports.bubbleChartValidateMsg = P;
exports['default'] = V;
exports.defaultError = a;
exports.duplicateSeriesID = s;
exports.enumValues = p;
exports.invalidSeriesType = h;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = r;
exports.maxItems = u;
exports.minItems = l;
exports.minLength = d;
exports.negativeValueInDataForLogTransformation = e;
exports.negativeValueInDataForSqrtTransformation = t;
exports.nonNumericAggregation = n;
exports.or = I;
exports.queryError = g;
exports.requiredProperty = o;
exports.whiteSpacePattern = m;
