// @arcgis/charts-components/dist/cjs/messages.vi-7bf62f1e-b3a7504e-68bf2b37.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a="Đã xảy ra lỗi khi tải biểu đồ.",t="Không thể áp dụng phép biến đổi log lên giá trị âm hoặc bằng không.",n="Không thể áp dụng phép biến đổi căn bậc hai lên giá trị âm.",e="Đã xảy ra lỗi khi tải lớp. URL = ${ url }. ID mục cổng thông tin = ${ portalItemId }.",i="Không tìm thấy trong lớp này các trường [${ fieldNames }] được xác định trong cấu hình.",h="${ dataPath } phải là duy nhất. Chuỗi có tên ${ seriesName } có một ID (${ seriesID }) đã được chuỗi khác sử dụng.",r="${ dataPath } không thể thực hiện tổng hợp không đếm trên trường không phải là số.",g="${ dataPath } đang thiếu một tên phù hợp ${ missingProperty }.",o="${ dataPath } không được ngắn hơn ${ limit } ký tự.",s="${ dataPath } không được chứa dưới ${ limit } mục.",l="${ dataPath } không được chứa nhiều hơn ${ limit } mục.",c="${ dataPath } phải chứa ít nhất một ký tự không phải khoảng trắng.",d="${ dataPath } không được chứa ${ additionalProperty }.",u="${ dataPath } phải bằng một trong những giá trị được phép này: ${ allowedValues }.",m="${ dataPath } phải khớp với lược đồ của một trong những thứ sau: ${ schemaOptions }.",p="Biểu đồ phân tán có ký hiệu tỷ lệ không được hỗ trợ. Đã áp dụng kích cỡ ký hiệu theo mặc định.",y="Không thể đọc dữ liệu đầu vào.",$="Loại chuỗi mong đợi tại chỉ mục ${ seriesIndex } là '${ expectedType }' nhưng thay vào đó lại nhận được '${ receivedType }'",k="hoặc";var P={defaultError:a,negativeValueInDataForLogTransformation:t,negativeValueInDataForSqrtTransformation:n,layerLoadFailure:e,layerFieldsNotFound:i,duplicateSeriesID:h,nonNumericAggregation:r,requiredProperty:g,minLength:o,minItems:s,maxItems:l,whiteSpacePattern:c,additionalProperty:d,enumValues:u,anyOfValues:m,bubbleChartValidateMsg:p,queryError:y,invalidSeriesType:$,or:k};

exports.additionalProperty = d;
exports.anyOfValues = m;
exports.bubbleChartValidateMsg = p;
exports['default'] = P;
exports.defaultError = a;
exports.duplicateSeriesID = h;
exports.enumValues = u;
exports.invalidSeriesType = $;
exports.layerFieldsNotFound = i;
exports.layerLoadFailure = e;
exports.maxItems = l;
exports.minItems = s;
exports.minLength = o;
exports.negativeValueInDataForLogTransformation = t;
exports.negativeValueInDataForSqrtTransformation = n;
exports.nonNumericAggregation = r;
exports.or = k;
exports.queryError = y;
exports.requiredProperty = g;
exports.whiteSpacePattern = c;
