// @arcgis/charts-components/dist/cjs/messages.sl-407c0dc1-b807e824.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Med nalaganjem grafikona je prišlo do napake.",a="Pretvorbe dnevnika ni mogoče uporabiti za negativne ali ničelne vrednosti.",i="Pretvorbe kvadratnega korena ni mogoče uporabiti za negativne vrednosti.",t="Med nalaganjem sloja je prišlo do napake. URL = ${ url }. ID elementa portala = ${ portalItemId }.",n="Polja [${ fieldNames }] opredeljena v konfiguraciji niso bila najdena v sloju.",o="${ dataPath } mora biti edinstveno. Serija z imenom ${ seriesName } ima ID (${ seriesID }), ki ga že uporablja druga serija.",r="${ dataPath } ne more izvajati združevanja brez števila na neštevilskem polju.",l="${ dataPath } manjka lastnost imenovana ${ missingProperty }.",s="${ dataPath } ne sme biti krajše od ${ limit } znakov.",d="${ dataPath } ne sme imeti manj kot ${ limit } elementov.",m="${ dataPath } ne sme imeti več kot ${ limit } elementov.",v="${ dataPath } mora vsebovati vsaj en znak, ki ni presledek.",p="${ dataPath } ne sme imeti ${ additionalProperty }.",j="${ dataPath } mora biti enak eni od teh dovoljenih vrednosti: ${ allowedValues }.",u="${ dataPath } se mora ujemati s shemo ene od teh: ${ schemaOptions }.",$="Raztreseni grafikoni s proporcionalnimi simboli niso podprti. Uporabljena je bila privzeta velikost simbolov.",k="Branje vhodnih podatkov ni uspelo.",g="Pričakovani tip niza z indeksom ${ seriesIndex } je '${ expectedType }', ampak '${ receivedType }' je bil prejet namesto",h="ali";var b={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:i,layerLoadFailure:t,layerFieldsNotFound:n,duplicateSeriesID:o,nonNumericAggregation:r,requiredProperty:l,minLength:s,minItems:d,maxItems:m,whiteSpacePattern:v,additionalProperty:p,enumValues:j,anyOfValues:u,bubbleChartValidateMsg:$,queryError:k,invalidSeriesType:g,or:"ali"};

exports.additionalProperty = p;
exports.anyOfValues = u;
exports.bubbleChartValidateMsg = $;
exports['default'] = b;
exports.defaultError = e;
exports.duplicateSeriesID = o;
exports.enumValues = j;
exports.invalidSeriesType = g;
exports.layerFieldsNotFound = n;
exports.layerLoadFailure = t;
exports.maxItems = m;
exports.minItems = d;
exports.minLength = s;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = i;
exports.nonNumericAggregation = r;
exports.or = h;
exports.queryError = k;
exports.requiredProperty = l;
exports.whiteSpacePattern = v;
