// @arcgis/charts-components/dist/cjs/messages.ar-01382838-4fee603c-61dd772b.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const t={chart:"الرسم البياني",barChart:"المخطط الشريطي",lineChart:"مخطط خطي",pieChart:"مخطط دائري",histogram:"مدرج تكراري",scatterPlot:"Scatterplot",boxPlot:"رسم مربع"},a={mean:"المتوسط",median:"متوسط",sum:"مجموع",discretePercentile:"النسبة المئوية المنفصلة",minimum:"الحد الأدنى",maximum:"الحد الأقصى",variance:"تنوع",count:"عدد",aggregation:"${ statistics } من ${ fieldName }",noAggregation:"لا يوجد تجميع"},e="الرسم البياني",i="سلسلة العدد",s="محور س",r="غير مرئي",n={start:"انطباق على نقطة البيانات الأولى",end:"انطباق على نقطة البيانات الأخيرة"};var o={chartType:t,statistics:a,defaultTitle:e,countSeries:i,xAxis:s,notAvailable:r,timeAggregationTypes:n};

exports.chartType = t;
exports.countSeries = i;
exports['default'] = o;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = a;
exports.timeAggregationTypes = n;
exports.xAxis = s;
