// @arcgis/charts-components/dist/cjs/messages.it-f3a3c18c-092ffc47.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Grafico",barChart:"Grafico a barre",lineChart:"Grafico a linee",pieChart:"Grafico a torta",histogram:"Istogramma",scatterPlot:"Grafico a dispersione",boxPlot:"Box plot"},i={mean:"Media",median:"Mediana",sum:"Somma",discretePercentile:"Percentile discreto",minimum:"Minimo",maximum:"Massimo",variance:"Varianza",count:"Conteggio",aggregation:"${ statistics } di ${ fieldName }",noAggregation:"Nessuna aggregazione"},e="Grafico",t="Serie Conteggio",o="Asse X",r="Non disponibile",n={start:"Snap al primo punto dati",end:"Snap al punto dati finale"};var s={chartType:a,statistics:i,defaultTitle:e,countSeries:t,xAxis:o,notAvailable:r,timeAggregationTypes:n};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = s;
exports.defaultTitle = e;
exports.notAvailable = r;
exports.statistics = i;
exports.timeAggregationTypes = n;
exports.xAxis = o;
