// @arcgis/charts-components/dist/cjs/messages.pt-PT-e9eb0f86-901ba69d-252215b8.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const a={chart:"Gráfico",barChart:"Gráfico de barras",lineChart:"Gráfico de linhas",pieChart:"Gráfico circular",histogram:"Histograma",scatterPlot:"Gráfico de dispersão",boxPlot:"Gráfico de caixas"},e={mean:"Média",median:"Mediana",sum:"Soma",discretePercentile:"Percentil discreto",minimum:"Mínimo",maximum:"Máximo",variance:"Variação",count:"Contagem",aggregation:"${ statistics } de ${ fieldName }",noAggregation:"Sem agregação"},i="Gráfico",t="Séries de Contagem",o="Eixo X",r="Não disponível",s={start:"Ajustar ao primeiro ponto de dados",end:"Ajustar ao último ponto de dados"};var n={chartType:a,statistics:e,defaultTitle:i,countSeries:t,xAxis:o,notAvailable:r,timeAggregationTypes:s};

exports.chartType = a;
exports.countSeries = t;
exports['default'] = n;
exports.defaultTitle = i;
exports.notAvailable = r;
exports.statistics = e;
exports.timeAggregationTypes = s;
exports.xAxis = o;
