// @arcgis/charts-components/dist/cjs/messages.pt-BR-56a76dfb-d3f9be29-e4f82b94.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const e="Ocorreu um erro ao carregar o gráfico.",a="Não é possível aplicar transformação de log para valores negativos ou zero.",o="Não é possível aplicar transformação de raiz quadrada para valores negativos.",r="Ocorreu um erro ao carregar a camada. URL = ${ url }. D de item do portal = ${ portalItemId }.",t="Os campos [${ fieldNames }] definidos na configuração não foram encontrados na camada.",d="${ dataPath } deve ser único. A série denominada ${ seriesName } tem um id (${ seriesID }) que já é usado por outra série.",s="${ dataPath } não deve realizar agregação sem contagem em campo não numérico.",i="${ dataPath } está sem uma propriedade denominada ${ missingProperty }.",n="${ dataPath } não deve ser menor que ${ limit } caracteres.",m="${ dataPath } não deve ter menos que ${ limit } itens.",l="${ dataPath } não deve ter mais de ${ limit } itens.",u="${ dataPath } deve ter pelo menos um caractere diferente de espaço em branco.",p="${ dataPath } não deve ter ${ additionalProperty }.",c="${ dataPath } deve ser igual a um destes valores permitidos: ${ allowedValues }.",$="${ dataPath } deve corresponder ao esquema de um destes: ${ schemaOptions }.",g="Gráficos de dispersão com símbolos proporcionais não são suportados. O tamanho padrão do símbolo foi aplicado.",v="Falha ao ler os dados de entrada.",f="O tipo de série esperado no índice${ seriesIndex } é '${ expectedType }' mas '${ receivedType }' foi recebido em seu lugar",h="ou";var P={defaultError:e,negativeValueInDataForLogTransformation:a,negativeValueInDataForSqrtTransformation:o,layerLoadFailure:r,layerFieldsNotFound:t,duplicateSeriesID:d,nonNumericAggregation:s,requiredProperty:i,minLength:n,minItems:m,maxItems:l,whiteSpacePattern:u,additionalProperty:p,enumValues:c,anyOfValues:$,bubbleChartValidateMsg:g,queryError:v,invalidSeriesType:f,or:"ou"};

exports.additionalProperty = p;
exports.anyOfValues = $;
exports.bubbleChartValidateMsg = g;
exports['default'] = P;
exports.defaultError = e;
exports.duplicateSeriesID = d;
exports.enumValues = c;
exports.invalidSeriesType = f;
exports.layerFieldsNotFound = t;
exports.layerLoadFailure = r;
exports.maxItems = l;
exports.minItems = m;
exports.minLength = n;
exports.negativeValueInDataForLogTransformation = a;
exports.negativeValueInDataForSqrtTransformation = o;
exports.nonNumericAggregation = s;
exports.or = h;
exports.queryError = v;
exports.requiredProperty = i;
exports.whiteSpacePattern = u;
