// @arcgis/charts-components/dist/cjs/t9n-strings-ceef62aa.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See LICENSE.md for details.
 * v4.32.1
 */
'use strict';

const restJsObjectLiterals = require('./rest-js-object-literals-acb59afb.js');
const index = require('./index-2e64cfc1.js');

/**
 * Fetches strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchStrings(props) {
    const { htmlElement, relativePath, filePrefix } = props;
    const assetsPath = index.getAssetPath(relativePath);
    return await restJsObjectLiterals.Dn({
        element: htmlElement,
        baseURL: assetsPath,
        filePrefix,
    });
}
/**
 * Fetches the error strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchErrorsStrings(htmlElement) {
    return await fetchStrings({
        htmlElement,
        relativePath: "assets/errors",
        filePrefix: "errors",
    });
}
/**
 * Fetches the error strings for a component
 * @param htmlElement The component's HTML element
 */
async function fetchCommonStrings(htmlElement) {
    return await fetchStrings({
        htmlElement,
        relativePath: "assets/common",
        filePrefix: "common",
    });
}

exports.fetchCommonStrings = fetchCommonStrings;
exports.fetchErrorsStrings = fetchErrorsStrings;
